
<!--==========================================================================
	FTL macros 
	========================================================================== -->


<!--==========================================================================
	CSS styles 
	========================================================================== -->


<style type="text/css">

body {
	background-color: #dddddd;
}

/* Table styles. */

table {
  border-color: #000;
  border-spacing: 0px;
  border-style: solid;
  border-width: 1px;
  cell-spacing: 0px;
}

.noBorder {
  border-width: 0px;
}

td, th {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 10pt;
  padding: 2px 0.5em;
  white-space: nowrap;
}

td.numeric {
  text-align: right;
}

th {
  background-color: #c0c0c0;
}

th.mainHeader {
  background-color: #808080;
  color: #ffffff;
  text-align: left;
}

th a {
  color: #000080;
  text-decoration: none;
}

th a:visited {
  color: #000080;
}

th a:active, th a:hover {
  color: #800000;
  text-decoration: underline;
}

.toc {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 95%;
}

.note {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 90%;
	text-align: left;
}

.main {
	margin-top: 15px;
	width: 1000px;
	text-align: left;
	background: #ffffff;
	border: 3px solid #777777;
}

.histo {
	table-layout:fixed;
	width:100%;
	border:1px solid #aaa;
	word-wrap:break-word;
}

</style>


<!--==========================================================================
	Index
	========================================================================== -->

<center>
<div class="main">

<center> <h3> SnpEff: Variant analysis </h3> </center>

<div style="margin-left: .5em">
<table class="toc"><tr><td>
	<center><b>Contents</b></center>
	<a href="#summary">Summary</a><br>
	<a href="#changeRateChr"> Variant rate by chromosome</a><br>
	<a href="#changesByType">Variants by type</a><br>
	<a href="#effectsImpact"> Number of variants by impact </a><br>
	<a href="#effectsImpact"> Number of variants by functional class </a><br>
	<a href="#effects"> Number of variants by effect </a><br>
	<a href="#quality">Quality histogram</a><br>
	<a href="#indels">InDel length histogram</a><br>
	<a href="#baseChages">Base variant table</a><br>
	<a href="#tstv">Transition vs transversions (ts/tv)</a><br>
	<a href="#alleleFreq"> Allele frequency </a><br>
	<a href="#alleleCount"> Allele Count </a><br>
	<a href="#codonChanges"> Codon change table </a><br>
	<a href="#aaChanges"> Amino acid change table </a><br>
	<a href="#chrChanges"> Chromosome variants plots </a><br>
	<a href="snpEff_genes.txt"> Details by gene </a><br>
</tr></td></table>
</div>


<!--==========================================================================
	Summary table 
	========================================================================== -->

<hr>
<a name="summary">
<center>
<b>Summary</b><p>

<table border=0>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Genome </b> </td>
		<td> test_ENSG00000158062_spaces </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Date </b> </td>
		<td> 2014-12-06 18:55 </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> SnpEff version </b> </td>
		<td> <pre>SnpEff 4.1alpha (build 2014-12-06), by Pablo Cingolani</pre> </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Command line arguments </b> </td>
		<td> <pre>SnpEff  test_ENSG00000158062_spaces tests/vcf_genes_spaces.vcf 
</pre> </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Warnings </b> </td>
		   
		<td bgcolor="#ffff00"> 26 </td>
	</tr>	
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Errors </b> </td>
		 
		<td bgcolor="#dddddd"> 0 </td>
	</tr>	
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of lines (input file) </b> </td>
		<td> 4 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants (before filter) </b> </td>
		<td> 4 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of not variants <br>(i.e. reference equals alternative) </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants processed <br> (i.e. after filter and non-variants) </b> </td>
		<td> 4 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of known variants <br>(i.e. non-empty ID) </b> </td>
		<td> 
		1
		( 25% ) 
		</td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of multi-allelic VCF entries <br>(i.e. more than two alleles) </b> </td>
		<td> 
		0
		</td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of effects </b> </td>
		<td> 55 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Genome total length </b> </td>
		<td> 3,095,693,983 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Genome effective length </b> </td>
		<td> 249,250,621 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Variant rate </b> </td>
		<td> 1 variant every 62,312,655 bases </td>
	</tr>
</table>
<p>
</center>

<!--==========================================================================
	Variant rate by chromosome
	========================================================================== -->

<hr> 
<a name="changeRateChr"> 

<center>
<b> Variants rate details </b><p>

<table border=1>
	<tr><th> Chromosome </th><th> Length </th><th> Variants </th><th> Variants rate </th></tr> 
		<tr>
			<td> 1 </td>
			<td class="numeric"> 249,250,621 </td>
			<td class="numeric"> 4 </td>
			<td class="numeric"> 62,312,655 </td>
		</tr> 
	<tr>
		<th> Total </th>
		<th class="numeric"> 249,250,621 </th>
		<th class="numeric"> 4 </th>
		<th class="numeric"> 62,312,655 </th>
	</tr> 
</table> 
</center>

<!--==========================================================================
	Variantss by type
	========================================================================== -->

<hr> 
<a name="changesByType"> 
<center>
<b> Number variantss by type</b><p>

<table border=1>
	<thead>
		<tr>
			<th> <b> Type   </b> </th>
			<th> <b> Total  </b> </th>
		</tr>
	</thead>
    <tbody>
	    <tr>
	    	<td> <b> SNP </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 4 </td>
	    </tr>
	    <tr>
	    	<td> <b> MNP </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> INS </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> DEL </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> MIXED </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> INTERVAL </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
    </tbody>
    <tfoot>
	    <tr>
	    	<th><b>Total </b> </th>
	    	<th class="numeric"> 4 </th>
	    </tr>
    </tfoot>
</table>
</center>

<!--==========================================================================
	Effects by impact
	========================================================================== -->

<hr> 
<a name="effectsImpact"> 
<center>
<b> Number of effects by impact </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> HIGH </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#36c800"> 9 </td> 
			<td class="numeric" bgcolor="#36c800"> 16.364% </td>
		</tr>
		<tr> 
			<td> <b> LOW </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#36c800"> 9 </td> 
			<td class="numeric" bgcolor="#36c800"> 16.364% </td>
		</tr>
		<tr> 
			<td> <b> MODERATE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 2 </td> 
			<td class="numeric" bgcolor="#00ff00"> 3.636% </td>
		</tr>
		<tr> 
			<td> <b> MODIFIER </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 35 </td> 
			<td class="numeric" bgcolor="#ff0000"> 63.636% </td>
		</tr>
</table><br>
<p>
</center>

<!--==========================================================================
	Effects by functional class
	========================================================================== -->

<hr> 
<a name="effectsFuncClass"> 
<center>
<b> Number of effects by functional class </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> MISSENSE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 2 </td> 
			<td class="numeric" bgcolor="#00ff00"> 25% </td>
		</tr>
		<tr> 
			<td> <b> NONSENSE </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 4 </td> 
			<td class="numeric" bgcolor="#ff0000"> 50% </td>
		</tr>
		<tr> 
			<td> <b> SILENT </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 2 </td> 
			<td class="numeric" bgcolor="#00ff00"> 25% </td>
		</tr>
</table><br>
<p>

Missense / Silent ratio: </th><td class="numeric"> 1

</center>

<!--==========================================================================
	Effects stats
	========================================================================== -->

<hr> 
<a name="effects"> 
<center>
<b> Number of effects by type and region </b> <p> 

<table border=0>
	<tr>
		<th> Type </th>
		<th> Region </th>
	</tr>
	<tr>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> downstream_gene_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#639b00"> 10 </td> 
			<td class="numeric" bgcolor="#639b00"> 18.182% </td>
		</tr>
		<tr> 
			<td> <b> intron_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 24 </td> 
			<td class="numeric" bgcolor="#ff0000"> 43.636% </td>
		</tr>
		<tr> 
			<td> <b> missense_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#0bf300"> 2 </td> 
			<td class="numeric" bgcolor="#0bf300"> 3.636% </td>
		</tr>
		<tr> 
			<td> <b> splice_donor_variant&intron_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#2cd200"> 5 </td> 
			<td class="numeric" bgcolor="#2cd200"> 9.091% </td>
		</tr>
		<tr> 
			<td> <b> splice_region_variant&intron_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#42bc00"> 7 </td> 
			<td class="numeric" bgcolor="#42bc00"> 12.727% </td>
		</tr>
		<tr> 
			<td> <b> stop_gained </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#21dd00"> 4 </td> 
			<td class="numeric" bgcolor="#21dd00"> 7.273% </td>
		</tr>
		<tr> 
			<td> <b> synonymous_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#0bf300"> 2 </td> 
			<td class="numeric" bgcolor="#0bf300"> 3.636% </td>
		</tr>
		<tr> 
			<td> <b> upstream_gene_variant </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 1.818% </td>
		</tr>
</table><br>
 </td>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> DOWNSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#639b00"> 10 </td> 
			<td class="numeric" bgcolor="#639b00"> 18.182% </td>
		</tr>
		<tr> 
			<td> <b> EXON </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#4db100"> 8 </td> 
			<td class="numeric" bgcolor="#4db100"> 14.545% </td>
		</tr>
		<tr> 
			<td> <b> INTRON </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#ff0000"> 24 </td> 
			<td class="numeric" bgcolor="#ff0000"> 43.636% </td>
		</tr>
		<tr> 
			<td> <b> SPLICE_SITE_DONOR </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#2cd200"> 5 </td> 
			<td class="numeric" bgcolor="#2cd200"> 9.091% </td>
		</tr>
		<tr> 
			<td> <b> SPLICE_SITE_REGION </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#42bc00"> 7 </td> 
			<td class="numeric" bgcolor="#42bc00"> 12.727% </td>
		</tr>
		<tr> 
			<td> <b> UPSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="#00ff00"> 1 </td> 
			<td class="numeric" bgcolor="#00ff00"> 1.818% </td>
		</tr>
</table><br>
 </td>
	</tr>
</table>
<p>

<img src="http://chart.apis.google.com/chart?cht=bvs&chbh=50,4,8&chs=800x300&chxt=y,y,x,x&chtt=Variations&chxp=1,50.0|3,50.0&chxr=0,0.0,43.63636363636363|1,0.0,100.0|3,0.0,100.0&chco=00FFFF,FA8072,FF0000,A52A2A,FFA500,DEB887,E9967A,7FFFD4,808080&chdl=Upstream|5%27UTR|Exon|Splice+Donor|Intron|Splice+Acceptor|3%27UTR|Downstream|Intergenic&chxl=1:|%25|2:|Intergenic|Up|5%27UTR|Exon|Donor|Intron|Acceptor|Exon|3%27UTR|Down|Intergenic|3:|&chd=e:AACrAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAVVAAAAAAVVAAAAAA,AAAAAAAANVAAAAAAAAAAAA,AAAAAAAAAA..AAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAaqAA,AAAAAAAAAAAAAAAAAAAAAA" border=1><p>

</center>

<!--==========================================================================
	Quality
	========================================================================== -->

<hr> 
<a name="quality"> 
<b> Quality:</b> 
<p>
<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>40</td> </tr>
		<tr> <th>Max</th><td>2,373</td> </tr>
		<tr> <th>Mean</th><td>650.75</td> </tr>
		<tr> <th>Median</th><td>95</td> </tr>
		<tr> <th>Standard deviation</th><td>1,148.656</td> </tr>
		<tr> <th>Values</th><td>40,69,121,2373</td> </tr>
		<tr> <th>Count</th><td>1,1,1,1</td> </tr>
		</tr>
	</table>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chd=e:........&chxp=1,50.0|3,50.0&chtt=Quality+histogram&chxl=1:|Count|2:|40|||2373|3:|Quality&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
</pre>

<!--==========================================================================
	InDels
	========================================================================== -->

<hr> 
<a name="indels"> 
<b> Insertions and deletions length:</b> 
<p>
<pre>
	<img src="http://chart.apis.google.com/chart?chd=e:&cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chxl=1:|Count|3:|Length&chtt=Insertion+deletion+length+histogram&chxr=0,2.147483647E9,-2.147483648E9|1,0.0,100.0|3,0.0,100.0"><br>
</pre>

<!--==========================================================================
	Base changes
	========================================================================== -->

<hr> 
<a name="baseChages"> 
<center>
<b> Base changes (SNPs) </b> <p>

<table border=1>
    <tr>
    	<td> &nbsp; </td>  <th> <b> A </b> </th> <th> <b> C </b> </th> <th> <b> G </b> </th> <th> <b> T </b> </th>
    </tr>
	    <tr> <th> <b> A </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="#00ff00"> 1 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> C </b> </th><td class="numeric" bgcolor="#00ff00"> 1 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> G </b> </th><td class="numeric" bgcolor="#ff0000"> 2 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> T </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
</table>
</center>

<p>

<!--==========================================================================
	Ts/Tv
	========================================================================== -->

<hr> 
<a name="tstv">
<center> <b> Ts/Tv (transitions / transversions) </b> <p> </center>

<small>
<b>Note:</b> Only SNPs are used for this statistic.<br>
<b>Note:</b> This Ts/Tv ratio is a 'raw' ratio (ratio of observed events).
</small>
<p>

<center>
<table border=1>
	<tr> <th> Transitions </th><td class="numeric"> 3 </td> </tr>
	<tr> <th> Transversions </th><td class="numeric"> 1 </td> </tr>
	<tr> <th> Ts/Tv ratio </th><td class="numeric"> 3 </td> </tr>
</table>
</center>
<p>

	<b>All variants:</b>
	<pre>Sample ,Total
Transitions ,3,3
Transversions ,1,1
Ts/Tv ,3.000,3.000
</pre>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
	<pre>Sample ,Total
Transitions ,0,0
Transversions ,1,1
Ts/Tv ,0.000,0.000
</pre>
<p>

<!--==========================================================================
	Allele frequency
	========================================================================== -->

<hr> 
<a name="alleleFreq">
<center> <b> Allele frequency </b> <p> </center>

	<img src="http://chart.apis.google.com/chart?chd=e:&cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Allele+Frequency+%25&chxl=1:|Count|3:|Alleles+frequency&chxr=0,2.147483647E9,-2.147483648E9|1,0.0,100.0|3,0.0,100.0"><br>
<p>
	
<!--==========================================================================
	Allele Count
	========================================================================== -->

<hr> 
<a name="alleleCount">
<center> <b> Allele Count </b> <p> </center>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chtt=Allele+Count&chxp=1,50.0|3,50.0&chxl=1:|Count|2:|1|3:|Allele+Count&chxr=0,0.0,4.0|1,0.0,100.0|3,0.0,100.0"><br>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>1</td> </tr>
		<tr> <th>Max</th><td>1</td> </tr>
		<tr> <th>Mean</th><td>1</td> </tr>
		<tr> <th>Median</th><td>1</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>1</td> </tr>
		<tr> <th>Count</th><td>4</td> </tr>
		</tr>
	</table>
<p>


<!--==========================================================================
	Hom/Het table
	========================================================================== -->

<hr> 
<a name="homHet">
<center> <b> Hom/Het per sample </b> <p> </center>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:AA&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Heterozygous+genotypes&chxr=0,0.0,0.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Heterozygous+count|2:|0|3:|Sample+number"><br>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:AA&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Homozygous+%28ALT%29+genotypes&chxr=0,0.0,0.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Homozygous+count|2:|0|3:|Sample+number"><br>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:AA&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Missing+genotypes&chxr=0,0.0,0.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Missing+count|2:|0|3:|Sample+number"><br>
	
	<pre>Sample_names , No_sample_name_found
Reference , 0
Het , 0
Hom , 0
Missing , 0
</pre>
<p>

<!--==========================================================================
	Codon change table
	========================================================================== -->
	
<hr> 
<a name="codonChanges">
<center>
<b> Codon changes</b> <p>

	<div class="note">
		How to read this table: <br>
		- Rows are reference codons and columns are changed codons. E.g. Row 'AAA' column 'TAA' indicates how many 'AAA' codons have been replaced by 'TAA' codons.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

  
<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> ACG </th>  <th> ATG </th>  <th> CAG </th>  <th> CGA </th>  <th> GCC </th>  <th> GCT </th>  <th> TAG </th>  <th> TGA </th> 
		</tr>
	</thead>
			<tr> 
				<th> ACG </th>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> ATG </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> CAG </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> CGA </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
			</tr>
			<tr> 
				<th> GCC </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> GCT </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> TAG </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> TGA </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Amino acid change table
	========================================================================== -->

<hr> 
<a name="aaChanges">
<center>
<b> Amino acid changes</b><p> 

	<div class="note">
		How to read this table: <br>
		- Rows are reference amino acids and columns are changed amino acids. E.g. Row 'A' column 'E' indicates how many 'A' amino acids have been replaced by 'E' amino acids.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> * </th>  <th> A </th>  <th> M </th>  <th> Q </th>  <th> R </th>  <th> T </th> 
		</tr>
	</thead>
			<tr> 
				<th> * </th>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> A </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> 2 </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> M </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> Q </th>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> R </th>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
			</tr>
			<tr> 
				<th> T </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> 2 </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Chromosome variants table
	========================================================================== -->

<hr> 
<a name="chrChanges">
<center>
<b> Variants by chromosome</b><p> 

<center>
	<pre>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Variants+histogram%3A+1&chxr=0,0.0,4.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Variants%2F1000Kb|2:|0|||||||||||||||||||||||||25000000|||||||||||||||||||||||||50000000|||||||||||||||||||||||||75000000|||||||||||||||||||||||||100000000|||||||||||||||||||||||||125000000|||||||||||||||||||||||||150000000|||||||||||||||||||||||||175000000|||||||||||||||||||||||||200000000|||||||||||||||||||||||||225000000||||||||||||||||||||||||249000000|3:|Position&chd=e:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA..AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"><br>
		1, Position,0,1000000,2000000,3000000,4000000,5000000,6000000,7000000,8000000,9000000,10000000,11000000,12000000,13000000,14000000,15000000,16000000,17000000,18000000,19000000,20000000,21000000,22000000,23000000,24000000,25000000,26000000,27000000,28000000,29000000,30000000,31000000,32000000,33000000,34000000,35000000,36000000,37000000,38000000,39000000,40000000,41000000,42000000,43000000,44000000,45000000,46000000,47000000,48000000,49000000,50000000,51000000,52000000,53000000,54000000,55000000,56000000,57000000,58000000,59000000,60000000,61000000,62000000,63000000,64000000,65000000,66000000,67000000,68000000,69000000,70000000,71000000,72000000,73000000,74000000,75000000,76000000,77000000,78000000,79000000,80000000,81000000,82000000,83000000,84000000,85000000,86000000,87000000,88000000,89000000,90000000,91000000,92000000,93000000,94000000,95000000,96000000,97000000,98000000,99000000,100000000,101000000,102000000,103000000,104000000,105000000,106000000,107000000,108000000,109000000,110000000,111000000,112000000,113000000,114000000,115000000,116000000,117000000,118000000,119000000,120000000,121000000,122000000,123000000,124000000,125000000,126000000,127000000,128000000,129000000,130000000,131000000,132000000,133000000,134000000,135000000,136000000,137000000,138000000,139000000,140000000,141000000,142000000,143000000,144000000,145000000,146000000,147000000,148000000,149000000,150000000,151000000,152000000,153000000,154000000,155000000,156000000,157000000,158000000,159000000,160000000,161000000,162000000,163000000,164000000,165000000,166000000,167000000,168000000,169000000,170000000,171000000,172000000,173000000,174000000,175000000,176000000,177000000,178000000,179000000,180000000,181000000,182000000,183000000,184000000,185000000,186000000,187000000,188000000,189000000,190000000,191000000,192000000,193000000,194000000,195000000,196000000,197000000,198000000,199000000,200000000,201000000,202000000,203000000,204000000,205000000,206000000,207000000,208000000,209000000,210000000,211000000,212000000,213000000,214000000,215000000,216000000,217000000,218000000,219000000,220000000,221000000,222000000,223000000,224000000,225000000,226000000,227000000,228000000,229000000,230000000,231000000,232000000,233000000,234000000,235000000,236000000,237000000,238000000,239000000,240000000,241000000,242000000,243000000,244000000,245000000,246000000,247000000,248000000,249000000
1,Count,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0

	</pre>
</center>

<!--==========================================================================
	Variants by gene 
	========================================================================== -->

<center>
	<hr>
	<b> Details by gene </b><p>
	<p>
	<b><a href="snpEff_genes.txt">Here</a></b> you can find a tab-separated table.
</center>
</div>
</center>