<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>SnpEff</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Styles -->
    <link href="css/bootstrap.css" rel="stylesheet">
    <link href="css/bootstrap-responsive.css" rel="stylesheet">
    <link href="css/docs.css" rel="stylesheet">
    <link href="js/google-code-prettify/prettify.css" rel="stylesheet">
    <link href="css/my.css" rel="stylesheet">

    <!-- hTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="js/html5shiv.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="ico/apple-touch-icon-144-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="ico/apple-touch-icon-114-precomposed.png">
      <link rel="apple-touch-icon-precomposed" sizes="72x72" href="ico/apple-touch-icon-72-precomposed.png">
                    <link rel="apple-touch-icon-precomposed" href="ico/apple-touch-icon-57-precomposed.png">
                                   <link rel="shortcut icon" href="ico/favicon.png">

  </head>

  <body data-spy="scroll" data-target=".bs-docs-sidebar">

    <!-- Navbar
    ================================================== -->
    <div class="navbar navbar-inverse navbar-fixed-top">
      <div class="navbar-inner">
        <div class="container-fluid">
          <button type="button" class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="brand" href="#">SnpEff</a>
          <div class="nav-collapse collapse">
            <p class="navbar-text pull-right">
            </p>
            <ul class="nav">
              <li class="active"><a href="index.html">Home</a></li>
              <li><a href="download.html">Download</a></li>
              <li><a href="SnpEff_manual.html">Documentation</a></li>
              <li><a href="SnpSift.html">SnpSift</a></li>
              <li><a href="about.html">About</a></li>
            </ul>
          </div><!--/.nav-collapse -->
        </div>
      </div>
    </div>

<!-- Subhead
================================================== -->
<header class="jumbotron subhead" id="overview">
  <div class="container">
    <h1>SnpEff</h1>
    <p class="lead">Documentation</p>
  </div>
</header>


  <div class="container">

    <!-- Docs nav
    ================================================== -->
    <div class="row">
      <div class="span3 bs-docs-sidebar">
        <ul class="nav nav-list bs-docs-sidenav">
          <li><a href="#intro"><i class="icon-chevron-right"></i> Introduction</a></li>
          <li><a href="#examples"><i class="icon-chevron-right"></i>Usage examples (for the impatient) </a></li>
          <li><a href="#install"><i class="icon-chevron-right"></i> Download & Install</a></li>
          <li><a href="#run"><i class="icon-chevron-right"></i> Running SnpEff </a></li>
          <li><a href="#seqData"><i class="icon-chevron-right"></i> Sequence data analysis </a></li>
          <li><a href="#input"><i class="icon-chevron-right"></i> Input files </a></li>
          <li><a href="#output"><i class="icon-chevron-right"></i> Output files </a></li>
          <li><a href="#outputSummary"><i class="icon-chevron-right"></i> Output Summary file </a></li>
          <li><a href="#eff"><i class="icon-chevron-right"></i> Effect prediction details </a></li>
          <li><a href="#effNc"><i class="icon-chevron-right"></i> Regulatory and Non-coding annotations </a></li>
          <li><a href="#filters"><i class="icon-chevron-right"></i> Annotations options </a></li>
          <li><a href="#cancer"><i class="icon-chevron-right"></i> Cancer samples </a></li>
          <li><a href="#ENCODE"><i class="icon-chevron-right"></i> Encode </a></li>
          <li><a href="#Epigenome"><i class="icon-chevron-right"></i> Epigenome project </a></li>
          <li><a href="#NextProt"><i class="icon-chevron-right"></i> NextProt </a></li>
          <li><a href="#Motif"><i class="icon-chevron-right"></i> Motif </a></li>
          <li><a href="#databases"><i class="icon-chevron-right"></i> Building databases</a></li>
          <li><a href="#databasesNc"><i class="icon-chevron-right"></i> Building databases: Regulatory and Non-coding</a></li>
          <li><a href="#gatk"><i class="icon-chevron-right"></i> Integration: GATK</a></li>
          <li><a href="#galaxy"><i class="icon-chevron-right"></i> Integration: Galaxy</a></li>
          <li><a href="#utils"><i class="icon-chevron-right"></i>Commands and utilities </a></li>
        </ul>
      </div>
      <div class="span9">


        <!-- Introduction
        ================================================== -->
        <section id="intro">
          <div class="page-header">
            <h1>1. Introduction</h1>
          </div>
				<p class="lead">
				SnpEff is a variant annotation and effect prediction tool.
				It annotates and predicts the effects of genetic variants (such as amino acid changes).
				</p>

				<p> <h3> License </h3>
				SnpEff is open source. It is released as "LGPLv3". <br>

				<p> <h3>Summary</h3>
				A typical SnpEff use case would be:
				<ul>
					<li> Input:
					The inputs are predicted variants (SNPs, insertions, deletions and MNPs). 
					The input file is usually obtained as a result of a sequencing experiment, and it is usually in variant call format (VCF).
					<p>

					<li> Output:
					SnpEff analyzes the input variants. 
					It annotates the variants and calculates the effects they produce on known genes (e.g. amino acid changes).
					A list of effects and annotations that SnpEff can calculate can be found <a class="body" href="faq.html#What_effects_are_predicted?">here</a>.
				</ul>
				</p>

				<p> <h3> Variants </h3>
				By genetic variant we mean difference between a genome and a "reference" genome.
				As an example, imagine we are sequencing a "sample".
				Here "sample" can mean anything that you are interested in studying, from a cell culture, to a mouse or a cancer patient.
				<br>
				It is a standard procedure to compare your sample sequences against the corresponding "reference genome".
				For instance you may compare the cancer patient genome against the current "reference genome" (at the time of this writing, it is 'hg19').
				<p>
				In a typical sequencing experiment, you will find many places in the genome where your sample differs from the reference genome.
				These are called "genomic variants" or just "variants".
				<br>
				Typically, variants are categorized as follows:
				<table class="table table-striped">
					<tr> <th> Name  </th> <th> What is means                    </th> <th> Example </th> </tr>
					<tr> <td> SNP   </td> <td> Single-Nucleotide Polymorphism   </td> <td> Reference = 'A', Sample = 'C' </td> </tr>
					<tr> <td> Ins   </td> <td> Insertion                        </td> <td> Reference = 'A', Sample = 'AGT' </td> </tr>
					<tr> <td> Del   </td> <td> Deletion                         </td> <td> Reference = 'AC', Sample = 'C' </td> </tr>
					<tr> <td> MNP   </td> <td> Multiple-nucleotide polymorphism </td> <td> Reference = 'ATA', Sample = 'GTC' </td> </tr>
				</table>
				This is not a comprehensive list, it is just to give you an idea.
				</p>

				<p> <h3>Annotations</h3>
				So, you have a huge file describing all the differences between your sample and the reference genome.
				But you want to know more about these variants than just their genetic coordinates. 
				E.g.: Are they in a gene? In an exon? Do they change protein coding? Do they cause premature stop codons?
				<br>
				SnpEff can help you answer all these questions.
				The process of adding this information about the variants is called "Annotation".
				<br>
				SnpEff provides several degrees of annotations, from simple (e.g. which gene is each variant affecting) to extremely complex annotations (e.g. will this non-coding variant affect the expression of a gene?).
				It should be noted that the more complex the annotations, the more it relies in computational predictions.
				Such computational predictions are sometimes incorrect. 
				<br>
				<i class="icon-warning-sign"></i> Results cannot be trusted blindly. SnpEff results may have to analyzed and independently validated.
				</p>

				<p> <h3>Databases</h3>
				In order to produce the annotations, SnpEff requires a database.
				We build these databases using informations from trusted resources.
				<br>
				Currently, there are pre-built database for over 8,500 reference genomes.
				This means that 99% of the cases are covered.
				<br>
				In some very rare occasions, people need to build a database for an organism not currently supported (e.g. the genome is not publicly available).
				In most cases, this can be done and there is a section of this manual teaching how to build your own SnpEff database.
				<br>
				<br>
				Which databases are supported? You can find out all the supported databases by running the <code>database</code> command:
				<br>
				<br>
<pre>
java -jar snpEff.jar databases
</pre>
				<br>
				This command shows the database name, genome name and source data (where was the genome reference data obtained from).
				Keep in mind that many times I use ENSEMBL reference genomes, so the name would be <code>GRCh37</code> instead of <code>hg19</code>, or <code>GRCm38</code> instead of <code>mm10</code>, and so on. 
				<br>
				<br>
				<b>Example: Finding a database:</b> So, let's say you want to find out the name of the latest mouse (Mus.Musculus) database. 
				You can runs something like this: 
				<br>
				<br>
<pre>
$ java -jar snpEff.jar databases | grep -i musculus
GRCm38.68                                                   	Mus_musculus                                                	OK        	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
GRCm38.69                                                   	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
GRCm38.70                                                   	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
GRCm38.71                                                   	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
GRCm38.72                                                   	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
GRCm38.73                                                   	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
GRCm38.74                                                   	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
NCBIM37.64                                                  	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
NCBIM37.65                                                  	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
NCBIM37.66                                                  	Mus_musculus                                                	          	http://sourceforge.net/projects/snpeff/files/databases/v3.4/snpEff_v3.4_Mus_musculus.zip
</pre>
				<br>
				At the time of writing this, you have 10 options (obviously this will change in the future). 
				Some are databases are GRCm version 37 (i.e. mm9) and some are version 38 (i.e. mm10).
				Since it is generally better to use the latest release, you should probably pick <code>GRCm38.74</code>.
				Again, this is an example of the version numbers at the time of writing this paragraph, in the future there will be other releases and you should update to the corresponding version.
				<br>
				<br>
				<b>Unsupported reference genomes:</b> If your reference genome of interest is not supported yet (i.e. there is no database avaialble), you can build a database yourself (see "Building dataases" section). If you have problems adding you own organism, send me an email and I'll do my best to help you out.
				<p>
				<p> <h3>Citing</h3>
				If you are using SnpEff or SnpSift, please cite our work as shown <a href="SnpEff.html#citing">here</a>. Thank you!
				</p>

				<p> <h3>Features</h3>
				The following table shows the main SnpEff features:
				<table class="table table-striped">
					<tr> <th> Feature </th> <th> Comment</th> </tr>
					<tr> 
						<td> Local install </td> 
						<td> 
							SnpEff can be installed in your local computer or servers.
							Local installations are preferred for processing genomic data. 
							As opposed to remote web-based services, running a program locally has many advantages:
							<ul>
								<li> There no need to upload huge genomic dataset. 
								<li> Processing doesn't depend on availability or processing capacity of remote servers.
								<li> Service continuity: no need to worry if a remote service will be maintained in the future.
								<li> Security and confidentiality issues of uploading data to third party servers are not a problem.
								<li> Avoid legal problems of processing clinical data on "outside" servers.
							</ul>
						</td> 
					</tr>
					<tr> <td> Multi platform </td> <td> SnpEff is written in Java. It runs on Unix / Linux, OS.X and Windows. </td> </tr>
					<tr> <td> Simple installation </td> <td> Installation is as simple as downloading a ZIP file and double clicking on it. </td> </tr>
					<tr> <td> Genomes </td> <td> Human genome, as well as all model organisms are supported. Over 2,500 genomes are supported, which includes most mammalian, plant, bacterial and fungal genomes with published genomic data. </td> </tr>
					<tr> <td> Speed </td> <td> SnpEff is really fast. It can annotate up to 1,000,000 variants per minute. </td> </tr>
					<tr> <td> GATK integration </td> <td> SnpEff can be easily integrated with <a href="http://www.broadinstitute.org/gatk/">GATK</a> pipelines.</td> </tr>
					<tr> <td> GUI </td> <td> Web based user interface via Galaxy project </td> </tr>
					<tr> <td> Input and Output formats </td> 
						<td> 
						SnpEff accepts input files in the following format:
							<ul>
								<li> <code>VCF</code> format, which is the de-facto standard for sequencing variants.
								<li> <code>TXT</code> format: Although TXT format is deprecated, some people still find it useful.
								<li> <code>BED</code> format: To annotate enrichment experiments (e.g. ChIP-Seq peaks) or other genomic data.
							</ul>
						</td> 
					</tr>
					<tr> <td> Variants supported </td> <td> SnpEff can annotate SNPs, MNPs, insertions and deletions. Support for mixed variants and structural variants is available (although sometimes limited). </td> </tr>
					<tr> <td> Effect supported </td> <td> Many effects are calculated: such as SYNONYMOUS_CODING, NON_SYNONYMOUS_CODING, FRAME_SHIFT, STOP_GAINED just to name a few.</td> </tr>
					<tr> <td> Variant impact </td> <td> SnpEff provides a simple assessment of the putative impact of the variant (e.g. HIGH, MODERATE or LOW impact).</td> </tr>
					<tr> <td> Cancer tissue analysis </td> <td> Somatic vs Germline mutations can be calculated on the fly. This is very useful for the cancer researcher community. </td> </tr>
					<tr> <td> Loss of Function (LOF) assessment </td> <td> SnpEff can estimate if a variant is deemed to have a loss of function on the protein. </td> </tr>
					<tr> <td> Nonsense mediate decay (NMD) assessment </td> <td> Some mutations may cause mRNA to be degraded thus not translated into a protein. NMD analysis marks mutations that are estimated to trigger nonsense mediated decay.</td> </tr>
					<tr> <td> HGVS notation</td> <td> SnpEff can provide output in HGVS notation, which is quite popular in clinical and translation research environments.</td> </tr>
					<tr> <td> User annotations</td> <td> A user can provide custom annotations (by means of BED files).</td> </tr>
					<tr> <td> Public databases </td> 
						<td> 
						SnpEff can annotate using publicly available data from well known databases, for instance:
						<ul>
							<li> <b>ENCODE</b> datasets are supported by SnpEff (by means of BigWig files provided by ENCODE project).
							<li> <b>Epigenome Roadmap</b> provides data-sets that can be used with SnpEff.
							<li> <b>TFBS</b> Transcription factor binding site predictions can be annotated. Motif data used in this annotations is generates by <code>Jasper</code> and <code>ENSEBML</code> projects
							<li> <b>NextProt</b> database can be used to annotate protein domains as well as important functional sites in a protein (e.g. phosphorilation site)
						</ul>
						</td> 
					</tr>
					<tr> <td> Common variants (dbSnp) </td> <td> Annotating "common" variants from <b>dbSnp</b> and <b>1,000 Genomes</b> can be easily done (see <code>SnpSift annotate</code>). </td> </tr>
					<tr> <td> Gwas catalog </td> <td> Support for GWAS catalog annotations (see <code>SnpSift gwasCat</code> )  </td> </tr>
					<tr> <td> Conservation scores </td> <td> PhastCons conservation score annotations support (see <code>SnpSift phastCons</code>)  </td> </tr>
					<tr>
						<td> DbNsfp </td> 
						<td> 
						A comprehensive database providing many annotations and scores, such as: <code>SIFT</code> ,<code>Polyphen2</code> ,<code>GERP++</code>, <code>PhyloP</code> ,<code>MutationTaster</code> ,<code>SiPhy</code> ,<code>Interpro</code> ,<code>Haploinsufficiency</code> ,etc. (via SnpSift). 
						<p>
						See <code>SnpSift dbnsfp</code> for details.
						</td> 
					</tr>
					<tr> <td> Non-coding annotations </td> <td> Regulatory and non-coding annotations are supported for different tissues and cell lines. Annotations supported include PolII ,H3K27ac ,H3K4me2 ,H3K4me3 ,H3K27me3 ,CTCF ,H3K36me3, just to name a few. </td> </tr>
					<tr> <td> Gene Sets annotations</td> <td> Gene sets (MSigDb, GO, BioCarta, KEGG, Reactome, etc.) can be used to annotate via <code>SnpSift geneSets</code> command.</td> </tr>
				</table>
				</p>
        </section>

		<!-- Examples
		================================================== -->
		<section id="examples">
		  <div class="page-header">
			<h1>2. Usage examples (for the impatient)</h1>
		  </div>
				<p class="lead">
				Here we'll show a few examples that should be enough to get you started.
				</p>

				<h3> Basic examples: Installing SnpEff </h3>
				<p>
				Obviously the first step to use the program is to install it.
				You have to download the core program from the "Download" page.
				Then you have to uncompress the ZIP file.
				In Windows systems, you can just double click and copy the contents of the ZIP file to wherever you want the program installed.
				If you have a Unix or a Mac system, the command line would be:
<pre>
unzip snpEff_latest_core.zip 
</pre>

				<p>
				<i class="icon-hand-right"></i> By default SnpEff stores all the databases in <code>$HOME/snpeff/data</code> (where <code>$HOME</code> is your home directory).
				If you want to change this, you can edit the config file and change the <code>data_dir</code> entry:
				<p>
<pre>
#---
# Databases are stored here
# E.g.: Information for 'hg19' is stored in data_dir/hg19/
#
# Note: Since version 2.1 you can use tilde ('~') as first character to refer to your home directory
#---
data_dir = ~/snpEff/data/
</pre>
				</p>

				<h3> Basic examples: Managing SnpEff databases </h3>
				<p>
				In order to use SnpEff, you need a database for the reference genome you want to use. 
				SnpEff databases for the most popular genomes are already pre-built and available for you to download.
				So there is no need for you to build it (this will save you a LOT of work).
				<p>
				<i class="icon-warning-sign"></i> Most of the time you don't need to build the SnpEff database. You can just download a pre-built one.
				<p>

				The easiest way to download and install a pre-built SnpEff database it using the "download" command.
				E.g. if you want to install the SnpEff database for the human genome, you can run the following command:
<pre>
java -jar snpEff.jar download -v GRCh37.66
</pre>
				<p>
				<i class="icon-hand-right"></i> If you are running SnpEff from a directory different than the one it was installed, you will have to specify where the config file is.
				This is done using the '-c' command line option:
<p><pre>
java -Xmx4g -jar snpEff.jar download <b>-c path/to/snpEff/snpEff.config</b> -v GRCh37.66 
</pre><p>

				<h3> Basic examples: Annotate using SnpEff </h3>
				Let's assume you have a VCF file ("file.vcf") and you want to annotate the variants in that file.
				You can download the file for this example <a class="body" href="examples/file.vcf">here</a>.
				This data is from the 1000 Genomes project, so the reference genome is the human genome (GRCh37.66).
				<p>
				Let's assume this is the first time you use SnpEff, so you don't have the database installed, so the first step is to download the database.
				Then you can annotate the file:

<p><pre>
# Go to the directory where snpEff is installed
cd snpEff

# Do this only if you don't already have the database installed.
java -jar snpEff.jar download -v GRCh37.66

# Annotate the file
java -Xmx4g -jar snpEff.jar eff -v GRCh37.66 file.vcf > file.eff.vcf
</pre><p>

				The annotated variants will be in the new file "file.eff.vcf".<p>
				<p>
				<i class="icon-warning-sign"></i> SnpEff creates a file called "snpEff_summary.html" showing basic statistics about the analyzed variants. 
				Take a quick look at it!

				<p>
				<i class="icon-hand-right"></i> We used the java parameter -Xmx4g to increase the memory available to the Java Virtual Machine to 4G. 
				SnpEff's human genome database is large and it has to be loaded into memory. 
				If your computer doesn't have at least 4G of memory, you probably won't be able to run this example.

				<p>
				<i class="icon-hand-right"></i> If you are running SnpEff from a directory different than the one it was installed, you will have to specify where the config file is.
				This is done using the '-c' command line option:
<p><pre>
java -Xmx4g -jar snpEff.jar eff <b>-c path/to/snpEff/snpEff.config</b> -v GRCh37.66 file.vcf > file.eff.vcf
</pre><p>


				<h3> Advanced examples: Filter out SNPs from dbSnp </h3>

				These are slightly more advanced examples. 
				Here we'll try to show how to perform specific tasks.

				If you want to filter out SNPs from dbSnp, you can do it using SnpSift.
				You can download SnpSift from the "Downloads" page.
				<p>
				You can download the file for this example <a class="body" href="examples/file.vcf">here</a>.<br>
				Here is how to do it:
				<p>
				<ol>
					<li> Annotate ID fields using <code>SnpSif annotate</code> and DbSnp<br>
						<i class="icon-hand-right"></i> We annotate using dbSnp before using SnpEff in order to 
								have 'known' and 'unknown' statistics in SnpEff's summary page.
								Those stats are based on the presence of an ID field. If the ID 
								is non-empty, then it is assumed to be a 'known variant'.
<pre>
# Download and uncompress dbSnp database.
wget -O dbSnp.vcf.gz ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/00-All.vcf.gz
gunzip dbSnp.vcf.gz

# Annotate ID field using dbSnp
java -jar SnpSif.jar annotate -v dbSnp.vcf.gz file.vcf > file.dbSnp.vcf
</pre>

					<li> Annotate using SnpEff:
<pre>
# Do this only if you don't already have the database installed.
java -jar snpEff.jar download -v GRCh37.66

# Annotate the file
java -Xmx4g -jar snpEff.jar eff -v GRCh37.66 file.dbSnp.vcf > file.eff.vcf
</pre>

					<li> Filter out variants that have a non-empty ID field. 
					These variants are the ones that are NOT in dbSnp, since we annotated the ID field using rs-numbers from dbSnp in step 1.
<pre>
java -jar SnpSift.jar filter -f file.eff.vcf "! exists ID" > file.eff.not_in_dbSnp.vcf
</pre>
					<i class="icon-hand-right"></i>The expression using to filter the file is "! exists ID".
					This means that the ID field does not exists (i.e. the value is empty) which is represented as a dot (".") in a VCF file.
				</ol>
				</p>
        <!-- Download
        ================================================== -->
        <section id="install">
          <div class="page-header">
            <h1>3. Download &amp; Install</h1>
          </div>
				<p class="lead"> Download and installing SnpEff it pretty easy. 
				</p>
				
				<p> <h3>System requirements </h3>
				SnpEff requires that you have Java v1.6 or later installed (any modern operating system has it). <br>
				You'll probably need at least 2Gb of memory. 
				The amount of memory used can vary significantly depending on genome size and data analysis type you are doing.

				<p> <h3>Downloading and Installing</h3>
				In order to download and install the program, you can follow the directions <a href="download.html">here</a>.

				<p> <h3>Configuration</h3>
				The only configuration file is <code>snpEff.config</code>.
				Most configuration parameters, are explained in the comments in the same config file, so I won't repeat the explanation here :-)
				<br>
				Usually you will only need to change the <code>data_dir</code> parameter.
				This parameter points to the <code>data</code> directory where you installed the tool (i.e. in unix systems, this is <code>$HOME/snpEff/data</code>).<p>

				<p> <h3>Source code </h3>
				The source code is in GitHub (although we keep the binary distribution is at SourceForge). 
				<br>
				<br>
				<b> Getting the souce:</b> Here is the <code>git</code> command to check out the development version of the code:
				<br>
				<br>
<pre>
git clone https://github.com/pcingola/SnpEff.git

# If you want to SnpSift as well
git clone https://github.com/pcingola/SnpSift.git
</pre>
				<br>
				<b>Install required libraries:</b> 
				Most libraries should be install using Maven, so you just need to run <code>mvn</code> command.
				<br>
				<br>
				You may need to install <code>sam-jdk</code> JAR file manually on your maven repository.
				Download the JAR file <a href="http://sourceforge.net/projects/picard/files/">here</a>, then run the following command to add the JAR to Maven:
				<br>
				<br>
<pre>
# Download JAR file
wget http://sourceforge.net/projects/picard/files/sam-jdk/1.97/sam-1.97.jar

# Manually add it to Maven
mvn install:install-file -Dfile=sam-1.97.jar -DgroupId=net.sf.samtools -DartifactId=Sam -Dversion=1.97 -Dpackaging=jar
</pre>
				<br>
				<b>Intalling test cases:</b>
				Test cases require special "test cases databases and genome", you can find them here:
				<br>
				<br>
<pre>
# Install test databases in SnpEff's development directory (not the soruce code dir!)
cd $HOME/snpEff

# Download dtabases and genome for test cases
wget http://sourceforge.net/projects/snpeff/files/databases/test_cases.tgz

# Uncompress
tar -xvzf test_cases.tgz

# Go to Eclipse's workspace directory (where the source code is)
cd $HOME/workspace/SnpEff

# Create a link to the 'data' dir, so that we can run test cases within Eclipse
ln -s $HOME/snpEff/data

# Add data dir to 'gitignore'
echo "/data" >> .gitignore
</pre>
        </section>

        <!-- Running SnpEff
        ================================================== -->
        <section id="run">
          <div class="page-header">
            <h1>4. Running SnpEff</h1>
          </div>
				<p class="lead"> We show some basic examples how to use SnpEff.</p>

				<p><h3>Command line vs Web interface</h3>
				In order to run SnpEff you need to be comfortable running command from a command line terminal.
				If you are not, then it is probably a good idea to ask you systems administrator to install a <a class="body"class="body" href="http://usegalaxy.org">Galaxy</a> server and use the web interface.
				You can also use the open Galaxy server, but functionality may be limited and SnpEff versions may not be updated frequently.
				</p>

				<p><h3>Usage example</h3>
				In this example, we will annotate a variants file using the human reference genome.<br>

				As an input, we will use a <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41"> Variant Call Format (VCF) </a> file.
				VCF is the de-facto standard for genetic variants.
				<br>
				In this example we will use a file 'demo.1kg.vcf.gz', which is provided in the distribution.
<pre>
# Go to SnpEff install directory
cd ~/snpEff 

# Download human database 'hg19'
java -jar snpEff.jar download -v hg19

# Annotate 
java -Xmx2g -jar snpEff.jar hg19 -v demo.1kg.vcf > demo.1kg.snpeff.vcf
</pre>

				Here is an example of some entries in the annotated output file. 
				You can see the 'EFF' field was added, predicting STOP_GAINED protein changes:

<pre>
$ cat demo.1kg.snpeff.vcf | grep STOP_
1    889455    .    G    A    100.0    PASS    ...;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q236*...
1    897062    .    C    T    100.0    PASS    ...;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*...
1    900375    .    G    A    100.0    PASS    ...;EFF=STOP_GAINED(HIGH|NONSENSE|tGg/tAg|W578*...
</pre>
				Note: The real output was edited for readibility reasons.
				</p>

				<p><h3>Running from another directory</h3>
				When you run SnpEff from a different directory than your install directory, you have to specify where the config file is located using the '-c' command line option.
<pre>
java -Xmx2g path/to/snpEff/snpEff.jar -c path/to/snpEff/snpEff.config GRCh37.70 path/to/snps.vcf
</pre>

				<p><h3>Java memory options</h3>
				By default the amount of memory set by a java process is set too low.
				If you don't assign more memory to the process, you will most likely have an "OutOfMemory" error.<br>

				You should set the amount of memory in your java virtual machine to, at least, 2 Gb. 
				This can be easily done using the Java command line option "-XmX".
				E.g. In this example I use 4Gb:
<pre>
# Run using 4 Gb of memory 
java -Xmx4G snpEff.jar hg19 path/to/your/files/snps.vcf
</pre>
				Note: There is no space between "-Xmx" and "4G".
				</p>

				<p><h3>Running SnpEff in the Cloud </h3>
				You can run SnpEff in a "the Cloud" exactly the same way as running it on your local computer.
				You should not have any problems at all.
				<br>
				Here is an example of installing it and running it on an Amazon EC2 instance (virtual machine):<p>
<pre>
$ ssh -i ./aws_amazon/pcingola_aws.pem ec2-user@ec2-54-234-14-244.compute-1.amazonaws.com

       __|  __|_  )
       _|  (     /   Amazon Linux AMI
      ___|\___|___|


[ec2-user@ip-10-2-202-163 ~]$ wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip
[ec2-user@ip-10-2-202-163 ~]$ unzip snpEff_latest_core.zip
[ec2-user@ip-10-2-202-163 ~]$ cd snpEff_3_1/
[ec2-user@ip-10-2-202-163 snpEff_3_1]$ java -jar snpEff.jar download -v hg19
00:00:00.000    Downloading database for 'hg19'
...
00:00:36.340    Done
[ec2-user@ip-10-2-202-163 snpEff_3_1]$ java -Xmx4G -jar snpEff.jar dump -v hg19 > /dev/null
00:00:00.000    Reading database for genome 'hg19' (this might take a while)
00:00:20.688    done
00:00:20.688    Building interval forest
00:00:33.110    Done.
</pre>

				<p>
				As you can see, it's very simple.

				<p><h3>Command line options</h3>
				In order to see all available options, you can run 
				<pre>java -jar snpEff.jar -help</pre>
				We'll explain many of those options in detail in the following chapters.
				</p>

				<p><h3>Loading the database</h3>
				One of the first thins SnpEff has to do is to load the database.
				Usually it takes frmo a few seconds to a couple of minutes, depending on database size.
				Complex databases, like human, require more time to load.
				After the database is loaded, SnpEff can analyze thousands of variants per second.
				</p>
        </section>

        <!-- Sequence data analysis
        ================================================== -->
        <section id="seqData">
          <div class="page-header">
            <h1>5. Sequence data analysis</h1>
          </div>
				<p class="lead"> Here we show an example on how to get from Sequencing data to an annotated variants file </p>

				<p><h3>Sequencing data example</h3>
				<i class="icon-warning-sign"></i> This is an extremely simplified version on how to analyze the data from scratch
				This is not meant to be a tutorial on sequencing analysis as it would be way beyond the scope of this handbook.
				<br>
				Let's assume you have sequence data in FASTQ format (file "s.fastq") and your reference genome is dm5.34 (fly genome)
<pre>
# Download the genome, uncompress and rename file
wget ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r5.34_FB2011_02/fasta/dmel-all-chromosome-r5.34.fasta.gz
gunzip dmel-all-chromosome-r5.34.fasta.gz
mv dmel-all-chromosome-r5.34.fasta dm5.34.fasta

# Create a genome index (we assume you installed BWA http://bio-bwa.sourceforge.net/)
bwa index -bwtsw dm5.34.fasta

# Map sequences to the genome: Create SAI file
bwa aln -bwtsw dm5.34.fasta s.fastq > s.sai

# Map sequences to the genome: Create SAM file
bwa samse dm5.34.fasta s.sai s.fastq > s.sam

# Create BAM file (we assume you installed SamTools http://samtools.sourceforge.net/)
samtools view -S -b s.sam > s.bam

# Sort BAM file (will create s_sort.bam)
samtools sort s.bam s_sort

# Create VCF file (BcfTools is part of samtools distribution)
samtools mpileup -uf dm5.34.fasta s_sort.bam | bcftools view -vcg - > s.vcf

# Analyze variants using snpEff
java -Xmx4g -jar snpEff.jar dm5.34 s.vcf > s.eff.vcf
</pre>

				This highly simplified sequencing data analysis pipeline, has the common basic steps
				<ul>
					<li> Index the reference genome (bwa)
					<li> Map reads to reference genome (bwa)
					<li> Call variants (bcftools)
					<li> Annotate variants (SnpEff)
				</ul>
				</p>
        </section>
				
        <!-- Input
        ================================================== -->
        <section id="input">
          <div class="page-header">
            <h1>6. Input files </h1>
          </div>
				<p class="lead"> 
				Files used as input to SnpEff must comply with standard formats.
				Here we describe supported input data formats.
				</p>

				<i class="icon-hand-right"></i> You can use "-" as input file name to spedicfy <b>STDIN</b>. 
				<br>
				For example, you can easily stream files like this:
				<br>
				<br>
<pre>
cat file.vcf | java -Xmx4g -jar snpEff.jar hg19 - > file.eff.vcf
</pre>


				<p><h3>VCF</h3>
				As we mentioned before, <a class="body"class="body" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">Variant Call Format (VCF)</a> is the recomended format for input files.
				This is the format used by the "1000 Genomes Project", and is currently considered the de-facto standard for genomic variants.
				Furthermore, most variant calling programs output VCF files, so it is also the simplest format to use.
				<br>
				In a nutshell, VCF format is tab-separated text file having the following columns:

				<ol>
					<li> Chromosome name
					<li> Position
					<li> Variant's ID
					<li> Reference genome
					<li> Alternative (i.e. variant)
					<li> Quality score
					<li> Filter (whether or not the variant passed quality filters)
					<li> INFO : Generic information about this variant. SnpEff adds annotation information in this column.
				</ol>

				Here is an example of a few lines in a VCF file:
<pre>
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                    
20     14370   rs6054257 G      A       29   PASS   NS=3;DP=14;AF=0.5;DB;H2
20     17330   .         T      A       3    q10    NS=3;DP=11;AF=0.017   
</pre>
				Note that the first line is header information. Header lines start with '#'
				<br>
				<br>
				</p>

				<p><h3>TXT</h3>
				<i class="icon-warning-sign"></i> TXT format is currently deprecated and discouraged.
				<br>
				TXT file format must be tab-separated, containing columns that correspond to:
				<ul>
					<li>chromosome_name
					<li>chromosome_position
					<li>reference sequence
					<li>changed sequence: A slash is used to separate two alleles
					<li>strand: {+,-} (optional)
					<li>quality (optional)
					<li>coverage (optional)
				</ul>

				This is an example of an input TXT file:

<pre>
5   140532    T    C       +
12  1017956   T    A       +    45    12
2   946507    G    C       +    23    8
14  19584687  C    T       -
19  66520     G    -G/-G   +
8   150029    A    */+T    +
</pre>

				<p><b>TXT: Representing an heterozygous variant</b><br> 
				<i class="icon-warning-sign"></i> TXT format is currently deprecated and discouraged.
				<br>
				When representing a heterozygous SNP you can use IUB codes: M=A/C, R=A/G, W=A/T, S=C/G, Y=C/T and K=G/T. 
				Indels are indicated by, for example, */+A, -A/* or +CC/-C. There is no difference between */+A or +A/*.
				<br>
				<br>
				</p>

				<p><h3>BED</h3>
				In an enrichment experiment, such as ChIP-Seq, the results are enrichment regions, usually called "peaks".
				It is common for "peak callers" (algorithms that detect enrichment), write the results in a BED file.
				SnpEff can annotate BED files in order to facilitate interpretation of enrichment experiments.
				<br>
				<br>
				<i class="icon-warning-sign"></i> Column fifth onwards are ignored when using BED file format and they will be lost in the output file.
				</p>
        </section>

        <!-- Output
        ================================================== -->
        <section id="output">
          <div class="page-header">
            <h1>7. Output files </h1>
          </div>
				<p class="lead"> 
				Here we describe the format of SnpEff's output files and the information added.
				</p>

				<p><h3>VCF output format</h3>
				As we mentioned in the previous chapter, VCF is the default input format.
				It is highly recommended to use VCF as input and output format, since it is a standard format that can be also used by other tools and software packages.
				Thus VCF makes it much easier to integrate genomic data processing pipelines.
				<br>
				SnpEff adds annotation information to the INFO field of a VCF file.
				The INFO field is the eight column of a VCF file, see previous section for a quick example or take a look at the <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF specification</a> for details.
				<br>
				<br>
				Here is an example of a file before and after being annotated using SnpEff:
				<br>
				VCF file before annotations
<pre>
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                    
1	889455	.	G	A	100.0	PASS	AF=0.0005
1	897062	.	C	T	100.0	PASS	AF=0.0005
</pre>

				VCF file after being annotated using SnpEff
<pre>
#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                    
1	889455	.	G	A	100.0	PASS	AF=0.0005;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q236*|749|NOC2L||CODING|NM_015658|)
1	897062	.	C	T	100.0	PASS	AF=0.0005;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*|642|KLHL17||CODING|NM_198317|)
</pre>
				A you can see, SnpEff added an 'EFF' tag to the INFO field (eigth column).
				</p>

				<p><b>VCF EFF field</b>
				<br>
				<br>
				Effects information is added to the INFO field using an 'EFF' tag. 

				There can be multiple effects separated by comma. The format for each effect is:
<pre>
EFF= Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_Change| Amino_Acid_Length | Gene_Name | Transcript_BioType | Gene_Coding | Transcript_ID | Exon_Rank  | Genotype_Number [ | ERRORS | WARNINGS ] )
</pre>

				
				<center>
				<table class="table table-striped">
					<tr> <th>  EFF Sub-field </th> <th>  Meaning </th> </tr>
					<tr> <td>  Effect </td> <td>  Effect of this variant. See details <a class="body" href="#What_effects_are_predicted?">here</a>. </td> </tr>
					<tr> <td>  Effect impact </td> <td>  Effect impact {High, Moderate, Low, Modifier}. See details <a class="body" href="#How_is_impact_categorized?_(VCF_output)">here</a>. </td> </tr>
					<tr> <td>  Functional Class </td> <td>  Functional class {NONE, SILENT, MISSENSE, NONSENSE}. </td> </tr>
					<tr> <td>  Codon_Change / Distance </td> <td>  Codon change: old_codon/new_codon OR distance to transcript (in case of upstream / downstream) </td> </tr>
					<tr> <td>  Amino_Acid_Change </td> <td>  Amino acid change: old_AA AA_position/new_AA (e.g. 'E30K') </td> </tr>
					<tr> <td>  Amino_Acid_Length</td> <td>  Length of protein in amino acids (actually, transcription length divided by 3). </td> </tr>
					<tr> <td>  Gene_Name </td> <td>  Gene name </td> </tr>
					<tr> <td>  Transcript_BioType </td> <td>  Transcript bioType, if available. </td> </tr>
					<tr> <td>  Gene_Coding </td> <td>  [CODING | NON_CODING]. This field is 'CODING' if any transcript of the gene is marked as protein coding. </td> </tr>
					<tr> <td>  Transcript_ID </td> <td>  Transcript ID (usually ENSEMBL IDs) </td> </tr>
					<tr> <td>  Exon/Intron Rank </td> <td>  Exon rank or Intron rank (e.g. '1' for the first exon, '2' for the second exon, etc.) </td> </tr>
					<tr> <td>  Genotype_Number </td> <td>  Genotype number corresponding to this effect (e.g. '2' if the effect corresponds to the second ALT) </td> </tr>
					<tr> <td>  Warnings / Errors </td> <td>  Any warnings or errors (not shown if empty). </td> </tr>
				</table>
				</center>
				</p>

				<p><b>Errors and Warnings </b>
				<br>
				<br>
				As mentioned int the previous section, the last sub-field in EFF field shows errors or warnings (if any).
				Here is a description of the errors and warnings:
				<center>
				<table class="table table-striped">
					<tr> <th>  Error </th> <th>  Meaning and possible solutions</th> </tr>
					<tr> <td> ERROR_CHROMOSOME_NOT_FOUND </td> 
						<td>
							Chromosome does not exits in reference database. 
							Typically this means that there is a mismatch between the chromosome names in your input file and the chromosome names used in the reference genome to build SnpEff's database. 
							<br>
							<br>
							<b>Warning:</b> This could be caused because you are trying to annotate using a reference genome that is different than the one you used for sequence alignment.
							Obviously doing this makes no sense and the annotation information you'll get will be garbage.
							That's why SnpEff shows you an error message.
							<br>
							<br>
							<b>Possible solution:</b> Sometime SnpEff database matches the reference genome for your organism, and it's just that the chromosome names are changed.
							You can fix this by changing the input file.
							<br>
							You can see chromosome names used by SnpEff by using <code>-v</code> (verbose) option. 
							SnpEff will show a line like this one:
							<pre># Chromosomes names [sizes]  : '1' [249250621] '2' [243199373] ...</pre>
							Then you can modify your input file chromosome names to match the database (e.g. using <code>sed</code> unix command).
						</td> 
					</tr>
					<tr> <td> ERROR_OUT_OF_CHROMOSOME_RANGE </td> 
						<td> 
							This means that the position is higher than chromosome's length. 
							Probably an indicator that your data is not from this reference genome. 
						</td> 
					</tr>
					<tr> <td> ERROR_OUT_OF_EXON </td> <td> Exonic information not matching the coordinates. Indicates a problem (or even a bug?) in the database </td> </tr>
					<tr> <td> ERROR_MISSING_CDS_SEQUENCE </td> <td> Transcript has no CDS info. Indicates a problem (or even a bug?) in the database </td> </tr>
				</table>

				<table class="table table-striped">
					<tr> <th>  Warning </th> <th>  Meaning and possible solutions</th> </tr>
					<tr> <td> WARNING_REF_DOES_NOT_MATCH_GENOME </td> 
						<td> 
						This means that the REF field does not match the reference genome.
						<br> 
						<br> 
						<i class="icon-warning-sign"></i> This warning probably indicated there is something really wrong with your data!
						<br> 
						<br> 
						This happens when your data was aligned to a different reference genome than the one used to create SnpEff's database.
						If there are many of these warnings, it's a strong indicator that the data doesn't match and all the annotations will be garbage (because you are using the wrong database).
						<br> 
						<br> 
						<b>Solution:</b> Use the right database to annotate!
						<br> 
						<br> 
						<i class="icon-hand-right"></i> Due to performance and memory optimizations, SnpEff only checks reference sequence on Exons.
						</td> 
					</tr>
					<tr> <td> WARNING_SEQUENCE_NOT_AVAILABLE </td> <td> For some reason the exon sequence is not available, so we cannot calculate effects.</td> </tr>
					<tr> <td> WARNING_TRANSCRIPT_INCOMPLETE</td> 
						<td> 
						A protein coding transcript whose length is non-multiple of 3. 
						This means that information is missing for one or more amino acids.
						<br>
						This is usually due to errors in the genomic information (e.g. the genomic databases provided by UCSC or ENSEMBL).
						Genomic information databases are constantly being improved and are getting more accurate, but some errors still remain.
						</td> 
					</tr>
					<tr> <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS </td> 
						<td> 
						A protein coding transcript has two or more STOP codons in the middle of the coding sequence (CDS).
						This should not happen and it usually means the genomic information may have an error in this transcript.
						<br>
						This is usually due to errors in the genomic information (e.g. the genomic databases provided by UCSC or ENSEMBL).
						Genomic information databases are constantly being improved and are getting more accurate, but some errors still remain.
						</td> 
					</tr>
					<tr> <td> WARNING_TRANSCRIPT_NO_START_CODON </td>
						<td> 
						A protein coding transcript does not have a proper START codon.
						It is rare that a real transcript does not have a START codon, so this probably indicates errors in genomic information for this transcript (e.g. the genomic databases provided by UCSC or ENSEMBL).
						<br>
						Genomic information databases are constantly being improved and are getting more accurate, but some errors still remain.
						</td> 
					</tr>
				</table>
				</center>
				</p>
				
				<p><b>VCF Header lines </b>
				<br>
				<br>
				SnpEff updates the header of the VCF file to reflect additional fields.
				This is required by the VCF specification.
				SnpEff also adds the command line options used to annotate the file as well as SnpEff's version, so you can keep track of what exactly was done.
				<br>
				<br>
				Here is an example of some header lines added to an annotated file:
<pre>
##SnpEffVersion="SnpEff 3.1m (build 2013-02-08)"
##SnpEffCmd="SnpEff  hg19 demo.1kg.vcf "
##INFO=&lt;ID=EFF,Number=.,Type=String,Description="Predicted effects for this variant.Format: 'Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_change| Amino_Acid_length | Gene_Name | Gene_BioType | Coding | Transcript | Exon [ | ERRORS | WARNINGS ] )' \"&gt;
</pre>
				</p>

				<br> 
				<p><b>Multiple effects per VCF line</b>
				<br> 
				<br> 
				Usually there is more than one effect reported in each EFF field.
				<br>
				There are several reasons for this:
				<ul>
					<li> A variant can affect multiple genes. E.g a variant can be DOWNSTREAM from one gene and UPSTREAM from another gene.
					<li> In complex organisms, genes usually have multiple transcripts. So SnpEff reports the effect of a variant on each transcript.
					<li> A VCF line can have more then one variant. 
						 For instance if reference genome is 'G', but the sample has either 'A' or 'T', then this will be reported as one line, having multiple alternative variants. 
						 E.g.: 
<pre>
#CHROM  POS       ID   REF  ALT      QUAL  FILTER  INFO                    
1       889455    .    G    A,T      .     .       AF=0.0005
</pre>
						In this case SnpEff will report the effect of each variant on each gene and each transcript.
				</ul>
				</p>

				<br>
				<p> <b> Counting total number of effects of a given type </b>
				<br>
				<br>
				Some people try to count the number of effects in a file by doing (assuming we want to count how many MODIFIER effects we have): 
				<br>
				<pre> grep -o MODIFIER output.eff.vcf | wc -l </pre>
				<br>
				This is incorrect because a VCF line can have multiple effects (e.g. when there are multiple transcripts in a gene).
				A proper way to count effects would be:

<pre>
cat output.eff.vcf \
	| cut -f 8 \
	| tr ";" "\n" \
	| grep ^EFF= \
	| cut -f 2 -d = \
	| tr "," "\n" \
	| grep MODIFIER \
	| wc -l
</pre>

				Brief explanation:
				<table class="table table-striped">
					<tr> <td> <tt>cut -f 8 </tt> </td> <td>  Extract INFO fields </td> </tr>
					<tr> <td> <tt>tr ";" "\n" </tt> </td> <td>  Expand each field into one line </td> </tr>
					<tr> <td> <tt>grep ^EFF= </tt> </td> <td>  Only keep 'EFF' fields </td> </tr>
					<tr> <td> <tt>cut -f 2 -d = </tt> </td> <td>  Keep only the effect data (drop the 'EFF=' part) </td> </tr>
					<tr> <td> <tt>tr "," "\n" </tt> </td> <td>  Expand effects to multiple lines </td> </tr>
					<tr> <td> <tt>grep MODIFIER | wc -l </tt> </td> <td>  Count the ones you want (in this example 'MODIFIER')</td> </tr>
				</table>
				</p>

				<br> 
				<p><h3>TXT output format</h3>
				<i class="icon-warning-sign"></i> TXT format is currently deprecated and discouraged.
				<br>
				The TXT output format consist of one line per effect. 
				This means that you usually get more than one line per SNP since the same SNP may affect several different transcripts in the same gene.
				<br>
				<br>
				The format consist of tab separated columns:
				<center>
				<table class="table table-striped">
					<tr> <th>  Column </th> <th>  Meaning </th> </tr>
					<tr> <td>  Chromosome </td> <td>  Chromosome name (usually without any leading 'chr' string) </td> </tr>
					<tr> <td>  Position </td> <td>  One based position </td> </tr>
					<tr> <td>  Reference </td> <td>  Reference </td> </tr>
					<tr> <td>  Change </td> <td>  Sequence change </td> </tr>
					<tr> <td>  Change type </td> <td>  Type of change {SNP, MNP, INS, DEL} </td> </tr>
					<tr> <td>  Homozygous </td> <td>  Is this homozygous or heterozygous {Hom, Het} </td> </tr>
					<tr> <td>  Quality </td> <td>  Quality score (from input file) </td> </tr>
					<tr> <td>  Coverage </td> <td>  Coverage (from input file) </td> </tr>
					<tr> <td>  Warnings </td> <td>  Any warnings or errors. </td> </tr>
					<tr> <td>  Gene_ID </td> <td>  Gene ID (usually ENSEMBL) </td> </tr>
					<tr> <td>  Gene_name </td> <td>  Gene name </td> </tr>
					<tr> <td>  Bio_type </td> <td>  BioType, as reported by ENSEMBL. </td> </tr>
					<tr> <td>  Transcript_ID </td> <td>  Transcript ID (usually ENSEMBL) </td> </tr>
					<tr> <td>  Exon_ID </td> <td>  Exon ID (usually ENSEMBL) </td> </tr>
					<tr> <td>  Exon_Rank </td> <td>  Exon number on a transcript </td> </tr>
					<tr> <td>  Effect </td> <td>  Effect of this variant. See details below. </td> </tr>
					<tr> <td>  old_AA/new_AA </td> <td>  Amino acid change </td> </tr>
					<tr> <td>  old_codon/new_codon </td> <td>  Codon change </td> </tr>
					<tr> <td>  Codon_Num(CDS) </td> <td>  Codon number in CDS </td> </tr>
					<tr> <td>  Codon_degenaracy </td> <td>  Codon degenaracy (see below). </td> </tr>
					<tr> <td>  CDS_size </td> <td>  CDS size in bases </td> </tr>
					<tr> <td>  Custom_interval_ID </td> <td>  If any custom interval was used, add the IDs here (may be more than one). </td> </tr>
				</table>
				</center>
				<br>
				This is an example of an output file:
				<br>
				<small>
<pre>
chr2L   35041   T       K       SNP     Het     50                      FBgn0051973     Cda5            FBtr0078164     FBgn0051973:11  11      SYNONYMOUS_CODING       Q/Q     CAA/CAA 332     5997    
chr2L   35041   T       K       SNP     Het     50                      FBgn0051973     Cda5            FBtr0078163     FBgn0051973:11  11      NON_SYNONYMOUS_CODING   Q/H     CAA/CAC 332     3120    
chr2L   200401  C       Y       SNP     Het     228                     FBgn0016977     spen            FBtr0078121     CG18497:10      10      STOP_GAINED     Y/*     TAC/TAG 5240    16602   
chr2L   194601  C       Y       SNP     Het     228                     FBgn0016977     spen            FBtr0078122     CG18497:6       6       NON_SYNONYMOUS_CODING   P/A     CCG/GCG 3473    16683   
chr2L   779563  A       R       SNP     Het     168                     FBgn0031277     CG13947         FBtr0078005     FBgn0031277:1   1       STOP_LOST       */Y     TAA/TAC 120     360   
chr2L   856021  *       +CGGAGGAGG/*    INS     Het     267                     FBgn0031288     CG13949         FBtr0078017     FBgn0031288:3   3       FRAME_SHIFT                     118     438     
chr2L   856890  C       Y       SNP     Het     201                     FBgn0029095     aru             FBtr0078031                     DOWNSTREAM: 43 bases   
chr2L   871556  G       R       SNP     Het     228                     FBgn0053526     PNUTS           FBtr0091487     CG33526:2       2       5PRIME_UTR: 18 bases from TSS  
chr2L   878267  T       W       SNP     Het     228             WARNING: Base is 'G' but SNP says 'T'.  FBgn0031292     CG15824         FBtr0078029     CG15824:14      14      3PRIME_UTR: 14 bases from transcript end  
</pre>
				</small>
				</p>

				<br>
				<i> Meaning of 'Degeneracy' column </i>: Here is an explanation form Stephen Wright (Univ. Toronto), who requested to add this feature
	  
<pre>
	     "...a fourfold degenerate site would be a site where any change is synonymous. So the 
	     third codon position for the arginine CGN, is a fourfold degenerate site, as is the 
	     third codon position for valine, alanine, etc.
	     Similarly, all second positions of a codon are zerofold degenerate, since any change is 
	     nonsynonymous. Many first codon positions are also zerofold degenerate, however, for 
	     example, the first codon position of AGG is NOT zerofold, because a shift to CGG is a 
	     synonymous change."
</pre>
				</p>

				<br>
				<p> <b> Details about Hom/Het calculation (summary file and TXT output format) </b>
				<br>
				<br>
				In older formats, there used to be one sample per file, so 
				identifying Hom/Het in a per variant basis made sense.
				<br>
				<br>
				In VCF files, where many samples are analyzed simultaneously, 
				identification of Hom/Het should be done in a per sample basis 
				instead of per variant basis. In an attempt to output that field I 
				used a majority vote (i.e. if most samples are Hom, I output 
				Hom, otherwise, Het). This is not strictly correct and, it 
				may also be confusing.
				<br>
				<br>
				For this reason, the output is empty when using multi-sample files, such as VCF. 
				Unless the file contains only one sample (i.e. only one genotype field)
				</p>

				<br>
				<p> <b> Coordinates differences between VCT and TXT outputs </b>
				<br>
				<br>
				There are some situations when there seems to be a difference between TXT output coordinates and VCF output coordinates.
				This happens because sometimes VCF files have a coordinate that does not start exactly were the variant is. 
				Here is an example:
				<pre>
VCF line:                   "chr10   102770293       .       tgctgcggctgcggctgcggctacggctgcggct      tGCTGCGgctgcggctgcggctgcggctacggctgcggct"
INS reported by SnpEff:     "chr10   102770315       *       +GCGGCT INS"
</pre>
<p>

				At first glace, these two look different, since 102770293 is not the same as 102770315. 
				But if you look at the sequences:

<pre>
tgctgcggctgcggctgcggct      acggctgcggct    REF (from VCF file)
||||||||||||||||||||||      ||||||||||||
tgctgcggctgcggctgcggctgcggctacggctgcggct    ALT (from VCF file)
|      |         |    |||||| 
|      |         |    gcggct                INS (reported by SnpEff TXT format)
|      |         |    |
102770293        |    |                     Position (VCF)
       |         |    |
       102770300 |    |
                 |    |
                 102770310
                      |
                      102770315             Position (TXT)
</pre>
				So the two VCF and TXT representations are equivalent.
				</p>
				<br>
				<br>

				<p><h3>BED output format</h3>
				<br>
				SnpEff can annotate BED files in order to facilitate interpretation of enrichment experiments.
				Annotations are added to the fourth column of the BED file.
				<br>
				E.g.: 
<pre>
$ java -Xmx4g -jar snpEff.jar eff -i bed BDGP5.69 chipSeq_peaks.bed 

# SnpEff version 3.3 (build 2013-05-15), by Pablo Cingolani
# Command line: SnpEff  -i bed BDGP5.69 /home/pcingola/fly_pvuseq/chipSeq/Sample_w1118_IP_w_5hmC/w1118_IP_w_5hmC_peaks.bed 
# Chromo  Start     End       Name;Effect|Gene|BioType        Score
2L        189463    190154    MACS_peak_1;Exon|exon_6_12_RETAINED|FBtr0078122|protein_coding|spen|protein_coding;Exon|exon_5_10_RETAINED|FBtr0078123|protein_coding|spen|protein_coding;Exon|exon_7_13_RETAINED|FBtr0306341|protein_coding|spen|protein_coding;Exon|exon_6_11_RETAINED|FBtr0078121|protein_coding|spen|protein_coding 245.41
2L        195607    196120    MACS_peak_2;Exon|exon_6_12_RETAINED|FBtr0078122|protein_coding|spen|protein_coding;Exon|exon_5_10_RETAINED|FBtr0078123|protein_coding|spen|protein_coding;Exon|exon_7_13_RETAINED|FBtr0306341|protein_coding|spen|protein_coding;Exon|exon_6_11_RETAINED|FBtr0078121|protein_coding|spen|protein_coding 51.22
2L        527253    527972    MACS_peak_3;Intron|intron_2_RETAINED-RETAINED|FBtr0078063|protein_coding|ush|protein_coding     55.97
2L        711439    711764    MACS_peak_4;Intron|intron_1_RETAINED-RETAINED|FBtr0078045|protein_coding|ds|protein_coding      61.16
2L        1365255   1365556   MACS_peak_5;Upstream|FBtr0077927|protein_coding|CG14346|protein_coding;Upstream|FBtr0077926|protein_coding|CG14346|protein_coding;Intergenic|NLaz...CG14346;Upstream|FBtr0077942|protein_coding|NLaz|protein_coding     62.78
2L        1970199   1970405   MACS_peak_6;Upstream|FBtr0077813|protein_coding|Der-1|protein_coding;Intergenic|tRNA:CR31942...Der-1;Downstream|FBtr0077812|tRNA|tRNA:CR31942|tRNA      110.34
2L        3345637   3346152   MACS_peak_7;Intron|intron_2_ALTTENATIVE_3SS-ALTTENATIVE_3SS|FBtr0089979|protein_coding|E23|protein_coding;Intron|intron_3_ALTTENATIVE_3SS-ALTTENATIVE_3SS|FBtr0089981|protein_coding|E23|protein_coding 65.49
2L        4154734   4155027   MACS_peak_8;Intergenic|CG2955...Or24a;Downstream|FBtr0077468|protein_coding|CG2955|protein_coding       76.92
2L        4643232   4643531   MACS_peak_9;Downstream|FBtr0110769|protein_coding|BG642163|protein_coding;Exon|exon_2_2_RETAINED|FBtr0300354|protein_coding|CG15635|protein_coding      76.92

</pre>

				When a peak intersects multiple transcripts or even multiple genes, each annotation is separated by a semicolon.
				So if you look into the previous results in more detail, the fisrt line looks like this (format editted for readibility pourposes):
<pre>
2L  189463  190154  MACS_peak_1;Exon|exon_6_12_RETAINED|FBtr0078122|protein_coding|spen|protein_coding
                               ;Exon|exon_5_10_RETAINED|FBtr0078123|protein_coding|spen|protein_coding
                               ;Exon|exon_7_13_RETAINED|FBtr0306341|protein_coding|spen|protein_coding
                               ;Exon|exon_6_11_RETAINED|FBtr0078121|protein_coding|spen|protein_coding
</pre>
				This peak is hitting four transcripts (FBtr0078122, FBtr0078123, FBtr0306341, FBtr0078121) in gene 'spen'.
				</p>
				<br>

				<p><h3>Exon naming convention</h3>
				The format for the exon identifier is <code>exon_Rank_Total_Type</code>, where:
				<ul>
					<li> <code>rank</code> is the exon rank in the transcript (position in the transcript)
					<li> <code>total</code> is the total number of exons in that transcript
					<li> <code>type</code> is the exon splice type.
				</ul>
				<br>
				For instance <code>exon_5_10_RETAINED</code> would be the fifth exon in a 10 exon transcript. 
				This exon is type "RETAINED", which means it is not spliced out.
				<br>
				<br>
				Exons are categorized by splicing as follows:
				<ul>
					<li> <code>NONE</code> : Not spliced
					<li> <code>RETAINED</code> : All transcripts have this exon
					<li> <code>SKIPPED</code> : Some transcripts skip it
					<li> <code>ALTTENATIVE_3SS</code> : Some transcripts have and alternative 3' exon start 
					<li> <code>ALTTENATIVE_5SS</code> : Some transcripts have and alternative 5' exon end
					<li> <code>MUTUALLY_EXCLUSIVE</code> : Mutually exclusive (respect to other exon)
					<li> <code>ALTTENATIVE_PROMOMOTER</code> : The first exon is different in some transcripts.
					<li> <code>ALTTENATIVE_POLY_A</code> : The last exon.
				</ul>
				See <a href="http://en.wikipedia.org/wiki/Alternative_splicing#Modes"> this Wikipedia</a> entry for more information on exon splice types.
				</p>

				<p><h3>Intron naming convention</h3>
				Similarly to exons, introns are named as<code>intron_Rank_ExonTypeBefore-ExonTypeAfter</code>, where:
				<ul>
					<li> <code>Rank</code> :  the rank number for this inton in the transcript
					<li> <code>ExonTypeBefore</code> :  the splicing type of the exon preceding this intron (see exon naming convention for details).
					<li> <code>ExonTypeAfter</code> :  the splicing type of the after this intron (see exon naming convention for details).
				</ul>
				For instance <code>intron_9_SKIPPED-RETAINED</code> would be the ninth intron of the transcript. 
				The intron is preceded by a <code>SKIPPED</code> exon and followed by a <code>RETAINED</code> exon.
				
				</p>
        </section>


        <!-- Output Summary
        ================================================== -->
        <section id="outputSummary">
          <div class="page-header">
            <h1>8. Output summary files </h1>
          </div>
				<p class="lead"> 
				SnpEff creates an additional output file showing overall statistics.
				This "stats" file is an HTML file which can be opened using a web browser.
				You can find an example of a 'stats' file <a href="1kg.html"> here</a>.
				</p>
				<p>
				The program performs some statistics and saves them to the file 'snpEff_summary.html' on the 
				directory where snpEff is being executed. You can see the file, by opening it in your browser.
				<br>
				<br>
				<i class="icon-hand-right"></i>You can change the default location by using the '-stats' command line option. This also changes the location of the TXT summary file.
				<br>
				<br>
				<i class="icon-hand-right"></i> Summary can be create in CSV format using command line option <code>-csvStats</code>. This allows easy downstream processing.
				<br>
				<br>
				E.g.: In the stats file, you can see coverage histogram plots like this one <img src="http://chart.apis.google.com/chart?cht=bvg&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chg=11.0,40.0,3,2&chxp=1,50.0|3,50.0&chtt=Coverage+histogram&chxr=0,1.0,409.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Count|2:|1||||||||||11||||||||||21||||||||||31||||||||||41||||||||||51||||||||||61||||||||||76||||||||||3:|Coverage&chd=e:DNBSEfGaI9MKRSWZfXnrlwmZuE1wvW630e6O..638y3r7g8JsKxRlHgAkeZmY9XCR7PXRSOuOFMzNcIUJmHrHrFIFIFxEfCkDNDNBSBSB7BSB7BSApApApBSBSApAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA">
				</p>

				<br>
				<p> <b> "Effects by type" vs "Effects by region" </b>
				<br>
				<br>
				SnpEff annotates variants. 
				Variants produce effect of difference "types" (e.g. NON_SYNONYMOUS_CODING, STOP_GAINED). 
				These variants affect regions of the genome (e.g. EXON, INTRON). 
				The two tables count how many effects for each type and for each region exists.
				<br>
				E.g.: In an EXON region, you can have all the following effect types:	NON_SYNONYMOUS_CODING, SYNONYMOUS_CODING, FRAME_SHIFT, STOP_GAINED, etc.
				<br>
				<br>
				The complicated part is that some effect types affect a region that has the same name (yes, I know, this is confusing).
				<br>
				E.g.: In a UTR_5_PRIME region you can have UTR_5_PRIME and START_GAINED effect type. <br>
				<br>
				<br>
				This means that the number of both tables are not exactly the same, because the labels don't mean the same.
				See the next figure as an example
				<br>
				<img width=700 src="images/type_vs_region.png">
				<br>
				So the number of effects that affect a UTR_5_PRIME <i>region</i> is 206. Of those, 57 are effects <i>type</i> START_GAINED and 149 are effects <i>type</i> UTR_5_PRIME.
				<br>
				How exactly are effect <i>type</i> and effect <i>region</i> related? See the following table<p>
				<center>
				<table class="table table-striped">
					<tr> <th> Effect Type </th><th> Region </th> </tr> 
					 <tr>
						<td>
						NONE<br>
						CHROMOSOME<br>
						CUSTOM<br>
						CDS<br>
						</td>
						<td>NONE</td>
					</tr>
					<tr>
						<td>
						INTERGENIC<br>
						INTERGENIC_CONSERVED<br>
						</td>
						<td>INTERGENIC</td>
					</tr>
					<tr>
						<td>
						UPSTREAM<br>
						</td>
						<td>UPSTREAM</td>
					</tr>
					<tr>
						<td>
						UTR_5_PRIME<br>
						UTR_5_DELETED<br>
						START_GAINED<br>
						</td>
						<td>UTR_5_PRIME</td>
					</tr>
					<tr>
						<td>
						SPLICE_SITE_ACCEPTOR<br>
						</td>
						<td>SPLICE_SITE_ACCEPTOR</td>
					</tr>
					<tr>
						<td>
						SPLICE_SITE_DONOR<br>
						</td>
						<td>SPLICE_SITE_DONOR</td>
					</tr>
					<tr>
						<td>
						SPLICE_SITE_REGION<br>
						</td>
						<td>SPLICE_SITE_REGION</td>
					</tr>
					<tr>
						<td>
						INTRAGENIC<br>
						START_LOST<br>
						SYNONYMOUS_START<br>
						NON_SYNONYMOUS_START<br>
						GENE<br>
						TRANSCRIPT<br>
						</td>
						<td>EXON or NONE</td>
					</tr>
					<tr>
						<td>
						EXON<br>
						EXON_DELETED<br>
						NON_SYNONYMOUS_CODING<br>
						SYNONYMOUS_CODING<br>
						FRAME_SHIFT<br>
						CODON_CHANGE<br>
						CODON_INSERTION<br>
						CODON_CHANGE_PLUS_CODON_INSERTION<br>
						CODON_DELETION<br>
						CODON_CHANGE_PLUS_CODON_DELETION<br>
						STOP_GAINED<br>
						SYNONYMOUS_STOP<br>
						STOP_LOST<br>
						RARE_AMINO_ACID<br>
						</td>
						<td>EXON</td>
					</tr>
					<tr>
						<td>
						INTRON<br>
						INTRON_CONSERVED<br>
						</td>
						<td>INTRON</td>
					</tr>
					<tr>
						<td>
						UTR_3_PRIME<br>
						UTR_3_DELETED<br>
						</td>
						<td>UTR_3_PRIME</td>
					</tr>
					<tr>
						<td>
						DOWNSTREAM<br>
						</td>
						<td>DOWNSTREAM</td>
					</tr>
					<tr>
						<td>
						REGULATION<br>
						</td>
						<td>REGULATION</td>
					</tr>
				</table>
				</center>
				</p>
        </section>

        <!-- Effects
        ================================================== -->
        <section id="eff">
          <div class="page-header">
            <h1>9. Effect prediction details</h1>
          </div>
				<p class="lead">
				We describe the effects predicted by SnpEff
				</p>
				Notes:
				<ul>
					<li> <b> Sequence Ontology </b>
					Sequence ontology (<a href="http://www.sequenceontology.org/">SO</a>) allows to standardize terminology used for assessing sequence changes and impact. 
					This allows for a common language across all variant annotation programs and makes it easier to communicate using a uniform terminology.
					<br>
					<li> <b> Effect (Classic)</b> This are the "classic" effect names usd by SnpEff. We are moving towards a standardized classification system using Sequence Ontology.
					<br>
					<li><b>Effect impact</b>
					Effects are categorized by 'impact': {High, Moderate, Low, Modifier}. 
					This are pre-defined categories to help users find more significant variants.
					<br>
					<i class="icon-warning-sign"></i> Impact categories must be used with care, they were created only to help and simplify the filtering process. 
					Obviously, there is no way to predict whether a "high impact" or a "low impact" variant is the one producing you phenotype of interest.
					<br>
					<br>
				</ul>
				Here is a list of effects and some brief explanations:
				<br>
				<br>

				<center>
				<table class="table table-striped">
					<tr> <th> Effect<br>Seq. Ontology </th><th> Effect<br>Classic </th><th> Note &amp; Example </th><th>Impact</th> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001580"> coding_sequence_variant </a>  </td><td> CDS  </td><td> The variant hits a CDS. </td> <td> MODIFIER </td></tr> 
					<tr> <td> chromosome  </td><td> CHROMOSOME_LARGE DELETION  </td><td> A large parte (over 1%) of the chromosome was deleted. </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001580"> coding_sequence_variant </a>  </td><td> CODON_CHANGE  </td><td> One or many codons are changed <br>e.g.:  An MNP of size multiple of 3 </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001821"> inframe_insertion </a>  </td><td> CODON_INSERTION  </td><td> One or many codons are inserted <br>e.g.:  An insert multiple of three in a codon boundary </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001824"> disruptive_inframe_insertion</a>  </td><td> CODON_CHANGE_PLUS CODON_INSERTION  </td><td> One codon is changed and one or many codons are inserted <br>e.g.:  An insert of size multiple of three, not at codon boundary </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001822"> inframe_deletion </a>  </td><td> CODON_DELETION  </td><td> One or many codons are deleted <br>e.g.:  A deletion multiple of three at codon boundary </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001826"> disruptive_inframe_deletion </a>  </td><td> CODON_CHANGE_PLUS CODON_DELETION  </td><td> One codon is changed and one or more codons are deleted <br>e.g.:  A deletion of size multiple of three, not at codon boundary </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001632"> downstream_gene_variant </a>  </td><td> DOWNSTREAM  </td><td> Downstream of a gene (default length: 5K bases) </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001791"> exon_variant </a>  </td><td> EXON  </td><td> The vairant hits an exon. </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572"> exon_loss_variant </a>  </td><td> EXON_DELETED  </td><td> A deletion removes the whole exon. </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001589"> frameshift_variant </a>  </td><td> FRAME_SHIFT  </td><td> Insertion or deletion causes a frame shift <br>e.g.:  An indel size is not multple of 3 </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001564"> gene_variant </a>  </td><td> GENE  </td><td> The variant hits a gene. </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000605"> intergenic_region</a> </td><td> INTERGENIC </td><td> The variant is in an intergenic region </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002017"> conserved_intergenic_variant</a> </td><td> INTERGENIC_CONSERVED </td><td> The variant is in a highly conserved intergenic region </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002011"> intragenic_variant</a> </td><td> INTRAGENIC </td><td> The variant hits a gene, but no transcripts within the gene</td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001627"> intron_variant </a>  </td><td> INTRON  </td><td> Variant hits and intron. Technically, hits no exon in the transcript. </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002018"> conserved_intron_variant </a>  </td><td> INTRON_CONSERVED  </td><td> The variant is in a highly conserved intronic region </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000276"> miRNA </a>  </td><td> MICRO_RNA  </td><td> Variant affects an miRNA </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001583"> missense_variant </a>  </td><td> NON_SYNONYMOUS_CODING  </td><td> Variant causes a codon that produces a different amino acid <br>e.g.:  Tgg/Cgg, W/R </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001582"> initiator_codon_variant </a>  </td><td> NON_SYNONYMOUS_START  </td><td> Variant causes start codon to be mutated into another start codon (the new codon produces a different AA). <br>e.g.: Atg/Ctg, M/L (ATG and CTG can be START codons) </td> <td> LOW </td> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001567"> stop_retained_variant </a>  </td><td> NON_SYNONYMOUS_STOP  </td><td> Variant causes stop codon to be mutated into another stop codon (the new codon produces a different AA). <br>e.g.: Atg/Ctg, M/L (ATG and CTG can be START codons) </td> <td> LOW </td> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002008">rare_amino_acid_variant</a> </td><td> RARE_AMINO_ACID </td><td> The variant hits a rare amino acid thus is likely to produce protein loss of function </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001574"> splice_acceptor_variant </a>  </td><td> SPLICE_SITE_ACCEPTOR  </td><td> The variant hits a splice acceptor site (defined as two bases before exon start, except for the first exon). </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001575"> splice_donor_variant </a>  </td><td> SPLICE_SITE_DONOR  </td><td> The variant hits a Splice donor site (defined as two bases after coding exon end, except for the last exon). </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630"> splice_region_variant </a>  </td><td> SPLICE_SITE_REGION  </td><td> A sequence variant in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8 bases of the intron.</td> <td> LOW </td> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630"> splice_region_variant </a>  </td><td> SPLICE_SITE_BRANCH  </td><td> A varaint affective putative (Lariat) branch point, located in the intron.</td> <td> LOW </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630"> splice_region_variant </a>  </td><td> SPLICE_SITE_BRANCH_U12  </td><td> A varaint affective putative (Lariat) branch point from U12 splicing machinery, located in the intron.</td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001578"> stop_lost </a>  </td><td> STOP_LOST  </td><td> Variant causes stop codon to be mutated into a non-stop codon <br>e.g.: Tga/Cga, */R </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001988"> 5_prime_UTR_premature start_codon_gain_variant </a>  </td><td> START_GAINED  </td><td> A variant in 5'UTR region produces a three base sequence that can be a START codon. </td> <td> LOW </td> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002012"> start_lost</a> </td><td> START_LOST   </td><td> Variant causes start codon to be mutated into a non-start codon. <br> e.g.: aTg/aGg, M/R </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001587"> stop_gained </a>  </td><td> STOP_GAINED  </td><td> Variant causes a STOP codon <br>e.g.: Cag/Tag, Q/* </td> <td> HIGH </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001819"> synonymous_variant </a>  </td><td> SYNONYMOUS_CODING  </td><td> Variant causes a codon that produces the same amino acid <br>e.g.:  Ttg/Ctg, L/L </td> <td> LOW </td> </tr> 
					<tr> <td> <a href=""> start_retained </a>  </td><td> SYNONYMOUS_START  </td><td> Variant causes start codon to be mutated into another start codon. <br>e.g.:  Ttg/Ctg, L/L (TTG and CTG can be START codons) </td> <td> LOW </td> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001567"> stop_retained_variant </a>  </td><td> SYNONYMOUS_STOP  </td><td> Variant causes stop codon to be mutated into another stop codon. <br>e.g.: taA/taG, */* </td> <td> LOW </td> </tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001576"> transcript_variant </a>  </td><td> TRANSCRIPT  </td><td> The variant hits a transcript. </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001566"> regulatory_region_variant </a>  </td><td> REGULATION  </td><td> The variant hits a known regulatory feature (non-coding). </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001631"> upstream_gene_variant </a>  </td><td> UPSTREAM  </td><td> Upstream of a gene (default length: 5K bases) </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001624"> 3_prime_UTR_variant </a>  </td><td> UTR_3_PRIME  </td><td> Variant hits 3'UTR region </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002015"> 3_prime_UTR_truncation</a> + <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572"> exon_loss </a>  </td><td> UTR_3_DELETED  </td><td> The variant deletes an exon which is in the 3'UTR of the transcript </td> <td> MODERATE </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001623"> 5_prime_UTR_variant </a>  </td><td> UTR_5_PRIME  </td><td> Variant hits 5'UTR region </td> <td> MODIFIER </td></tr> 
					<tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002013"> 5_prime_UTR_truncation</a> + <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a> </td><td> UTR_5_DELETED  </td><td> The variant deletes an exon which is in the 5'UTR of the transcript </td> <td> MODERATE </td></tr> 
				</table> 
				</center>

				<br>
				<p> <b>Details about Rare amino acid effect </b>
				<br>
				<br>
				These are amino acids that occurs very rarely in an organism.
				For instance, humans are supposed to use 20 amino acids, but
				there is also one rare AA. Selenocysteine, single letter
				code 'U', appears roughly 100 times in the whole genome.
				The amino acid is so rare that usually it does not appear 
				in codon translation tables. It is encoded as UGA, which usually 
				means a STOP codon. Secondary RNA 
				structures are assumed to enable this special translation.
				<br>
				<br>
				A variant in one of these sites is likely to cause a loss of 
				function in the protein. E.g. in case of a Selenocysteine, a 
				loss of a selenium molecule is likely to cause loss of function.
				Put it simply, the assumption is that there is a great deal of trouble 
				to get that non-standard amino acid there, so it must be important.
				RARE_AMINO_ACID mark is used to show that special attention should 
				be paid in these cases.
				<br>
				<br>
				<i class="icon-warning-sign"></i> When the variant hits a RARE_AMINO_ACID mark, it is likely that 
				the 'old_AA/new_AA' field will be incorrect. This may happen because 
				the amino acid is not predictable using a codon table. 
				</p>
				<br>
				<p> <h3> Loss of function and nonsense-mediated decay predictions </h3>
				Loss of function ('LOF') and nonsense-mediated decay ('NMD') predictions can be activated by using <code>-lof</code> command line option.
				Details on how these variants work, can be found in <a class="body" href="snpEff_lof_nmd.pdf">these slides </a>
				<br>
				Usage example:
<pre>
java Xmx4g -jar snpEff.jar v \
	-lof \
	GRCh37.68 \
	file.vcf.gz > file.eff.vcf
</pre>

				SnpEff adds 'LOF' and 'NMD' tags to INFO fields (column 8 in VCF format).  LOF and NMD tags have the following format:
<pre>
Gene | ID | num_transcripts | percent_affected
</pre>
				Where:
				<table class="table table-striped">
					<tr> <th> Field            </th><th> Description </th> </tr>
					<tr> <td> Gene             </td><td> Gene name </td> </tr>
					<tr> <td> ID               </td><td> Gene ID (usually ENSEMBL) </td> </tr>
					<tr> <td> Num_transcripts  </td><td> Number of transcripts in this gene </td> </tr>
					<tr> <td> percent_affected </td><td> Percentage of transcripts affected by this variant. </td> </tr>
				</table>

				<br>
				Example: If we have this effect
<pre>
EFF=, SPLICE_SITE_DONOR(HIGH||||639|ILDR2|protein_coding|CODING|ENST00000271417|1)
</pre>

				and the corresponding LOF tag is
<pre>
LOF=ILDR2|ENSG00000143195|7|1.00
</pre>

				The meaning of the LOF tag is:
				<table class="table table-striped">
					<tr> <th> Field            </th><th> Description </th> </tr>
					<tr> <td> Gene             </td><td> ILDR2 </td> </tr>
					<tr> <td> ID               </td><td> ENSG00000143195 </td> </tr>
					<tr> <td> Num_transcripts  </td><td> There are 7 transcripts in this gene </td> </tr>
					<tr> <td> percent_affected </td><td> 100% of transcripts are affected by this variant. </td> </tr>
				</table>
				</p>

				<h3> Troubleshooting Effect predictions </h3>

				<p> <b> Chromosome not found </b>
				<br>
				<br>
				<i class="icon-warning-sign"></i> This is by far the most common problem. 
				It means that the input VCF file has chromosome names that do not match SnpEff's database and don't match reference genome either, since SnpEff's database are created using reference genome chromosome names.
				<br>
				<br>
				The solution is simple: fix your VCF file to use standard chromosome names.
				You can see which chromosome names are used by SnpEff simply by using the <code>-v</code> (verbose) command line option.
				This shows all chromosome names and their respective lengths. Notice the last line ("Chromosomes names [sizes]"):
<pre>
$ java -Xmx4g eff -v GRCh37.70 test.1K.vcf 
00:00:00.000    Reading configuration file 'snpEff.config'
00:00:00.176    done
00:00:00.176    Reading database for genome version 'GRCh37.70' from file '/home/pcingola//snpEff/data/GRCh37.70/snpEffectPredictor.bin' (this might take a 
00:01:13.513    done
00:01:13.555    Setting '-treatAllAsProteinCoding' to 'false'
00:01:13.556    Building interval forest
00:01:32.642    done.
00:01:38.280    Genome stats :
# Genome name                : 'Homo_sapiens'
# Genome version             : 'GRCh37.70'
# Has protein coding info    : true
# Genes                      : 62069
# Protein coding genes       : 22643
# Transcripts                : 206046
# Avg. transcripts per gene  : 3.32
# Protein coding transcripts : 85685
# Cds                        : 756629
# Exons                      : 1262644
# Exons with sequence        : 1262644
# Exons without sequence     : 0
# Avg. exons per transcript  : 6.13
# Number of chromosomes      : 275
# Chromosomes names [sizes]  : '1' [249250621] '2' [243199373] '3' [198022430] '4' [191154276] '5' [180915260] '6' [171115067] '7' [159138663] 'X' [155270560] '8' [146364022] '9' [141213431] '10' [135534747] '11' [135006516] '12' [133851895] '13' [115169878] '14' [107349540] '15' [102531392] '16' [90354753] '17' [81195210] '18' [78077248] '20' [63025520] 'Y' [59373566] '19' [59128983] '22' [51304566] '21' [48129895] 'HG1287_PATCH' [249889652]  'HG1292_PATCH' [249822337]  'HG1473_PATCH' [249272860]  'HG1471_PATCH' [249269426] ...
...
</pre>
				</p>

				<p> <b> Apparent inconsistencies when using UCSC genome browser </b>
				<br>
				<br>
				<i class="icon-warning-sign"></i> <b>WARNING:</b> Usage of hg19 genome is deprecated and discouraged, you should use GRChXX.YY instead (e.g. the latest version at the time of writing is GRCh37.70)
				<br>
				<br>
				Reference sequence and annotations are made for an organism version and sub-version. 
				For examples human genome, version 37, sub-version 70 would be called (GRCh37.70).
				<br>
				UCSC doesn't specify sub-version. 
				They just say hg19.
				This annoying sub-version problem appeared often and, having reproducibility of results in mind, I dropped UCSC annotations in favor of ENSEMBL ones (they have clear versioning).
				</p>

				<p> <b> SnpEff reporting an effect that doesn't match ENSEMBL's web page </b>
				<br>
				<br>
				Please remember that databases are updated often (e.g. by ENSEMBL), so if you are using an old database, you might get different effects.<br>
				<br>
				For example, this transcript ENST00000487462 changed from protein_coding in GRCh37.63 
<pre>
1       protein_coding  exon    1655388 1655458 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "1"; gene_name "CDK11A"; transcript_name "CDK11A-013";
1       protein_coding  exon    1653905 1654270 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "2"; gene_name "CDK11A"; transcript_name "CDK11A-013";
</pre>
...to processed_transcript in GRCh37.64:
<pre>
1       processed_transcript    exon    1655388 1655458 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "1"; gene_name "CDK11A"; gene_biotype "protein_coding"; transcript_name "CDK11A-013";
1       processed_transcript    exon    1653905 1654270 .       -       .        gene_id "ENSG00000008128"; transcript_id "ENST00000487462"; exon_number "2"; gene_name "CDK11A"; gene_biotype "protein_coding"; transcript_name "CDK11A-013";
</pre>

				This means that you'll get different results for this transcript using sub-version 63 or 64. I assume that latest versions are improved, so I always encourage to upgrade.
				<br>
				Sometimes it might even be the case that latest released database and the one shown on the web interface may be out of sync.
				<br>
				</p>

				<p> <b> SnpEff reports a SYNONYMOUS and a NON_SYNONYMOUS effect on the same gene </b>
				<br>
				<br>
				This is not a bug.
				It is not uncommon for a gene to have more than one transcript (e.g. in human most genes have multiple transcripts). 
				A variant (e.g. a SNP) might affect different transcripts in different ways, as a result of different reading frames.
				<br>
				For instance: 
<pre>
chr5 137622242 . C T . . EFF=NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Gaa/Aaa|E/K|CDC25C|protein_coding|CODING|ENST00000514017|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000323760|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000348983|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000356505|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000357274|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000415130|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000513970|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000514555|exon_5_137622186_137622319),
                             SYNONYMOUS_CODING(LOW|SILENT|caG/caA|Q|CDC25C|protein_coding|CODING|ENST00000534892|exon_5_137622186_137622319)
</pre>

				in this example (it was divided into multiple lines for legibility), the first transcript ENST0000051401 has a NON_SYNONYMOUS effect, but all other transcripts have a SYNONYMOUS effect.
				</p>
        </section>


        <!-- Non-coding
        ================================================== -->
        <section id="effNc">
          <div class="page-header">
            <h1>10. Non-coding &amp; Regulatory annotations</h1>
          </div>
				<p class="lead">
				SnpEff can also provide non-coding and regulatory annotations.
				Here we show how to annotate them.
				</p>
				<p>
				<i class="icon-warning-sign"></i> Non-coding and regulatory annotations databases are available only for a few organisms (e.g. human, mouse, etc.).
				We intend to incorporate more non-coding annotations as soon as public databases are available, but your organism of choice might not have a non-coding/regulatory database available.
				</p>
				<p>
				First of all, you need to see if your organism has a regulatory database.
				You can just look into the database directory to see if <code>regulation_*.bin</code> files are there.
				For instance, for human genome:
<pre>
$ cd ~/snpEff
$ cd data/GRCh37.70/
$ ls -al
total 2399156
drwxr-xr-x    2 pcingola pcingola       4096 Feb 11 22:01 .
drwxr-xr-x 2472 pcingola pcingola     122880 Feb 11 19:55 ..
-rw-r--r--    1 pcingola pcingola    6292802 Feb 11 22:01 regulation_CD4.bin
-rw-r--r--    1 pcingola pcingola    2474396 Feb 11 22:01 regulation_GM06990.bin
-rw-r--r--    1 pcingola pcingola    7924641 Feb 11 22:00 regulation_GM12878.bin
-rw-r--r--    1 pcingola pcingola    6212489 Feb 11 22:01 regulation_H1ESC.bin
-rw-r--r--    1 pcingola pcingola    5152384 Feb 11 22:00 regulation_HeLa-S3.bin
-rw-r--r--    1 pcingola pcingola    7392256 Feb 11 22:01 regulation_HepG2.bin
-rw-r--r--    1 pcingola pcingola    3991817 Feb 11 22:01 regulation_HMEC.bin
-rw-r--r--    1 pcingola pcingola    4562241 Feb 11 22:01 regulation_HSMM.bin
-rw-r--r--    1 pcingola pcingola    5550176 Feb 11 22:01 regulation_HUVEC.bin
-rw-r--r--    1 pcingola pcingola    5528240 Feb 11 22:01 regulation_IMR90.bin
-rw-r--r--    1 pcingola pcingola     534786 Feb 11 22:01 regulation_K562b.bin
-rw-r--r--    1 pcingola pcingola    8371354 Feb 11 22:01 regulation_K562.bin
-rw-r--r--    1 pcingola pcingola    3063189 Feb 11 22:01 regulation_NH-A.bin
-rw-r--r--    1 pcingola pcingola    5621629 Feb 11 22:00 regulation_NHEK.bin
-rw-r--r--    1 pcingola pcingola   89937614 Feb 11 21:57 snpEffectPredictor.bin
</pre>

				So we can annotate using any of those tracks.
				<br>
				E.g. To use 'HeLa-S3' and 'NHEK' tracks, you can run:
<pre>
java -Xmx4g -jar snpEff.jar eff -reg HeLa-S3 -reg NHEK GRCh37.70 in.vcf > out.vcf
</pre>
				<br>
				<i class="icon-hand-right"></i> You can use <code>-onlyReg</code> command line option to perform only regulatory annotations (i.e. coding annotations will be omitted).
				</p>
        </section>

        <!-- Options
        ================================================== -->
        <section id="filters">
          <div class="page-header">
            <h1>11. Annotations options </h1>
          </div>
				<p class="lead">
				We describe how to change some defaults and filter input and outputs to narrow down the number of results.
				</p>

				<h3>Speed up options: Multithreaded &amp; No statistics </h3>
				<p>
				In order to speed up the annotation process, there are two options that can be activated:
				<ul>
					<li> No statistics: Calculating statistics can take a significant amount of time. 
							Particularly if there are hundreds or thousands of samples in the VCF file.
							The command line option <code>-noStats</code> disables the statistics and may result in a significant speedup.

					<li> Mutlithreaded: SnpEff has a mutlithreaded implementation. 
							This allows to use several cores available in the local machine. 
							It can be activated using the <code>-t</code> command line option.

							<br>
							<br>
							<i class="icon-warning-sign"></i> Some features are not available when running in mutlithreaded mode. 
							For instance, statistics are disabled in order to speed up the annotation process.
							<br>
							<br>
							<i class="icon-warning-sign"></i> In mutlithreaded mode, SnpEff will attempt to use all available cores. 
							Nevertheless is very rare that all cores get 100% usage.
				</ul>
				</p>

				<h3>HGVS notation</h3>
				<p>
				SnpEff partially implements <a href="http://www.hgvs.org/">HGVS notation</a>, which is somewhat popular amongst clinicians.
				You can activate HGVS notation using the command line option <code>-hgvs</code>.
				When activated, the "AA" sub-field in the EFF field (VCF file, INFO column) will be written in HGVS notation.
				<br>
				<br>
				<i class="icon-warning-sign"></i> HGVS notation is only partially implemented, some cases may be missing.
				</p>

				<h3>Logging</h3>
				<p>
				SnpEff will try to log usage statistics to our "log server".
				This is useful for us to understand user's needs and have some statistics on what users are doing with the program (e.g. decide whether a command or option is useful or not).
				<br>
				Logging can be deactivated by using the <code>-noLog</code> command line option.
				</p>

				<h3>Input filters</h3>
				<p>
				SnpEff supports filter on input files by using combinations of the following command line options:
				<br>
				<br>
				<i class="icon-warning-sign"></i> Input filters can be implemented using <code>SnpSift filter</code>, which allows to create more flexible and complex filters.
				<br>
				<br>
				<table class="table table-striped">
					<tr> <th>	Command line option     </th><th> Meaning </th></tr>
					<tr> <td>	-del                    </td><td> Analyze deletions only </td></tr>
					<tr> <td>	-ins                    </td><td> Analyze insertions only </td></tr>
					<tr> <td>	-hom                    </td><td> Analyze homozygous variants only </td></tr>
					<tr> <td>	-het                    </td><td> Analyze heterozygous variants only </td></tr>
					<tr> <td>	-minQ X, -minQuality X  </td><td> Filter out variants with quality lower than X </td></tr>
					<tr> <td>	-maxQ X, -maxQuality X  </td><td> Filter out variants with quality higher than X </td></tr>
					<tr> <td>	-minC X, -minCoverage X </td><td> Filter out variants with coverage lower than X </td></tr>
					<tr> <td>	-maxC X, -maxCoverage X </td><td> Filter out variants with coverage higher than X </td></tr>
					<tr> <td>	-mnp                    </td><td> Only MNPs (multiple nucleotide polymorphisms) </td></tr>
					<tr> <td>	-snp                    </td><td> Only SNPs (single nucleotide polymorphisms) </td></tr>
				</table>

				<h3>Output filters</h3>
				<p>
				SnpEff supports filter of output results by using combinations of the following command line options:
				<br>
				<br>
				<i class="icon-warning-sign"></i> Output filters can be implemented using <code>SnpSift filter</code>, which allows to create more flexible and complex filters.
				<br>
				<br>
				<table class="table table-striped">
					<tr> <th>	Command line option     </th><th> Meaning </th></tr>
					<tr> <td>	-no-downstream                  </td><td> Do not show DOWNSTREAM changes </td></tr>
					<tr> <td>	-no-intergenic                  </td><td> Do not show INTERGENIC changes </td></tr>
					<tr> <td>	-no-intron                      </td><td> Do not show INTRON changes </td></tr>
					<tr> <td>	-no-upstream                    </td><td> Do not show UPSTREAM changes </td></tr>
					<tr> <td>	-no-utr                         </td><td> Do not show 5_PRIME_UTR or 3_PRIME_UTR changes </td></tr>
					<tr> <td>	-no            EffectType       </td><td> Do not show 'EffectType' (it can be used several times) <br>e.g: <code>-no INTERGENIC -no SPLICE_SITE_REGION</code> </td></tr>
				</table>
				</p>

				<h3>Annotating selected intervals </h3>
				<p>
				You can use the <code>-fi intervals.bed</code> command line option (filterInterval). For instance, let's assume you have an interval file 'intervals.bed':
<pre>2L	10000	10999
2L	12000	12999
2L	14000	14999
2L	16000	16999
2L	18000	18999 </pre>
				</p>

				<h3>Canonical transcripts </h3>
				<p>
				SnpEff allows to annotate using canonical transcripts by using <code>-canon</code> command line option.
				<br>
				<br>
				<i class="icon-warning-sign"></i>Canonical transcripts are defined as the longest CDS of amongst the protein coding transcripts in a gene.
				If none of the transcripts in a gene is protein coding, then it is the longest cDNA. 
				<br>
				<br>
				<i class="icon-warning-sign"></i>Although this seems to be the standard definitions of "canonical transcript", there is no warranties that what SnpEff considers a canonical transcript will match exactly what UCSC or ENSEMBL consider a canonical transcript.
				</p>
				Example on how to use canonical transcripts annotations:
<pre> java -Xmx4G -jar snpEff.jar -canon GRCh37.70 file.vcf > file.snpeff.canon.vcf</pre>
				<br>
				In order to get only changes that match your intervals, you can use the command:
<pre> java -Xmx4G -jar snpEff.jar -fi intervals.bed dm5.30 file.vcf </pre>

				<br>
				In order to get a list of canonical transcipts, you can use the <code>-d</code> (debug) command line option. E.g.:
<pre>
$ java -Xmx4G -jar snpEff.jar eff -d -v -canon GRCh37.66 test.vcf
00:00:00.000    Reading configuration file 'snpEff.config'
00:00:00.173    done
00:00:00.173    Reading database for genome version 'GRCh37.66'
00:00:02.834    done
00:00:02.845    Filtering out non-canonical transcripts.
00:00:03.219    Canonical transcripts:
                geneName        geneId          transcriptId    cdsLength
                GGPS1           ENSG00000152904 ENST00000488594 903
                RP11-628K18.1.1 ENSG00000235112 ENST00000430808 296
                MIPEPP2         ENSG00000224783 ENST00000422560 1819
                FEN1P1          ENSG00000215873 ENST00000401028 1145
                AL591704.7.1    ENSG00000224784 ENST00000421658 202
                CAPNS1P1        ENSG00000215874 ENST00000401029 634
                ST13P20         ENSG00000215875 ENST00000447996 1061
                NCDN            ENSG00000020129 ENST00000373243 2190
                RP11-99H8.1.1   ENSG00000226208 ENST00000423187 432
                AL391001.1      ENSG00000242652 ENST00000489859 289
...
</pre> 


				<h3>Selected list of transcripts </h3>
				<p>
				SnpEff allows you to provide a list of transcripts to use for annotations by using the <code>-onlyTr file.txt</code> and providing a file with one transcript ID per line.
				Any other transcript will be ignored.
				<pre> java -Xmx4G -jar snpEff.jar -onlyTr my_transcripts.txt GRCh37.70 file.vcf > file.snpeff.vcf</pre>
				</p>

				<h3>Upstream and downstream</h3>
				<p>
				You can change the default upstream and downstream interval size (default is 5K) using the <code>-ud size_in_bases</code> option.
				This also allows to eliminate any upstream and downstream effect by using "-ud 0".
				<br>
				<br>
				Example: Make upstream and downstream size zero (i.e. do not report any upstream or downstream effect).
				<pre> java -Xmx4G -jar snpEff.jar -ud 0 GRCh37.70 file.vcf > file.snpeff.vcf</pre>
				</p>

				<h3>Splice site size</h3>
				<p>
				You can change the default splice site size (default is 2 bases) using the <code>-spliceSiteSize size_in_bases</code> option.
				<br>
				<br>
				Example: Make splice sites four bases long
				<pre> java -Xmx4G -jar snpEff.jar -spliceSiteSize 4 GRCh37.70 file.vcf > file.snpeff.vcf</pre>
				</p>

				<h3>Adding your own annotations</h3>
				<p>
				SnpEff allows user defined intervals to be annotated.
				This is achieved using the <code>-interval file.bed</code> command line option, which can be used multiple times in the same command line.
				Any variant that intersects an interval defined in those files, will be annotated using the "name" field (fourth column) in the input bed file.
				<br>
				<br>
				Example: We create our own annotations in <code>my_annotations.bed</code>
<pre>
$ cat my_annotations.bed 
1	10000	20000	MY_ANNOTATION

$ cat test.vcf 
1	10469	.	C	G	365.78	PASS	AC=30;AF=0.0732

$ java -Xmx4g -jar snpEff.jar eff -interval my_annotations.bed GRCh37.66 test.vcf
1	10469	.	C	G	365.78	PASS	AC=30;AF=0.0732;EFF=CUSTOM(MODIFIER||||||MY_ANNOTATION||||1),DOWNSTREAM(MODIFIER|||||WASH7P|unprocessed_pseudogene|NON_CODING|ENST00000423562||1),DOWNSTREAM(MODIFIER|||||WASH7P|unprocessed_pseudogene|NON_CODING|ENST00000438504||1),DOWNSTREAM(MODIFIER|||||WASH7P|unprocessed_pseudogene|NON_CODING|ENST00000488147||1),DOWNSTREAM(MODIFIER|||||WASH7P|unprocessed_pseudogene|NON_CODING|ENST00000538476||1),DOWNSTREAM(MODIFIER|||||WASH7P|unprocessed_pseudogene|NON_CODING|ENST00000541675||1),INTERGENIC(MODIFIER||||||||||1),UPSTREAM(MODIFIER|||||DDX11L1|processed_transcript|NON_CODING|ENST00000456328||1),UPSTREAM(MODIFIER|||||DDX11L1|transcribed_unprocessed_pseudogene|NON_CODING|ENST00000450305||1),UPSTREAM(MODIFIER|||||DDX11L1|transcribed_unprocessed_pseudogene|NON_CODING|ENST00000515242||1),UPSTREAM(MODIFIER|||||DDX11L1|transcribed_unprocessed_pseudogene|NON_CODING|ENST00000518655||1)
</pre>
				Notice that the variant was annotated using "MY_ANNOTATION" in the EFF field.
				</p>

				<h3>Gene ID instead of gene names</h3>
				<p>
				<br>
				You can obtain gene IDs instead of gene names by using the command line option <code>-geneId</code>.
				<br>
				Example:
<pre>
$ java -Xmx4g -jar snpEff.jar -geneId GRCh37.66 test.vcf 
1  902128  3617  C  T  .  PASS  AC=80;EFF=NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|gCt/gTt|A43V|576|ENSG00000187583|protein_coding|CODING|ENST00000379407|2|1),...
</pre>
				Note: The gene 'PLEKHN1' was annotated as 'ENSG00000187583'.

				</p>
				<h3>Chromosome names</h3>
				<p>
				<i class="icon-warning-sign"></i> This is only used in TXT output format, which is deprecated.
				<br>
				<br>
				By default SnpEff simplifies all chromosome names. For instance 'chr1' is just '1'. 
				Some people just hate this, so you can prepend any string you want to the chromosome name by using the <code>-chr string</code>
				<br>
				<br>
				Example:
<pre>
$ java -Xmx4g -jar snpEff.jar eff -v -o txt -chr "CHROMOSOME" GRCh37.66 test.vcf 
# Chromo	Position	Reference	Change	Change_type	Homozygous	Quality	Coverage	Warnings	Gene_ID	Gene_name	Bio_type	Transcript_ID	Exon_ID	Exon_Rank	Effect	old_AA/new_AA	Old_codon/New_codon	Codon_Num(CDS)	Codon_Degeneracy	CDS_size	Codons_around	AAs_around	Custom_interval_ID
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000227232	WASH7P	unprocessed_pseudogene	ENST00000423562			DOWNSTREAM: 3894 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000227232	WASH7P	unprocessed_pseudogene	ENST00000438504			DOWNSTREAM: 3894 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000227232	WASH7P	unprocessed_pseudogene	ENST00000541675			DOWNSTREAM: 3894 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000227232	WASH7P	unprocessed_pseudogene	ENST00000488147			DOWNSTREAM: 3935 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000227232	WASH7P	unprocessed_pseudogene	ENST00000538476			DOWNSTREAM: 3942 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000223972	DDX11L1	transcribed_unprocessed_pseudogene	ENST00000518655			UPSTREAM: 1405 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000223972	DDX11L1	transcribed_unprocessed_pseudogene	ENST00000450305			UPSTREAM: 1541 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0								INTERGENIC			
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000223972	DDX11L1	processed_transcript	ENST00000456328			UPSTREAM: 1400 bases								
CHROMOSOME1	10469	C	G	SNP	Hom	365.78	0		ENSG00000223972	DDX11L1	transcribed_unprocessed_pseudogene	ENST00000515242			UPSTREAM: 1403 bases								
</pre>
				</p>
        </section>


        <!-- Cancer
        ================================================== -->
        <section id="cancer">
          <div class="page-header">
            <h1>12. Cancer samples </h1>
          </div>
				<p class="lead">
				Here we describe details about annotating cancer samples.
				</p>

				<h3>Representing cancer data</h3>
				<p>
				In a typical cancer sequencing experiment, we want to measure and annotate differences between germline (healthy) and somatic (cancer) tissue samples from the same patient.
				The complication is that germline is not always the same as the reference genome, so a typical annotation does not work.
				<br>
				For instance, let's assume that at a given genomic position (e.g. chr1:69091), reference genome is 'A', germline is 'C' and somatic is 'C'.
				This should be represented in a VCF file as:
<pre>
#CHROM  POS    ID  REF  ALT    QUAL  FILTER  INFO  FORMAT  Patient_01_Germline  Patient_01_Somatic
1       69091  .   A    C,G    .     PASS    AC=1  GT      1/0                  2/1
</pre>
				<br>
				<i class="icon-warning-sign"></i> Some people tend to represent this by changing REF base 'A' using germline 'C'. 
				This is a mistake, REF must always represent the reference genome, not one of your samples.
				<br>
				<br>
				Under normal conditions, SnpEff would provide the effects of changes "A -&gt; C" and "A -&gt; G". 
				But in case of cancer samples, we are actually interested in the difference between somatic and germline, so we'd like to calculate the effect of a "C -&gt; G" mutation.
				Calculating this effect is not trivial, since we have to build a new "reference" by calculating the effect of the first mutation ("A -&gt; C") and then calculate the effect of the second one ("C -&gt; G") on our "new reference".
				<br>
				<br>
				<i class="icon-hand-right"></i> In order to activate cancer analysis, you must use <code>-cancer</code> command line option.
				<br>
				<br>
				<h3>Defining cancer samples</h3>
				As we already mentioned, cancer data is represented in a VCF file using multiple ALTs (REF field always is reference genome).
				In order to specify which samples are somatic and which ones are germline, there are two options:
				<br>
				<br>
				<ul>
					<li> Use a TXT file using <code>-canceSample</code> command line option.
					<li> Use the PEDIGREE meta information in your VCF file's header. This is the default, but some people might find hard to edit / change information in VCF file's headers.
				</ul>
				<br>
				<i class="icon-warning-sign"></i> If you do not provide either <code>PEDIGREE</code> meta information or a TXT samples file, SnpEff will not know which somatic samples derive from which germline samples. 
				Thus it will be unable to perform cancer effect analysis.
				<br>
				<br>
				<b> TXT file </b>
				<br>
				This is quite easy. 
				All you have to do is to create a tab-separated TXT file having two columns: the first column has the germline sample names and the second column has the somatic sample names.
				Make sure that sample names match exactly the ones in the VCF file.
				<br>
				<br>
				E.g.: Create a TXT file named 'samples_cancer.txt'
<pre>
Patient_01_Germline	Patient_01_Somatic
Patient_02_Germline	Patient_02_Somatic
Patient_03_Germline	Patient_03_Somatic
Patient_04_Germline	Patient_04_Somatic
</pre>

				Then you have to specify this TXT file when invoking SnpEff, using the <code>-canceSample</code> command line option.
				<br>
				<br>
				E.g. In our example, the file name is 'samples_cancer.txt', so the command line would look like this:
<pre>
java -Xmx4g -jar snpEff.jar eff -v -cancer -canceSample samples_cancer.txt GRCh37.70 cancer.vcf > cancer.snpeff.vcf 
</pre>
				<br>
				<br>
				<b> VCF header </b>
				<br>
				This is the default method and the main advantage is that you don't have to carry information on a separate TXT file (all the information is within your VCF file).
				You have to add the <code>PEDIGREE</code> header with the appropriate information to your VCF file.
				Obviously this requires you to edit you VCF file's header.
				<br>
				<br>
				<i class="icon-warning-sign"></i> How to edit VCF  headers is beyond the scope of this manual (we recommend using <code>vcf-annotate</code> from <a href="http://vcftools.sourceforge.net/docs.html#annotate">VCFtools</a>). But if you find adding PEDIGREE information to your VCF file difficult, just use the TXT file method described in the previous sub-section.
				<br>
				<br>
				E.g.: Pedigree information in a VCF file would look like this
<pre>
##PEDIGREE=&lt;Derived=Patient_01_Somatic,Original=Patient_01_Germline>
#CHROM  POS ID  REF ALT QUAL    FILTER  INFO        FORMAT  Patient_01_Germline Patient_01_Somatic
1   69091   .   A   C,G .       PASS    AF=0.1122   GT      1/0                 2/1
1   69849   .   G   A,C .       PASS    AF=0.1122   GT      1/0                 2/1
1   69511   .   A   C,G .       PASS    AF=0.3580   GT      1/1                 2/2
</pre>
				Here we say that the sample called <code>Patient_01_Somatic</code> is derived from the sample called <code>Patient_01_Germline</code>. 
				In this context, this means that cancer sample is derived from the healthy tissue. 

				<br>
				<br>

				<h3>Running in cancer analysis mode</h3>
				We added '-cancer'  command line option for somatic vs germline comparisons. 
				By default SnpEff does not perform them because they computationally intensive. 
				So your command line would be:

				<pre> java -Xmx4g -jar snpEff.jar eff -v -cancer GRCh37.70 cancer.vcf > cancer.snpeff.vcf </pre>

				<br>
				<i class="icon-hand-right"></i> HGVS notations seems to be preferred in the clinical and cancer community. 
				You can activate it by using <code>-hgvs</code> to provide HGVS notation in AA sub-field. 
				<br>
				<br>

				<h3>Interpreting Cancer effects</h3>
				Interpretation of EFF field cancer sample relies heavily on 'Genotype' sub-field.
				Just as a reminder, EFF field has the following format:

<pre>
EFF= Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_Change| Amino_Acid_Length | Gene_Name | Transcript_BioType | Gene_Coding | Transcript_ID | Exon_Rank  | Genotype_Number [ | ERRORS | WARNINGS ] )
</pre>
				The <code>GenotypeNum</code> field tells you which effect relates to which genotype. 
				More importantly, genotype difference between Somatic and Germline. 
				<br>
				Example: when there are multiple ALTs (e.g. REF='A' ALT='C,G') and the genotype field says:
				<ul>
					<li> GenotypeNum = "1": it means is the effect related to the first ALT ('C')
					<li> GenotypeNum = "2" if it's the effect related to the second ALT ('G') 
					<li> GenotypeNum = "2-1" means that this is the effect of having the second ALT as variant while using the first ALT as reference ("C -&gt; G").
				</ul>
				It is important that you understand the meaning of the last one, because you'll use it often for your cancer analysis.
				<br>
				<br>
				Example: Sample output for the previously mentioned VCF example would be (the output has been edited for readability reasons)
				<br>
				For the first line:
<pre>
1   69091   .   A   C,G .   PASS    AF=0.1122
    ;EFF=NON_SYNONYMOUS_START(LOW|MISSENSE|Atg/Ctg|M1L|305|OR4F5|protein_coding|CODING|ENST00000335137|1|1)
        ,START_LOST(HIGH|MISSENSE|Atg/Gtg|M1V|305|OR4F5|protein_coding|CODING|ENST00000335137|1|2)
        ,START_LOST(HIGH|MISSENSE|Ctg/Gtg|L1V|305|OR4F5|protein_coding|CODING|ENST00000335137|1|2-1)
    GT  1/0 2/1 
</pre>
				What does it mean: 
				<ol>
				   <li> In this case, we have two ALTs = 'C' and 'G'. 
				   <li> Germline sample is heterozygous 'C/A' (GT = '1/0') 
				   <li> Soamtic tissue is heterozygous 'G/C' (GT = '2/1')
				   <li> Change A -&gt; C and A -&gt; G are always calculated by SnpEff (this is the "default mode"). 
						<ul>
						 <li> A -&gt; C produces this effect
						 <code>NON_SYNONYMOUS_START(LOW|MISSENSE|Atg/Ctg|p.Met1Leu|305|OR4F5|protein_coding|CODING|ENST00000335137|1)</code>
						 Note that the last field (genotype field) is '1' indicating this is produced by the first ALT (C)

						 <li> A -&gt; G produces this effect
						 <code>START_LOST(HIGH|MISSENSE|Atg/Gtg|p.Met1Val|305|OR4F5|protein_coding|CODING|ENST00000335137|2)</code>
						 Note that the last field (genotype field) is '2' indicating this is produced by the first ALT (G)
						</ul>

					<li> Finally, this is what you were expecting for, the cancer comparisons. Since both germline and somatic 
						are heterozygous (GT are '1/0' and '2/1'), there are 4 possible comparisons to make:  
						<ul>
							<li> 2 vs 1 : This is the Somatic vs Germline we are interested in. SnpEff reports this one
							<li> 2 vs 0 : This compares ALT to REF, so it was already reported in "default mode". SnpEff doesn't report this one again.
							<li> 1 vs 1 : This is not a variant, since both og them ar '1'. SnpEff skips this one.
							<li> 1 vs 0 : This compares ALT to REF, so it was already reported in "default mode". SnpEff doesn't report this one again.
						</ul>
						I know is confusing, but the bottom line is that only the first comparison one makes sense, and is the one SnpEff reports.
						So 'C -&gt; G' produces the following effect:
						<br>
						<code>'START_LOST(HIGH|MISSENSE|Ctg/Gtg|p.Leu1Val|305|OR4F5|protein_coding|CODING|ENST00000335137|2-1)</code>
						<br>
						<br>
						<i class="icon-warning-sign"></i> Notice the genotype field is "2-1" meaning the we produce a new reference on the fly using ALT 1 ('C') and then used ALT 2 ('G') as the change.
				</ol>
				<br>
				If your brain hasn't exploded yet and want to have some more fun, the other lines produced in the above example are (again, lines edited for the sake of brevity):
<pre>
1   69849   .   G   A,C .   PASS    AF=0.1122
    ;EFF=STOP_GAINED(HIGH|NONSENSE|tgG/tgA|W253*|305|OR4F5|protein_coding|CODING|ENST00000335137|1|1)
        ,NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|tgG/tgC|W253C|305|OR4F5|protein_coding|CODING|ENST00000335137|1|2)
        ,STOP_LOST(HIGH|MISSENSE|tgA/tgC|*253C|305|OR4F5|protein_coding|CODING|ENST00000335137|1|2-1)
    GT  1/0 2/1 


1   69511   .   A   C,G .   PASS    AF=0.3580
    ;EFF=NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Aca/Cca|T141P|305|OR4F5|protein_coding|CODING|ENST00000335137|1|1)
        ,NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Aca/Gca|T141A|305|OR4F5|protein_coding|CODING|ENST00000335137|1|2)
        ,NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Cca/Gca|P141A|305|OR4F5|protein_coding|CODING|ENST00000335137|1|2-1)
    GT  1/1 2/2 
</pre>
				</p>
        </section>
				
		<!-- ENCODE
		================================================== -->
		<section id="ENCODE">
		  <div class="page-header">
			<h1>13. ENCODE </h1>
		  </div>
				<p class="lead">
				ENCODE project's goal is to find all functional elements in the human genome.
				You can perform annotations using ENCODE's data.
				</p>
				<p>
				<a href="http://genome.ucsc.edu/ENCODE/">ENCODE</a> project has produced huge amounts of data (see also <a href="http://www.nature.com/encode">Nature's</a> portal).
				This information is available for download and can be used to annotate genomic variants or regions.
				An overview of all the data available from ENCODE is shown as an <a href="http://genome.ucsc.edu/ENCODE/dataMatrix/encodeDataMatrixHuman.html">experimental data matrix</a>.
				The download site is <a href="http://genome.ucsc.edu/ENCODE/downloads.html">here</a>.
				<br>
				<br>
				Data is available in "BigBed" format, which can be feed into SnpEff using <code>-interval</code> command line option (you can add many <code>-interval</code> options).
				<br>
				Here is a simple example:
				<br>
				<br>
<pre>
# Create a directory for ENCODE files
mkdir -p db/encode

# Download ENCODE experimental results (BigBed file)
cd db/encode
wget "http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/openchrom/jan2011/fdrPeaks/wgEncodeDukeDnase8988T.fdr01peaks.hg19.bb"

# Annotate using ENCODE's data:
java -Xmx4g -jar snpEff.jar eff -v -interval db/encode/wgEncodeDukeDnase8988T.fdr01peaks.hg19.bb GRCh37.71 test.vcf > test.eff.vcf

# Annotations are added as "CUSTOM" intervals:
$ grep CUSTOM test.eff.vcf 
1	799463	.	T	C	.	PASS	AC=958;EFF=CUSTOM[wgEncodeDukeDnase8988T](MODIFIER||||||wgEncodeDukeDnase8988T:799426_799575||||1),DOWNSTREAM(MODIFIER|||||FAM41C|lincRNA|NON_CODING|ENST00000427857||1),DOWNSTREAM(MODIFIER|||||FAM41C|lincRNA|NON_CODING|ENST00000432963||1),DOWNSTREAM(MODIFIER|||||FAM41C|lincRNA|NON_CODING|ENST00000446136||1),DOWNSTREAM(MODIFIER|||||RP11-206L10.11|lincRNA|NON_CODING|ENST00000445118||1),INTERGENIC(MODIFIER||||||||||1)
1	902128	.	C	T	.	PASS	AC=80;EFF=CUSTOM[wgEncodeDukeDnase8988T](MODIFIER||||||wgEncodeDukeDnase8988T:902066_902215||||1),DOWNSTREAM(MODIFIER||||642|KLHL17|protein_coding|CODING|ENST00000338591||1),DOWNSTREAM(MODIFIER|||||KLHL17|nonsense_mediated_decay|CODING|ENST00000466300||1),DOWNSTREAM(MODIFIER|||||KLHL17|retained_intron|CODING|ENST00000463212||1),DOWNSTREAM(MODIFIER|||||KLHL17|retained_intron|CODING|ENST00000481067||1),NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|gCt/gTt|A43V|576|PLEKHN1|protein_coding|CODING|ENST00000379407|2|1),NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|gCt/gTt|A43V|611|PLEKHN1|protein_coding|CODING|ENST00000379410|2|1),NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|gCt/gTt|A43V|663|PLEKHN1|protein_coding|CODING|ENST00000379409|2|1),UPSTREAM(MODIFIER|||||PLEKHN1|retained_intron|CODING|ENST00000480267||1)
1	948921	.	T	C	.	PASS	AC=904;EFF=CUSTOM[wgEncodeDukeDnase8988T](MODIFIER||||||wgEncodeDukeDnase8988T:948786_948935||||1),UPSTREAM(MODIFIER|||||RP11-54O7.11|antisense|NON_CODING|ENST00000458555||1),UTR_5_PRIME(MODIFIER||||165|ISG15|protein_coding|CODING|ENST00000379389|1|1)
...
</pre>


				</p>
        </section>

		<!-- Epigenome
		================================================== -->
		<section id="Epigenome">
		  <div class="page-header">
			<h1>14. Epigenome </h1>
		  </div>
				<p class="lead">
				Epigenome Roadmap Project has produced large amounts of information that can be used by SnpEff.
				</p>
				<p>
				<a href="http://www.roadmapepigenomics.org"> Epigenome Roadmap Project</a> goal is 
				"to map DNA methylation, histone modifications, chromatin accessibility and small RNA transcripts 
				in stem cells and primary ex vivo tissues selected to represent the normal counterparts of tissues 
				and organ systems frequently involved in human disease".
				A <a href="http://www.roadmapepigenomics.org/data"> data matrix </a> shows the experimental set ups 
				currently available.
				<br>
				<br>
				Unfortunately the project is not (currently) providing results files that can be used directly by annotation software, such as SnpEff.
				They will be available later in the project. 
				So, for the time being, data has to be downloaded an pre-processed.
				We'll be processing these information and making it available (as SnpEff databases) as soon as we can.
				<br>
				<br>
				The latest Epigenome project processed information, can be found <a href="http://sourceforge.net/projects/snpeff/files/databases/epigenome_latest.tgz/download">here</a>.
				This includes genomic intervals for high confidence peaks in form of <code>BED</code> files.
				<br>
				<br>
				To annotate you can do:
<pre>
# Download Epigenome project database (pre-processed as BED files)
wget http://sourceforge.net/projects/snpeff/files/databases/epigenome_latest.tgz/download

# Open tar file
tar -xvzf epigenome_latest.tgz

# Annotate using SnpEff and "-interval" command line
java -Xmx4g -jar snpEff.jar eff -v -interval db/epigenome/BI_Pancreatic_Islets_H3K4me3.peaks.bed GRCh37.71 test.vcf > test.eff.vcf 

# See the data represented as "CUSTOM" EFF fields
$ grep CUSTOM test.eff.vcf 
1	894573	.	G	A	.	PASS	AC=725;EFF=CUSTOM[BI_Pancreatic_Islets_H3K4me3](MODIFIER||||||MACS_peak_8||||1),INTRON(MODIFIER||||749|NOC2L|protein_coding|CODING|ENST00000327044|1|1),INTRON(MODIFIER|||||NOC2L|processed_transcript|CODING|ENST00000487214|1|1),INTRON(MODIFIER|||||NOC2L|retained_intron|CODING|ENST00000469563|1|1),UPSTREAM(MODIFIER||||642|KLHL17|protein_coding|CODING|ENST00000338591||1),UPSTREAM(MODIFIER|||||KLHL17|nonsense_mediated_decay|CODING|ENST00000466300||1),UPSTREAM(MODIFIER|||||KLHL17|retained_intron|CODING|ENST00000463212||1),UPSTREAM(MODIFIER|||||KLHL17|retained_intron|CODING|ENST00000481067||1),UPSTREAM(MODIFIER|||||NOC2L|retained_intron|CODING|ENST00000477976||1)
1	948692	.	G	A	.	PASS	AC=896;EFF=CUSTOM[BI_Pancreatic_Islets_H3K4me3](MODIFIER||||||MACS_peak_9||||1),INTERGENIC(MODIFIER||||||||||1),UPSTREAM(MODIFIER||||165|ISG15|protein_coding|CODING|ENST00000379389||1),UPSTREAM(MODIFIER|||||RP11-54O7.11|antisense|NON_CODING|ENST00000458555||1)
1	948921	.	T	C	.	PASS	AC=904;EFF=CUSTOM[BI_Pancreatic_Islets_H3K4me3](MODIFIER||||||MACS_peak_9||||1),UPSTREAM(MODIFIER|||||RP11-54O7.11|antisense|NON_CODING|ENST00000458555||1),UTR_5_PRIME(MODIFIER||||165|ISG15|protein_coding|CODING|ENST00000379389|1|1)
1	1099342	.	A	C	.	PASS	AC=831;EFF=CUSTOM[BI_Pancreatic_Islets_H3K4me3](MODIFIER||||||MACS_peak_10||||1),INTERGENIC(MODIFIER||||||||||1),UPSTREAM(MODIFIER|||||MIR200A|miRNA|NON_CODING|ENST00000384875||1),UPSTREAM(MODIFIER|||||MIR200B|miRNA|NON_CODING|ENST00000384997||1)
</pre>


				</p>
        </section>

		<!-- NextProt
		================================================== -->
		<section id="NextProt">
		  <div class="page-header">
			<h1>15. NextProt </h1>
		  </div>
				<p class="lead">
				NextProt has useful proteomic annotations than can help to identify variants causing reduced protein functionality or even loss of function.
				</p>
				<p>
				<a href="http://www.nextprot.org/">Nextprot</a> project provides proteomic information that can be used for genomic annotations.
				NextProt povides only human data.
				<br>
				<br>
				This data is available by default in GRCh37 databases (latest version).
				You only need to add the <code>-nextprot</code> command line option to get this information.
<pre>
$ java -Xmx4g -jar snpEff.jar -v -nextprot GRCh37.71 test.vcf > test.eff.vcf

# Show results (edited for readibility)
$ cat test.eff.vcf
$ grep glycosylation_site test.eff.vcf 
1  1147655  .  G  A  .  PASS  AC=7;EFF=DOWNSTREAM(MODIFIER||||348|SDF4|protein_coding|CODING|ENST00000263741||1)
                                       ,DOWNSTREAM(MODIFIER|||||SDF4|retained_intron|CODING|ENST00000494748||1)
                                       ,INTRON(MODIFIER|||||TNFRSF4|retained_intron|CODING|ENST00000497869|3|1)
                                       ,NEXT_PROT[glycosylation_site:N-linked__GlcNAc..._](MODERATE||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|4|1)
                                       ,NEXT_PROT[glycosylation_site:N-linked__GlcNAc..._](MODERATE||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|5|1)
                                       ,NEXT_PROT[repeat:TNFR-Cys_4](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|4|1)
                                       ,NEXT_PROT[repeat:TNFR-Cys_4](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|5|1)
                                       ,NEXT_PROT[topological_domain:Extracellular](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|1|1)
                                       ,NEXT_PROT[topological_domain:Extracellular](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|2|1)
                                       ,NEXT_PROT[topological_domain:Extracellular](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|3|1)
                                       ,NEXT_PROT[topological_domain:Extracellular](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|4|1)
                                       ,NEXT_PROT[topological_domain:Extracellular](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|5|1)
                                       ,NEXT_PROT[topological_domain:Extracellular](LOW||||277|TNFRSF4|protein_coding|CODING|ENST00000379236|6|1)
</pre>
				This example shows a <code>glycosylation_site</code> marked as <code>MODERATE</code> impact, since a modification of such a site might impair protein function.
				</p>

        </section>

		<!-- Motif
		================================================== -->
		<section id="Motif">
		  <div class="page-header">
			<h1>16. Motif </h1>
		  </div>
				<p class="lead">
				Motif annotations provided by ENSEMBL and Jaspar can be added to the standard annotations.
				</p>
				<p>
				ENSEMBL provides trancription factor binding sites prediction, for human and mouse genomes, using <a href="http://jaspar.genereg.net/">Jasper</a> motif database.
				<br>
				Using <code>-motif</code> command line option, you can add TFBS prediction to your annotations:
				<br>
				<br>
<pre>
$ java -Xmx4g -jar snpEff.jar -v -motif GRCh37.71 test.vcf > test.eff.vcf
00:00:00.000	Reading configuration file 'snpEff.config'
00:00:00.228	done
00:00:00.228	Reading database for genome version 'GRCh37.71' from file '/Users/pablocingolani//snpEff/data/GRCh37.71/snpEffectPredictor.bin' (this might take a while)
00:00:23.506	done
00:00:23.552	Loading Motifs and PWMs
00:00:23.552		Loading PWMs from : /Users/pablocingolani//snpEff/data/GRCh37.71/pwms.bin
00:00:23.699		Loading Motifs from file '/Users/pablocingolani//snpEff/data/GRCh37.71/motif.bin'
00:00:24.811		Motif database: 393522 markers loaded.
...

$ grep MOTIF test.eff.vcf 
1	1060235	.	G	A	.	PASS	AC=16;EFF=MOTIF[PB0134.1:HNF4A](MODIFIER||||||||||1),INTERGENIC(MODIFIER||||||||||1)
1	1250957	.	G	A	.	PASS	AC=9;EFF=MOTIF[MA0058.1:Max](LOW||||||||||1),MOTIF[MA0093.1:USF1](LOW||||||||||1),MOTIF[PB0043.1:Max](MODIFIER||||||||||1),MOTIF[PL0007.1:Max](MODIFIER||||||||||1),MOTIF[PL0014.1:mxl-1](MODIFIER||||||||||1),START_GAINED(LOW||||342|CPSF3L|protein_coding|CODING|ENST00000421495|2|1),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H128|571|CPSF3L|protein_coding|CODING|ENST00000545578|6|1),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H157|600|CPSF3L|protein_coding|CODING|ENST00000435064|5|1),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H163|208|CPSF3L|protein_coding|CODING|ENST00000527719|6|1|WARNING_TRANSCRIPT_INCOMPLETE),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H163|606|CPSF3L|protein_coding|CODING|ENST00000540437|7|1),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H204|248|CPSF3L|protein_coding|CODING|ENST00000530031|6|1|WARNING_TRANSCRIPT_INCOMPLETE),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H33|108|CPSF3L|protein_coding|CODING|ENST00000526332|3|1),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H56|499|CPSF3L|protein_coding|CODING|ENST00000419704|3|1),SYNONYMOUS_CODING(LOW|SILENT|caC/caT|H59|502|CPSF3L|protein_coding|CODING|ENST00000411962|3|1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000461514||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000470030||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000478641||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000485710||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000497304||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000525164||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000525769||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000531020||1),UPSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000531292||1),UTR_3_PRIME(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000430786|7|1),UTR_3_PRIME(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000458452|7|1),UTR_3_PRIME(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000526797|5|1),UTR_3_PRIME(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000527098|6|1),DOWNSTREAM(MODIFIER||||124|PUSL1|protein_coding|CODING|ENST00000467712||1),DOWNSTREAM(MODIFIER||||139|CPSF3L|protein_coding|CODING|ENST00000534345||1),DOWNSTREAM(MODIFIER||||166|CPSF3L|protein_coding|CODING|ENST00000498476||1),DOWNSTREAM(MODIFIER||||303|PUSL1|protein_coding|CODING|ENST00000379031||1),DOWNSTREAM(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000470679||1),DOWNSTREAM(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000525285||1),DOWNSTREAM(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000531019||1),DOWNSTREAM(MODIFIER|||||CPSF3L|processed_transcript|CODING|ENST00000490853||1),DOWNSTREAM(MODIFIER|||||CPSF3L|processed_transcript|CODING|ENST00000493534||1),DOWNSTREAM(MODIFIER|||||CPSF3L|processed_transcript|CODING|ENST00000532952||1),DOWNSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000429572||1),DOWNSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000496353||1),DOWNSTREAM(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000498173||1),DOWNSTREAM(MODIFIER|||||PUSL1|processed_transcript|CODING|ENST00000470520||1),DOWNSTREAM(MODIFIER|||||PUSL1|retained_intron|CODING|ENST00000493657||1),DOWNSTREAM(MODIFIER|||||RP5-890O3.9|processed_transcript|NON_CODING|ENST00000444968||1),EXON(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000527383|1|1),EXON(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000528879|5|1),EXON(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000531377|2|1),EXON(MODIFIER|||||CPSF3L|processed_transcript|CODING|ENST00000462432|2|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000323275|4|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000467408|1|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000488042|5|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000525603|3|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000526113|2|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000526904|2|1),EXON(MODIFIER|||||CPSF3L|retained_intron|CODING|ENST00000533916|1|1),INTRON(MODIFIER||||343|CPSF3L|protein_coding|CODING|ENST00000434694|4|1),INTRON(MODIFIER||||578|CPSF3L|protein_coding|CODING|ENST00000450926|4|1),INTRON(MODIFIER|||||CPSF3L|nonsense_mediated_decay|CODING|ENST00000532772|2|1)
1	1310924	.	T	C	.	PASS	AC=948;EFF=MOTIF[PB0020.1:Gabp](MODIFIER||||||||||1),UPSTREAM(MODIFIER||||199|AURKAIP1|protein_coding|CODING|ENST00000321751||1),UPSTREAM(MODIFIER||||199|AURKAIP1|protein_coding|CODING|ENST00000338338||1),UPSTREAM(MODIFIER||||199|AURKAIP1|protein_coding|CODING|ENST00000338370||1),UPSTREAM(MODIFIER||||199|AURKAIP1|protein_coding|CODING|ENST00000378853||1),UPSTREAM(MODIFIER|||||AURKAIP1|processed_transcript|CODING|ENST00000489799||1),UPSTREAM(MODIFIER|||||AURKAIP1|processed_transcript|CODING|ENST00000496905||1),UPSTREAM(MODIFIER|||||RP5-890O3.3|processed_pseudogene|NON_CODING|ENST00000435351||1),INTERGENIC(MODIFIER||||||||||1)
1	1368599	.	A	C	.	PASS	AC=920;EFF=MOTIF[MA0065.1:PPARG](MODIFIER||||||||||1),MOTIF[MA0065.2:PPARG](MODIFIER||||||||||1),MOTIF[MA0114.1:HNF4A](MODIFIER||||||||||1),MOTIF[MA0115.1:RNR1H2](MODIFIER||||||||||1),UPSTREAM(MODIFIER||||147|VWA1|protein_coding|CODING|ENST00000471398||1),UPSTREAM(MODIFIER||||166|VWA1|protein_coding|CODING|ENST00000495558||1),UPSTREAM(MODIFIER||||233|VWA1|protein_coding|CODING|ENST00000404702||1),UPSTREAM(MODIFIER||||445|VWA1|protein_coding|CODING|ENST00000476993||1),UPSTREAM(MODIFIER||||64|VWA1|protein_coding|CODING|ENST00000338660||1),UPSTREAM(MODIFIER|||||RP4-758J18.10|lincRNA|NON_CODING|ENST00000434150||1),INTRON(MODIFIER|||||RP4-758J18.10|lincRNA|NON_CODING|ENST00000417917|1|1),INTRON(MODIFIER|||||RP4-758J18.10|lincRNA|NON_CODING|ENST00000430109|1|1),INTRON(MODIFIER|||||RP4-758J18.10|lincRNA|NON_CODING|ENST00000454562|1|1)
1	2182470	.	G	A	.	PASS	AC=743;EFF=MOTIF[PB0134.1:HNF4A](MODIFIER||||||||||1),INTRON(MODIFIER||||728|SKI|protein_coding|CODING|ENST00000378536|1|1),INTRON(MODIFIER|||||SKI|processed_transcript|CODING|ENST00000478223|1|1),INTRON(MODIFIER|||||SKI|processed_transcript|CODING|ENST00000508416|1|1)
1	2466633	.	A	G	.	PASS	AC=770;EFF=MOTIF[PB0191.1:Ap2gamma](MODIFIER||||||||||1),UPSTREAM(MODIFIER||||166|HES5|protein_coding|CODING|ENST00000378453||1),INTERGENIC(MODIFIER||||||||||1)
1	2480337	.	G	A	.	PASS	AC=16;EFF=MOTIF[MA0058.1:Max](MODIFIER||||||||||1),MOTIF[MA0059.1:Cmyc](MODIFIER||||||||||1),MOTIF[MA0093.1:USF1](MODIFIER||||||||||1),MOTIF[MA0147.1:Cmyc](LOW||||||||||1),MOTIF[PB0043.1:Max](MODIFIER||||||||||1),MOTIF[PL0007.1:Max](MODIFIER||||||||||1),MOTIF[PL0014.1:mxl-1](MODIFIER||||||||||1),DOWNSTREAM(MODIFIER|||||RP3-395M20.8|processed_transcript|NON_CODING|ENST00000416860||1),DOWNSTREAM(MODIFIER|||||RP3-395M20.8|processed_transcript|NON_CODING|ENST00000432521||1),DOWNSTREAM(MODIFIER|||||RP3-395M20.8|processed_transcript|NON_CODING|ENST00000443892||1),DOWNSTREAM(MODIFIER|||||RP3-395M20.8|processed_transcript|NON_CODING|ENST00000448624||1),DOWNSTREAM(MODIFIER|||||RP3-395M20.8|processed_transcript|NON_CODING|ENST00000449660||1),INTERGENIC(MODIFIER||||||||||1)

</pre>
				<br>
				So, for instance, the first MOTIF annotation is <code>MOTIF[PB0134.1:HNF4A]</code> corresponding to motif <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0134.1&rm=present&collection=CORE">PB0134.1</a>, which you can look up in Jaspar.
				</p>
        </section>

        <!-- Databases
        ================================================== -->
        <section id="databases">
          <div class="page-header">
            <h1>17. Building databases </h1>
          </div>
				<p class="lead">
				SnpEff needs a database to perform genomic annotations. 
				There are pre-built databases for over 2,500 genomes, so chances are that your organism of choice already has a SnpEff database available.
				In the (unlikely?) event that you need to build one yourself, here we describe how to it.
				</p>

				<p>
				<i class="icon-hand-right"></i>You can know which genomes are supported by running the following command:
<pre>
java -jar snpEff.jar databases
</pre>
				<br>
				<i class="icon-warning-sign"></i> Most people do NOT need to build a database, and can safely use a pre-built one. 
				So unless you are working with an rare genome you most likely don't need to do it either.
				<br>
				<br>
				<i class="icon-warning-sign"></i> Again, chances are that you DO NOT NEED to build a database.
				Sorry to repeat this, but I cannot tell how many times I get emails asking for help to build database that is already available.
				<br>
				<br>

				<h3>Building a database</h3>
				In order to build a database for a new genome, you need to:
				<br>
				<br>
				<ol>
					<li> Configure a new genome in SnpEff's config file <code>snpEff.config</code>.
						<ol>
							<li> <a href="#buildAddConfig">Add genome entry</a> to snpEff's configuration
							<li> If the genome uses a non-standard codon table: <a href="#buildAddConfigCodonTable">Add codon table parameter</a>
						</ol>
					<li> Get the reference genome sequence (e.g. in FASTA format).
					<li> Get genome annotations. There are four different ways you can do this:
						<ol>
							<li> <a href="#buildGtf">Option 1:</a> Building a database from GTF files (the easiest way)
							<li> <a href="#buildGff">Option 2:</a> Building a database from GFF files
							<li> <a href="#buildRefSeq">Option 3:</a> Building a database from RefSeq table from UCSC
							<li> <a href="#buildGenBank">Option 4:</a> Building a database from GenBank files
						</ol>
					<li> Run a command to create the database (i.e. "java -jar snpEff.jar build ...")
				</ol>

				<b>Note:</b> All files can be compressed using gzip. E.g. the reference file 'hg19.fa' can be compressed to 'hg19.fa.gz', snpEff will automatically decompress the file.
				<br>
				<br>
				<i class="icon-warning-sign"></i><b>Warning:</b> Some files claimed to be compressed using GZIP are actually not or even use a block compression variant not supported by Java's gzip library.
				If you notice that your build process finishes abruptly for no apparent reason, try uncompressing the files. 
				This sometimes happens with ENSEMBL files.
				<br>
				<br>
				</p>

				<h3>Configuring a new genome</h3>
				<p>
				In order to tell SnpEff that there is a new genome available, you must update SnpEff's configuration file <code>snpEff.config</code>.
				<br>
				You must add a new genome entry to <code>snpEff.config</code>.
				<br>
				If your genome, or a chromosome, uses non-standard codon tables you must update <code>snpEff.config</code> accordingly.
				A typical case is when you use mitochondrial DNA. Then you specify that chromosome 'MT' uses <code>codon.Invertebrate_Mitochondrial</code> codon table. 
				Another common case is when you are adding a bacterial genome, then you specify that the codon table is <code>Bacterial_and_Plant_Plastid</code>.
				<br>
				<br>
				</p>

				<a name="buildAddConfig"> </a>
				<h4>Add a genome to the configuration file</h4>
				<p>
				This example shows how to add a new genome to the config files. For this example we'll use the mouse genome (mm37.61):
				<br>
				<br>
				<ol>
					<li> Edit the config file to create the new genome

						<pre> vi snpEffect.config </pre>
							
						Add the following lines (you are editing snpEffect.config)<br>
<pre>
# Mouse genome, version mm37.61
mm37.61.genome : Mouse 
</pre>
				<br>
				<i class="icon-warning-sign"></i><b>Warning:</b> You may need to add codon table information for the genome or some parts of it (e.g. mitochondrial "chromosome"). 
				See next section for details.
				<br>
				<br>

					<li> Optional: Add genome to Galaxy's menu 
<pre>
cd /path/to/galaxy
cd tools/snpEffect/
vi snpEffect.xml 
</pre>

						Add the following lines to the file
<pre>
	&lt;param name="genomeVersion" type="select" label="Genome"&gt;
		&lt;option value="hg37"&gt;Human (hg37)&lt;option&gt;
		&lt;option value="mm37.61"&gt;Mouse (mm37.61)&lt;option&gt;
	&lt;param&gt; 
</pre>
				</ol>
				<br>
				<br>
				</p>

				<a name="buildAddConfigCodonTable"> </a>
				<h4>Configuring codon tables (not always required)</h4>
				<p>
				Codon tables are provided in the <code>snpEff.config</code> configuration file under the section <code>codon.Name_of_your_codon_table</code>.
				The format is a comma separated list of <code>CODON/AMINO_ACID</code>. 
				<br>
				E.g.: 
<pre>
codon.Invertebrate_Mitochondrial:  TTT/F, TTC/F, TAC/Y, TAA/*, ATG/M+, ATG/M+, ACT/T, ...
</pre>
				Note that codons marked with '*' are STOP codons and codons marked with a '+' are START codons.
				<br>
				<br>
				In order for you to use them, you have to specify that a given "chromosome" uses one of the tables (otherwise the default codon table is used).
				<br>
				E.g. Here we say the chromosome 'M' from fly genome (dm3) uses Invertebrate_Mitochondrial codon table:
<pre>
dm3.M.codonTable : Invertebrate_Mitochondrial
</pre>
				...of course, chromosome 'M' is not a real chromosome, it is just a way to mark the sequence as mitochondrial DNA in the reference genome.
				<br>
				<br>
				</p>

				<h3>Reference genome: GTF, GFF, RefSeq or GenBank</h3>
				<p>
				As we previously mentioned, reference genome information can be in different formats: GTF, GFF, RefSeq or GenBank.
				<br>
				In the following sub-sections, we show how to build a database for each type of genomic information file.
				<br>
				<br>
				</p>

				<a name="buildGtf"></a>
				<h4>Option 1: Building a database from GTF files</h4>
				<p>
				GTF 2.2 files are supported by SnpEff (e.g. ENSEMBL releases genome annotations in this format).
				<ol>
					<li> Get the genome and uncompress it
<pre>
# Create directoy for this new genome
cd /path/to/snpEff/data/
mkdir mm37.61
cd mm37.61

# Get annotation files
wget ftp://ftp.ensembl.org/pub/current/gtf/mus_musculus/Mus_musculus.NCBIM37.61.gtf.gz
mv Mus_musculus.NCBIM37.61.gtf.gz genes.gtf.gz

# Get the genome
cd /path/to/snpEff/data/genomes
wget ftp://ftp.ensembl.org/pub/current/fasta/mus_musculus/dna/Mus_musculus.NCBIM37.61.dna.toplevel.fa.gz
mv Mus_musculus.NCBIM37.61.dna.toplevel.fa.gz mm37.61.fa.gz
</pre>

					<li> <b>Note:</b> The FASTA file can be either in <br> <tt>/path/to/snpEff/data/genomes/mm37.61.fa</tt> <br> or in <br> <tt>/path/to/snpEff/data/mm37.61/sequences.fa</tt><p>

					<li> Add the new genome to the config file (see section "How to add a new genome to the configuration file" for details)

					<li> Create database
<pre>
cd /path/to/snpEff
java -jar snpEff.jar build -gtf22 -v mm37.61 
</pre>
				</ol>
				<br>
				<br>
				</p>

				<a name="buildGff"> </a>
				<h4>Option 2: Building a database from GFF files </h4>
				<br>
				<i class="icon-warning-sign"></i><b>Warning:</b> Using GFF is discouraged, we recommend you use GTF files instead (whenever possible).
				<br>
				<br>
				<p>
				This example shows how to create a database for a new genome using GFF file ((e.g. FlyBase, WormBase, BeeBase release GFF files). 
				For this example we'll use the Drosophila melanogaster genome (dm5.31):
				<ol>
					<li> Get a GFF file (into path/to/snpEff/data/dm5.31/genes.gff):

<pre>
mkdir path/to/snpEff/data/dm5.31
cd path/to/snpEff/data/dm5.31
wget ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r5.31_FB2010_08/gff/dmel-all-r5.31.gff.gz
mv dmel-all-r5.31.gff.gz genes.gff.gz 
</pre>

					<li> <b>Note:</b> GFF3 files can include the reference sequence in the same file. This is done by dumping the fasta file after a '##FASTA' line. You can also add the sequence fasta file to the 'data/genomes/' directory, like it is done in when using GTF format.

					<li> Add the new genome to the config file (see section "How to add a new genome to the configuration file" for details)

					<li> Create database (note the "-gff3" flag):
<pre>
cd /path/to/snpEff
java -jar snpEff.jar build -gff3 -v dm5.31 
</pre>
				</ol>
				<br>
				<br>
				</p>

				<a name="buildRefSeq"> </a>
				<h4>Option 3: Building a database from RefSeq table from UCSC</h4>
				<p>
				This example shows how to create a database for a new genome. 
				For this example we'll use the Human genome (hg19).
				<br>
				<br>
				<i class="icon-warning-sign"></i><b>Warning:</b> Using UCSC genome tables is highly discouraged, we recommend you use ENSEMBL versions instead.
				<br>
				<br>
				<i class="icon-warning-sign"></i><b>Warning:</b> UCSC tables sometimes change for different species. 
				This means that even if these instructions work for human genome, it might not work for other genomes.
				Obviously creating a new parser for each genome is impractical, so working with UCSC genomes is highly discouraged.
				We recommend to use ENSEMBL genomes instead.
				<br>
				<br>
				<i class="icon-warning-sign"></i><b>Warning:</b> UCSC genomes provide only major release version, but NOT sub-versions.
				E.g. UCSC's "hg19" has major version 19 but there is no "sub-version", whereas ENSEMBL's GRCh37.70 clearly has major version 37 and minor version 70. 
				Not providing a minor version means that they might change the database and two "hg19" genomes are actually be different.
				This creates all sorts of consistency problems (e.g. the annotations may not be the same that you see in the UCSC genome browser, even though both of them are 'hg19' version).
				Using UCSC genome tables is highly discouraged, we recommend you use ENSEMBL versions instead.
				<br>
				<br>
				In order to build a genome using UCSC tables, you can follow these instructions:
				<ol>
					<li> Go to UCSC genome browser (http://genome.ucsc.edu/)
					<li> Click on "Table" menu
					<li> Select parameters as shown here:
							<br>
							<img src="./images/ucsc_refSeq.png">
							<br>
					<li> Click on "get output" and save the data to file "/path/to/snpEff/data/hg19/genes.txt".

					<li> Add the fasta reference genome. The FASTA file can be either in <br> <tt>/path/to/snpEff/data/genomes/hg19.fa</tt> <br> or in <br> <tt>/path/to/snpEff/data/hg19/sequences.fa</tt><p>

					<li> <a class="body" href="#conf">Add</a> the new genome to the config file (see section "How to add a new genome to the configuration file" for details)

					<li> Create database (note the "-refSeq" flag):
<pre>
cd /path/to/snpEff
java -jar snpEff.jar build -refSeq -v hg19 
</pre>
				</ol>
				<br>
				<br>
				</p>

				<a name="buildGenBank"> </a>
				<h4>Option 4: Building a database from GenBank files</h4>
				<p>
				This example shows how to create a database for a new genome. 
				For this example we'll use "Staphylococcus aureus":
				<ol>
					<li> Go to NIH page for <a class="body" href="http://www.ncbi.nlm.nih.gov/nuccore/CP000730.1">CP000730</a>

					<li> Download the features in geneBank format, by clicking as shown in the following images (red arrows)
						<br>
						<br>
						<center> <img width=800 src="images/genBank_0.png"> </center> 
						<br>
						<br>
						Make sure you click the "Update" button!
						<br>
						Then you go to the "Send" menu
						<br>
						<br>
						<center> <img width=800 src="images/genBank_1.png"> </center>
						<br>
						<br>
						and then
						<br>
						<br>
						<center> <img width=800 src="images/genBank_2.png"> </center>
						<br>
						<br>

					<li> Save the GenBank data to "/path/to/snpEff/data/CP000730/genes.gb".
						<br>
						<br>
						 <b>Note:</b> If there are more than one genbank file for an organism (e.g. multiple chromosomes), then you can download each file and  concatenate them.
						<br>
						 E.g.: Vibrio Cholerae has two chromosomes with GenBank accessions: NC_002505.1 and NC_002506.1.
						 You can download both files and save them as snpEff/data/vibrio/NC_002505.1.gb and snpEff/data/vibrio/NC_002506.1.gb respectively, and then concatenate both files:
<pre>
cat NC_002505.1.gb NC_002506.1.gb > genes.gb
</pre>
						 Add the following entries in the config file:
						<br>
						<br>
<pre>
# Vibrio Cholerae
vibrio.genome : Vibrio Cholerae
	vibrio.chromosomes : NC_002505.1, NC_002506.1
	vibrio.NC_002505.1.codonTable : Bacterial_and_Plant_Plastid
	vibrio.NC_002506.1.codonTable : Bacterial_and_Plant_Plastid
</pre>


					<li> Create database (note the "-genbank" flag):
						<br>
						<br>
<pre>
	cd /path/to/snpEff
	java -jar snpEff.jar build -genbank -v CP000730 
</pre>
				</ol>
				</pre>
				<br>
				<br>
				</p>

				<h3>Example: Building the Human Genome database</h3>
				<p>
				This is a full example on how to build the human genome database (using GTF file from ENSEBML), it includes support for regulatory features, sanity check, rare amino acids, etc..
				<br>
				<br>
<pre>
# Go to SnpEff's install dir
cd ~/snpeff

# Create database dir
mkdir data/GRCh37.70
cd data/GRCh37.70

# Download annotated genes
wget ftp://ftp.ensembl.org/pub/release-70/gtf/homo_sapiens/Homo_sapiens.GRCh37.70.gtf.gz
mv Homo_sapiens.GRCh37.70.gtf.gz genes.gtf.gz

# Download proteins 
# This is used for:
#	- "Rare Amino Acid" annotations
#	- Sanity check (checking protein predicted from DNA sequences match 'real' proteins)
wget ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/pep/Homo_sapiens.GRCh37.70.pep.all.fa.gz
mv Homo_sapiens.GRCh37.70.pep.all.fa.gz protein.fa.gz

# Download CDSs
# Note: This is used as "sanity check" (checking that CDSs prediscted from gene sequences match 'real' CDSs)
wget ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh37.70.cdna.all.fa.gz
mv Homo_sapiens.GRCh37.70.cdna.all.fa.gz cds.fa.gz

# Download regulatory annotations
wget ftp://ftp.ensembl.org/pub/release-70/regulation/homo_sapiens/AnnotatedFeatures.gff.gz
mv AnnotatedFeatures.gff.gz regulation.gff.gz

# Uncompress
gunzip *.gz

# Download genome
cd ../genomes/
wget ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.70.dna.toplevel.fa.gz
mv Homo_sapiens.GRCh37.70.dna.toplevel.fa.gz GRCh37.70.fa.gz

# Uncompress: 
# Why do we need to uncompress? 
# Because ENSEMBL compresses files using a block compress gzip which is not compatibles whith Java's library Gunzip 
gunzip GRCh37.70.fa.gz

# Edit snpEff.config file
#
# WARNING! You must do this yourself. Just copying and pasting this into a terminal won't work.
#
# Add lines:
#		GRCh37.70.genome : Homo_sapiens
#		GRCh37.70.reference : ftp://ftp.ensembl.org/pub/release-70/gtf/

# Now we are ready to build tha database
cd ~/snpeff
java -Xmx20g -jar snpEff.jar build -v GRCh37.70 2>&1 | tee GRCh37.70.build
</pre>

				<br>
				<br>
				</p>
				<h3> Troubleshooting Database builds </h3>
				<i class="icon-warning-sign"></i><b>Warning:</b> By far the most common problem is that the FASTA file chromosome names are different than the GFF chromosome names.
				Make sure chromosome names are consistent in all the files you use.
				<br>
				<br>
				<p> 
				<b> When I build the database using GFF 3 SnpEff reports that Exons don't have sequences</b>
				<br>
				<br>
				GFF3 files can have sequence information either in the same file or in a separate fasta file.
				<br>
				In order to add sequence information in the GFF file, you can do this:
<pre>
cat annotations.gff > genes.gff
echo "###"  >> genes.gff
echo "##FASTA"  >> genes.gff
cat sequence.fa  >> genes.gff
</pre>
				</p>

				<p> <b> When building a database, I get zero protein coding genes </b>
				<br>
				<br>
				When building a database, snpEff tries to find which transcripts are protein coding. This is done using the 'bioType' information.
				<br>
				The bioType information is not a standard GFF or GTF feature. So I follow ENSEMBL's convention of using the second column ('source') for bioType, as well as the gene_biotype attribute.
				<br>
				If your file was not produced by ENSEMBL, it probably doesn't have this information. This means that snpEff doesn't know which genes are protein coding and which ones are not. 
				<br>
				Having no information, snpEff will treat all genes as protein coding (assuming you have '-treatAllAsProteinCoding Auto' option in the command line, which is the default).
				<br>
				So you will get effects as if all genes were protein coding, then you can filter out the irrelevant genes. Unfortunately, this is the best I can do if there is no 'bioType' information 
				</p>

				<p> <b> When building a database, I get too many warnings </b>
				<br>
				<br>
				There are plenty of GFF and GTF files that, unfortunately, do not follow the specification. 
				SnpEff usually complains about this, but tries hard to correct the problems. 
				So the database may be OK even after you see many warnings.
				<br>
				You can check the database to see if the features (genes, exons, UTRs) have been correctly incorporated, by taking a look at the database:
<pre>
java -jar snpEff.jar dump myGenome | less
</pre>
				</p>
        </section>

        <!-- Databases Non-coding
        ================================================== -->
        <section id="databasesNc">
          <div class="page-header">
            <h1>18. Building databases: Regulatory and Non-coding </h1>
          </div>
				<p class="lead">
				SnpEff supports regulatory and non-coding annotations.
				In this sections we show how to build those databases.
				As in the previous section, most likely you will never have to do it yourself and can just use available pre-built databases.
				</p>

				<p>
				There are two ways to add support for regulatory annotations (these are not mutually exclusive, you can use both at the same time): 
				<ol>
					<li> GFF regulation file (from ENSEMBL).
					<li> BED files.
				</ol>

				<i class="icon-warning-sign"></i> Adding regulation support and analyzing data using regulation tracks can take much more memory. For instance, for the human genome I use 10Gb to 20Gb of RAM.
				<br>
				<br>
				<i class="icon-warning-sign"></i> It is assumed the the genome is already installed, only regulatory tracks are added.

				<h3> <h3>Option 1: Using a GFF file </h3>

				This example shows how to create a regulation database for human (GRCh37.65):

				<ol>
					<li> Get the GFF regulatory annotations (into path/to/snpEff/data/GRCh37.65/regulation.gff):

<pre>
cd path/to/snpEff/data/GRCh37.65
wget ftp:/ftp.ensembl.org/pub/release-65/regulation/homo_sapiens/AnnotatedFeatures.gff.gz
mv AnnotatedFeatures.gff.gz regulation.gff.gz 
</pre>

					<li> Create databases. Note that we use "-onlyReg" flag, because we are only creating regulatory databases. If you omit it, it will create both of "normal' and regulatory databases:
<pre>
cd /path/to/snpEff
java -Xmx20G -jar snpEff.jar build -v -onlyReg GRCh37.65 
</pre>

						The output looks like this
<pre>
Reading regulation elements (GFF)
	Chromosome '11'	line: 226964
	Chromosome '12'	line: 493780
	...
	Chromosome '9'	line: 4832434
	Chromosome 'X'  line: 5054301
	Chromosome 'Y'  line: 5166958
Done
	Total lines                 : 5176289
	Total annotation count      : 3961432
	Percent                     : 76.5%
	Total annotated length      : 3648200193
	Number of cell/annotations  : 266
Saving database 'HeLa-S3' in file '/path/to/snpEff/data/GRCh37.65/regulation_HeLa-S3.bin'
Saving database 'HepG2' in file '/path/to/snpEff/data/GRCh37.65/regulation_HepG2.bin'
Saving database 'NHEK' in file '/path/to/snpEff/data/GRCh37.65/regulation_NHEK.bin'
Saving database 'GM12878' in file '/path/to/snpEff/data/GRCh37.65/regulation_GM12878.bin'
Saving database 'HUVEC' in file '/path/to/snpEff/data/GRCh37.65/regulation_HUVEC.bin'
Saving database 'H1ESC' in file '/path/to/snpEff/data/GRCh37.65/regulation_H1ESC.bin'
Saving database 'CD4' in file '/path/to/snpEff/data/GRCh37.65/regulation_CD4.bin'
Saving database 'GM06990' in file '/path/to/snpEff/data/GRCh37.65/regulation_GM06990.bin'
Saving database 'IMR90' in file '/path/to/snpEff/data/GRCh37.65/regulation_IMR90.bin'
Saving database 'K562' in file '/path/to/snpEff/data/GRCh37.65/regulation_K562.bin'
Done.
</pre>

						As you can see, annotations for each cell type are saved in different files. This makes it easier to load annotations only for the desired cell types when analyzing data.
				</ol>
				</p>

				<h3>Option 2: Using an BED filed </h3>
				<p>
				This example shows how to create a regulation database for human (GRCh37.65). 
				We assume we have a file called "my_regulation.bed" which has information for H3K9me3 in Pancreatic Islets (for instance, as a result of a Chip-Seq experiment and peak enrichment analysis).

				<ol>
					<li> Add all your BED files to 'path/to/snpEff/data/GRCh37.65/regulation.bed/' dir:

<pre>
cd path/to/snpEff/data/GRCh37.65
mkdir regulation.bed
cd regulation.bed
mv where/everh/your/bed/file/is/my_regulation.bed ./regulation.Pancreatic_Islets.H3K9me3.bed 
</pre>

						<b>Note:</b> The name of the file must be 'regulation.CELL_TYPE.ANNOTATION_TYPE.bed'. In this case, 'CELL_TYPE=Pancreatic_Islets' and 'ANNOTATION_TYPE=H3K9me3' <br>

					<li> Create databases (note the "-onlyReg" flag):
<pre>
cd /path/to/snpEff
java -Xmx20G -jar snpEff.jar build -v -onlyReg GRCh37.65 
</pre>

						The output looks like this
<pre>
Building database for 'GRCh37.65'
Reading regulation elements (GFF)
Cannot read regulation elements form file '/path/to/snpEff/data/GRCh37.65/regulation.gff'
Directory has 1 bed files and 1 cell types
Creating consensus for cellType 'Pancreatic_Islets', files: [/path/to/snpEff/data/GRCh37.65/regulation.bed/regulation.Pancreatic_Islets.H3K9me3.bed]
Reading file '/path/to/snpEff/data/GRCh37.65/regulation.bed/regulation.Pancreatic_Islets.H3K9me3.bed'
	Chromosome '10'	line: 5143
	Chromosome '11'	line: 8521
	...
	Chromosome 'X'	line: 52481
	Chromosome 'Y'	line: 53340
Done
	Total lines                 : 53551
	Total annotation count      : 53573
	Percent                     : 100.0%
	Total annotated length      : 75489402
	Number of cell/annotations  : 1
Creating consensus for cell type: Pancreatic_Islets
Sorting: Pancreatic_Islets	, size: 53573
Adding to final consensus
Final consensus for cell type: Pancreatic_Islets	, size: 53549
Saving database 'Pancreatic_Islets' in file '/path/to/snpEff/data/GRCh37.65/regulation_Pancreatic_Islets.bin'
Done
Finishing up
</pre>

						<b>Note:</b> If there are many annotations, they are saved in one binary file for each cell type (i.e. several BED files for different cell types are collapsed together). 
						This makes it easier to load annotations only for the desired cell types when analyzing data.
				</ol>
				</p>
        </section>


		<!-- Integration
		================================================== -->
		<section id="gatk">
		  <div class="page-header">
			<h1>19. Integration: GATK </h1>
		  </div>
				<p class="lead">
				SnpEff is integrated with Broad Institute's Genome Analysis Toolkit (<a href="http://www.broadinstitute.org/gatk/">GATK</a>).
				</p>
				In order to make sure SnpEff and GATK understand each other, you must activate GATK compatibility in SnpEff by using the <code>-o gatk</code> command line option.
				The reason for using '-o gatk' is that, even though both GATK and SnpEff use VCF format, SnpEff has recently updated the <code>EFF</code> sub-field format and this might cause some trouble (since GATK still uses the original version).
				<br>
				<br>
				<i class="icon-warning-sign"></i> GATK only picks one effect.
				Indeed,  the GATK team decided to only report the effect having the highest impact. 
				This was done intentionally for the sake of brevity, in a 'less is more' spirit.
				You can get the full effect by using snpEff independently, instead of using it within GATK framework.
				<br>
				<br>
				Script example: In this example we combine SnpEff and GATK's VariantAnnotator (you can find this script in <code>snpEff/scripts/</code> directory of the distribution)
<pre>
#!/bin/sh

#-------------------------------------------------------------------------------
# Files
#-------------------------------------------------------------------------------

in=$1                                                   # Input VCF file
eff=`dirname $in`/`basename $in .vcf`.snpeff.vcf        # SnpEff annotated VCF file
out=`dirname $in`/`basename $in .vcf`.gatk.vcf          # Output VCF file (annotated by GATK)

ref=$HOME/snpEff/data/genomes/hg19.fa                   # Reference genome file
dict=`dirname $ref`/`basename $ref .fa`.dict            # Reference genome: Dictionary file

#-------------------------------------------------------------------------------
# Path to programs and libraries
#-------------------------------------------------------------------------------

gatk=$HOME/tools/gatk/GenomeAnalysisTK.jar
picard=$HOME/tools/picard/
snpeff=$HOME/snpEff/snpEff.jar

#-------------------------------------------------------------------------------
# Main
#-------------------------------------------------------------------------------

# Create genome index file
echo
echo "Indexing Genome reference FASTA file: $ref"
samtools faidx $ref

# Create dictionary
echo
echo "Creating Genome reference dictionary file: $dict"
java -jar $picard/CreateSequenceDictionary.jar R= $ref O= $dict

# Annotate
echo
echo "Annotate using SnpEff"
echo "    Input file  : $in"
echo "    Output file : $eff"
java -Xmx4G -jar $snpeff -c $HOME/snpEff/snpEff.config -v -o gatk hg19 $in > $eff

# Use GATK
echo
echo "Annotating using GATK's VariantAnnotator:"
echo "    Input file  : $in"
echo "    Output file : $out"
java -Xmx4g -jar $gatk \
    -T VariantAnnotator \
    -R $ref \
    -A SnpEff \
    --variant $in \
    --snpEffFile $eff \
    -L $in \
    -o $out
</pre>

				<br>
				<i class="icon-warning-sign"></i> <b>Important:</b> In order for this to work, GATK requires that the Genome Reference file should have the chromosomes in karotyping order (largest to smallest chromosomes, followed by the X, Y, and MT). Your VCF file should also respect that order.
				<br>
				<br>

				Now we can use the script:
<pre>
$ ~/snpEff/scripts/gatk.sh zzz.vcf

Indexing Genome reference FASTA file: /home/pcingola/snpEff/data/genomes/hg19.fa

Creating Genome reference dictionary file: /home/pcingola/snpEff/data/genomes/hg19.dict
[Fri Apr 12 11:23:12 EDT 2013] net.sf.picard.sam.CreateSequenceDictionary REFERENCE=/home/pcingola/snpEff/data/genomes/hg19.fa OUTPUT=/home/pcingola/snpEff/data/genomes/hg19.dict    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false
[Fri Apr 12 11:23:12 EDT 2013] Executing as pcingola@localhost.localdomain on Linux 3.6.11-4.fc16.x86_64 amd64; OpenJDK 64-Bit Server VM 1.6.0_24-b24; Picard version: 1.89(1408)
[Fri Apr 12 11:23:12 EDT 2013] net.sf.picard.sam.CreateSequenceDictionary done. Elapsed time: 0.00 minutes.
Runtime.totalMemory()=141164544
To get help, see http://picard.sourceforge.net/index.shtml#GettingHelp
Exception in thread "main" net.sf.picard.PicardException: /home/pcingola/snpEff/data/genomes/hg19.dict already exists.  Delete this file and try again, or specify a different output file.
	at net.sf.picard.sam.CreateSequenceDictionary.doWork(CreateSequenceDictionary.java:114)
	at net.sf.picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:177)
	at net.sf.picard.sam.CreateSequenceDictionary.main(CreateSequenceDictionary.java:93)

Annotate using SnpEff
    Input file  : zzz.vcf
    Output file : ./zzz.snpeff.vcf
00:00:00.000	Reading configuration file '/home/pcingola/snpEff/snpEff.config'
00:00:00.173	done
00:00:00.173	Reading database for genome version 'hg19' from file '/home/pcingola//snpEff/data/hg19/snpEffectPredictor.bin' (this might take a while)
00:00:11.860	done
00:00:11.885	Building interval forest
00:00:17.755	done.
00:00:18.391	Genome stats :
# Genome name                : 'Homo_sapiens (USCS)'
# Genome version             : 'hg19'
# Has protein coding info    : true
# Genes                      : 25933
# Protein coding genes       : 20652
# Transcripts                : 44253
# Avg. transcripts per gene  : 1.71
# Protein coding transcripts : 36332
# Cds                        : 365442
# Exons                      : 429543
# Exons with sequence        : 409789
# Exons without sequence     : 19754
# Avg. exons per transcript  : 9.71
# Number of chromosomes      : 50
# Chromosomes names [sizes]  : '1' [249250621]	'2' [243199373]	'3' [198022430]	'4' [191154276]	'5' [180915260]	'6' [171115067]	'7' [159138663]	'X' [155270560]	'8' [146364022]	'9' [141213431]	'10' [135534747]	'11' [135006516]	'12' [133851895]	'13' [115169878]	'14' [107349540]	'15' [102531392]	'16' [90354753]	'17' [81195210]	'18' [78077248]	'20' [63025520]	'Y' [59373566]	'19' [59128983]	'22' [51304566]	'21' [48129895]	'6_ssto_hap7' [4905564]	'6_mcf_hap5' [4764535]	'6_cox_hap2' [4734611]	'6_mann_hap4' [4679971]	'6_qbl_hap6' [4609904]	'6_dbb_hap3' [4572120]	'6_apd_hap1' [4383650]	'17_ctg5_hap1' [1574839]	'4_ctg9_hap1' [582546]	'Un_gl000220' [156152]	'19_gl000209_random' [145745]	'Un_gl000213' [139339]	'17_gl000205_random' [119732]	'Un_gl000223' [119730]	'4_gl000194_random' [115071]	'Un_gl000228' [114676]	'Un_gl000219' [99642]	'Un_gl000218' [97454]	'Un_gl000211' [93165]	'Un_gl000222' [89310]	'4_gl000193_random' [88375]	'7_gl000195_random' [86719]	'1_gl000192_random' [79327]	'Un_gl000212' [60768]	'1_gl000191_random' [50281]	'M' [16571]	
00:00:18.391	Predicting variants
00:00:20.267	Creating summary file: snpEff_summary.html
00:00:20.847	Creating genes file: snpEff_genes.txt
00:00:25.026	done.
00:00:25.036	Logging
00:00:26.037	Checking for updates...

Annotating using GATK's VariantAnnotator:
    Input file  : zzz.vcf
    Output file : ./zzz.gatk.vcf
INFO  11:23:41,316 ArgumentTypeDescriptor - Dynamically determined type of zzz.vcf to be VCF 
INFO  11:23:41,343 HelpFormatter - -------------------------------------------------------------------------------- 
INFO  11:23:41,344 HelpFormatter - The Genome Analysis Toolkit (GATK) v2.4-9-g532efad, Compiled 2013/03/19 07:35:36 
INFO  11:23:41,344 HelpFormatter - Copyright (c) 2010 The Broad Institute 
INFO  11:23:41,344 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk 
INFO  11:23:41,347 HelpFormatter - Program Args: -T VariantAnnotator -R /home/pcingola/snpEff/data/genomes/hg19.fa -A SnpEff --variant zzz.vcf --snpEffFile ./zzz.snpeff.vcf -L zzz.vcf -o ./zzz.gatk.vcf 
INFO  11:23:41,347 HelpFormatter - Date/Time: 2013/04/12 11:23:41 
INFO  11:23:41,348 HelpFormatter - -------------------------------------------------------------------------------- 
INFO  11:23:41,348 HelpFormatter - -------------------------------------------------------------------------------- 
INFO  11:23:41,353 ArgumentTypeDescriptor - Dynamically determined type of zzz.vcf to be VCF 
INFO  11:23:41,356 ArgumentTypeDescriptor - Dynamically determined type of ./zzz.snpeff.vcf to be VCF 
INFO  11:23:41,399 GenomeAnalysisEngine - Strictness is SILENT 
INFO  11:23:41,466 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 
INFO  11:23:41,480 RMDTrackBuilder - Loading Tribble index from disk for file zzz.vcf 
INFO  11:23:41,503 RMDTrackBuilder - Loading Tribble index from disk for file ./zzz.snpeff.vcf 
WARN  11:23:41,505 RMDTrackBuilder - Index file /data/pcingola/Documents/projects/snpEff/gatk_test/./zzz.snpeff.vcf.idx is out of date (index older than input file), deleting and updating the index file 
INFO  11:23:41,506 RMDTrackBuilder - Creating Tribble index in memory for file ./zzz.snpeff.vcf 
INFO  11:23:41,914 RMDTrackBuilder - Writing Tribble index to disk for file /data/pcingola/Documents/projects/snpEff/gatk_test/./zzz.snpeff.vcf.idx 
INFO  11:23:42,076 IntervalUtils - Processing 33411 bp from intervals 
INFO  11:23:42,125 GenomeAnalysisEngine - Creating shard strategy for 0 BAM files 
INFO  11:23:42,134 GenomeAnalysisEngine - Done creating shard strategy 
INFO  11:23:42,134 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] 
INFO  11:23:42,135 ProgressMeter -        Location processed.sites  runtime per.1M.sites completed total.runtime remaining 
INFO  11:23:49,268 VariantAnnotator - Processed 9966 loci.
 
INFO  11:23:49,280 ProgressMeter -            done        3.34e+04    7.0 s        3.6 m    100.0%         7.0 s     0.0 s 
INFO  11:23:49,280 ProgressMeter - Total runtime 7.15 secs, 0.12 min, 0.00 hours 
INFO  11:23:49,953 GATKRunReport - Uploaded run statistics report to AWS S3 
</pre>
				<br>
        </section>

		<!-- IntegrationL Galaxy
		================================================== -->
		<section id="galaxy">
		  <div class="page-header">
			<h1>20. Integration: Galaxy </h1>
		  </div>
				<p class="lead">
				SnpEff provides integrations with <a href="http://galaxyproject.org/">Galaxy</a> project.
				</p>
				In order to install SnpEff in your own Galaxy server, you can use the <code>galaxy/*.xml</code> files provided in the main distribution. 
				<br>
				<br>
 				This is a screen capture from a Galaxy server (click to enlarge) <p>
				<center> <a href="images/snpEff_galaxy.png"><img src="images/snpEff_galaxy_small.png"></a> </center>
				<br>
				<br>
				Installing SnpEff in a Galaxy server:
				<p>
<pre>
# Set variable to snpEff install dir (we only use it for this install script)
export snpEffDir="$HOME/snpEff"

# Go to your galaxy 'tools' dir
cd galaxy-dist/tools

# Create a directory and copy the XML config files from SnpEff's distribution
mkdir snpEff
cd snpEff/
cp $snpEffDir/galaxy/* .

# Create links to JAR files
ln -s $snpEffDir/snpEff.jar
ln -s $snpEffDir/SnpSift.jar

# Link to config file
ln -s $snpEffDir/snpEff.config

# Allow scripts execution
chmod a+x *.{pl,sh}

# Copy genomes information
cd ../..
cp $snpEffDir/galaxy/tool-data/snpEff_genomes.loc tool-data/

# Edit Galaxy's tool_conf.xml and add all the tools
vi tool_conf.xml

-------------------- Begin: Edit tool_conf.xml --------------------
  &lt;!-- 
       Add this section to tool_conf.xml file in your galaxy distribution
	
       Note: The following lines should be added at the end of the 
             file, right before "&lt;/toolbox&gt;" line
  --&gt;
  &lt;section name="snpEff tools" id="snpEff_tools"&gt;
    &lt;tool file="snpEff/snpEff.xml" /&gt;
    &lt;tool file="snpEff/snpEff_download.xml" /&gt;
    &lt;tool file="snpEff/snpSift_annotate.xml" /&gt;
    &lt;tool file="snpEff/snpSift_caseControl.xml" /&gt;
    &lt;tool file="snpEff/snpSift_filter.xml" /&gt;
    &lt;tool file="snpEff/snpSift_int.xml" /&gt;
  &lt;/section&gt;
-------------------- End: Edit tool_conf.xml --------------------

# Run galaxy and check that the new menues appear
./run.sh

</pre>

				</p>
        </section>
				

		<!-- Utilities
		================================================== -->
		<section id="utils">
		  <div class="page-header">
			<h1>21. Commands and utilities </h1>
		  </div>
				<p class="lead">
				SnpEff provides several other commands and utilities that can be useful for genomic data analysis.
				</p>

				<p>
				Most of this manual was dedicated the <code>SnpEff eff</code> and <code>SnpEff build</code> commands, which annotate effects and build databases respectively.
				Here we describe all the other commands and some scripts provided, that are useful for genomic data analysis.
				</p>

				<h3> SnpEff closest</h3>
				<p>
				Annotates using the closest genomic region (e.g. exon, transcript ID, gene name) and distance in bases.
				<br>
				Example:
<pre>
$ java -Xmx4g -jar snpEff.jar closest GRCh37.66 test.vcf
##INFO=&lt;ID=CLOSEST,Number=4,Type=String,Description="Closest exon: Distance (bases), exons Id, transcript Id, gene name">
1       12078   .       G       A       25.69   PASS    AC=2;AF=0.048;CLOSEST=0,exon_1_11869_12227,ENST00000456328,DDX11L1
1       16097   .       T       G       42.42   PASS    AC=9;AF=0.0113;CLOSEST=150,exon_1_15796_15947,ENST00000423562,WASH7P
1       40261   .       C       A       366.26  PASS    AC=30;AF=0.484;CLOSEST=4180,exon_1_35721_36081,ENST00000417324,FAM138A
1       63880   .       C       T       82.13   PASS    AC=10;AF=0.0400;CLOSEST=0,exon_1_62948_63887,ENST00000492842,OR4G11P
</pre>

				For intance, in the third line (1:16097 T G), it added the tag <code>CLOSEST=150,exon_1_15796_15947,ENST00000423562,WASH7P</code>
				, which means that the variant is 150 bases away from exon "exon_1_15796_15947". 
				The exon belongs to transcript "ENST00000423562" of gene "WASH7P".
				<br>
				<br>
				<i class="icon-hand-right"></i> If multiple markers are available (at the same distance) the one belonging to the longest mRna transcript is shown.
				<br>
				<br>
				The input can also be a BED file, the output file has the same information as CLOSEST info field, added to the fourth column of the output BED file:
<pre>
$ snpeff closest -bed GRCh37.66 test.bed 
1	12077	12078	line_1;0,exon_1_11869_12227,ENST00000456328,DDX11L1
1	16096	16097	line_2;150,exon_1_15796_15947,ENST00000423562,WASH7P
1	40260	40261	line_3;4180,exon_1_35721_36081,ENST00000417324,FAM138A
1	63879	63880	line_4;0,exon_1_62948_63887,ENST00000492842,OR4G11P
</pre>
				</p>

				<a id="count">&nbsp;</a>
				<h3> SnpEff count </h3>
				<p>
				As the name suggests, <code>snpEff count</code> command counts how many reads and bases from a BAM file hit a gene, transcript, exon, intron, etc.
				Input files can be in BAM, SAM, VCF, BED or BigBed formats.
				<br>
				A summary HTML file with charts is generated. Here are some examples:
				<img src="images/snpeff_count_01.png">
				<p>
				<img src="images/snpeff_count_02.png">
				<br>
				If you need to count how many reads (and bases) from a BAM file hit each genomic region, you can use 'count' utility.<br>
				The command line is quite simple. E.g. in order to count how many reads (from N BAM files) hit regions of the human genome, you simply run: <br>
<pre>
java -Xmx4g -jar snpEff.jar count GRCh37.68 readsFile_1.bam readsFile_2.bam ...  readsFile_N.bam > countReads.txt 
</pre>
				<br>
				<br>
				The output is a TXT (tab-separated) file, that looks like this:
				<br>
<pre>
chr  start  end       type                IDs                         Reads:readsFile_1.bam  Bases:readsFile_1.bam  Reads:readsFile_2.bam  Bases:readsFile_2.bam ...
1    1      11873     Intergenic          DDX11L1                     130                    6631                   50                     2544
1    1      249250621 Chromosome          1                           2527754                251120400              2969569                328173439
1    6874   11873     Upstream            NR_046018;DDX11L1           130                    6631                   50                     2544
1    9362   14361     Downstream          NR_024540;WASH7P            243                    13702                  182                    9279
1    11874  12227     Exon                exon_1;NR_046018;DDX11L1    4                      116                    2                      102
1    11874  14408     Gene                DDX11L1                     114                    7121                   135                    6792
1    11874  14408     Transcript          NR_046018;DDX11L1           114                    7121                   135                    6792
1    12228  12229     SpliceSiteDonor     exon_1;NR_046018;DDX11L1    3                      6                      0                      0
1    12228  12612     Intron              intron_1;NR_046018;DDX11L1  13                     649                    0                      0
1    12611  12612     SpliceSiteAcceptor  exon_2;NR_046018;DDX11L1    0                      0                      0                      0
1    12613  12721     Exon                exon_2;NR_046018;DDX11L1    3                      24                     1                      51
1    12722  12723     SpliceSiteDonor     exon_2;NR_046018;DDX11L1    3                      6                      0                      0
1    12722  13220     Intron              intron_2;NR_046018;DDX11L1  22                     2110                   20                     987
1    13219  13220     SpliceSiteAcceptor  exon_3;NR_046018;DDX11L1    5                      10                     1                      2
1    13221  14408     Exon                exon_3;NR_046018;DDX11L1    82                     4222                   113                    5652
1    14362  14829     Exon                exon_11;NR_024540;WASH7P    37                     1830                   7                      357
1    14362  29370     Transcript          NR_024540;WASH7P            704                    37262                  524                    34377
1    14362  29370     Gene                WASH7P                      704                    37262                  524                    34377
1    14409  19408     Downstream          NR_046018;DDX11L1           122                    7633                   39                     4254
</pre>
				The columns are:
				<ul>
					<li> Column 1: Chromosome name
					<li> Column 2: Genomic region start
					<li> Column 3: Genomic region end
					<li> Column 4: Genomic region type (e.g. Exon, Gene, SpliceSiteDonor, etc.)
					<li> Column 5: ID (e.g. exon ID ; transcript ID; gene ID)
					<li> Column 6: Count of reads (in file readsFile_1.bam) intersecting genomic region.
					<li> Column 7: Count of bases (in file readsFile_1.bam) intersecting genomic region, i.e. each read is intersected and the resulting number of bases added.
					<li> Column ...:  (repeat count reads and bases for each BAM file provided)
				</ul>
				<br>
				<b>Totals and Binomial model</b>
				<br>
				Using command line option <code>-p</code>, you can calculate p-values based on a Binomial model.
				For example (output edited for the sake of brevity):
<pre>
$ java -Xmx4g -jar snpEff.jar count -v BDGP5.69 fly.bam > countReads.txt
00:00:00.000	Reading configuration file 'snpEff.config'
...
00:00:12.148	Calculating probability model for read length 50
...
type               p.binomial             reads.fly  expected.fly  pvalue.fly
Chromosome         1.0                    205215     205215        1.0
Downstream         0.29531659795589793    59082      60603         1.0
Exon               0.2030262729897713     53461      41664         0.0
Gene               0.49282883664487515    110475     101136        0.0
Intergenic         0.33995644860241336    54081      69764         0.9999999963234701
Intron             0.3431415343615103     72308      70418         9.06236369003514E-19
RareAminoAcid      9.245222303207472E-7   3          0             9.879186871519377E-4
SpliceSiteAcceptor 0.014623209601955131   3142       3001          0.005099810118785825
SpliceSiteDonor    0.015279075154423956   2998       3135          0.9937690786738507
Transcript         0.49282883664487515    110475     101136        0.0
Upstream           0.31499087549896493    64181      64641         0.9856950416729887
Utr3prime          0.03495370828296416    8850       7173          1.1734134297889064E-84
Utr5prime          0.02765432673262785    8146       5675          7.908406840800345E-215
</pre>
				<br>
				The columns in for this table are (in the previous example the input file was 'fly.bam' so fileName is 'fly'):
				<ul>
					<li> type : Type of interval
					<li> p.binomial : Probability that a random read hits this 'type' of interval (in binomial model)
					<li> reads.fileName : Total number of reads in 'fileName' (user provided BAM/SAM file)
					<li> expected.fileName : Expected number of reads hitting this 'type' of interval (for user provided BAM/SAM file)
					<li> pvalue.fileName : p-value that 'reads.fileName' reads or more hit this 'type' of interval (for user provided BAM/SAM file)
					<li> Column ...:  (repeat last three column for each BAM/SAM file provided by the user)
				</ul>
				<br>
				<b>User defined intervals </b>
				<br>
				You can add user defined intervals using <code>-i file.bed</code> command line option.
				The option can be used multiple times, thus allowing multiple BED files to be added.
				<br>
				Example : You want to know how many reads intersect each peak from a peak detection algorithm:
<pre>
java -Xmx4g -jar snpEff.jar count -i peaks.bed GRCh37.68 reads.bam
</pre>
				</p>

				<h3> SnpEff databases </h3>
				<p>
				This command provides a list of configured databases, i.e. available in snpEff.config file.
				<br>
				Example:
<pre>
$ java -jar snpEff.jar databases
Genome                                                          Organism                                                        Database download link
------                                                          --------                                                        ----------------------
ADWO01                                                          Prevotella bryantii B14                                         http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Prevotella bryantii B14.zip
Acholeplasma_laidlawii_PG_8A_uid58901                           Acholeplasma_laidlawii_PG_8A_uid58901                           http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acholeplasma_laidlawii_PG_8A_uid58901.zip
Achromobacter_xylosoxidans_A8_uid59899                          Achromobacter_xylosoxidans_A8_uid59899                          http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Achromobacter_xylosoxidans_A8_uid59899.zip
Acidaminococcus_fermentans_DSM_20731_uid43471                   Acidaminococcus_fermentans_DSM_20731_uid43471                   http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidaminococcus_fermentans_DSM_20731_uid43471.zip
Acidaminococcus_intestini_RyC_MR95_uid74445                     Acidaminococcus_intestini_RyC_MR95_uid74445                     http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidaminococcus_intestini_RyC_MR95_uid74445.zip
Acidianus_hospitalis_W1_uid66875                                Acidianus_hospitalis_W1_uid66875                                http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidianus_hospitalis_W1_uid66875.zip
Acidilobus_saccharovorans_345_15_uid51395                       Acidilobus_saccharovorans_345_15_uid51395                       http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidilobus_saccharovorans_345_15_uid51395.zip
Acidimicrobium_ferrooxidans_DSM_10331_uid59215                  Acidimicrobium_ferrooxidans_DSM_10331_uid59215                  http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidimicrobium_ferrooxidans_DSM_10331_uid59215.zip
Acidiphilium_cryptum_JF_5_uid58447                              Acidiphilium_cryptum_JF_5_uid58447                              http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidiphilium_cryptum_JF_5_uid58447.zip
Acidiphilium_multivorum_AIU301_uid63345                         Acidiphilium_multivorum_AIU301_uid63345                         http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidiphilium_multivorum_AIU301_uid63345.zip
Acidithiobacillus_caldus_SM_1_uid70791                          Acidithiobacillus_caldus_SM_1_uid70791                          http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidithiobacillus_caldus_SM_1_uid70791.zip
Acidithiobacillus_ferrivorans_SS3_uid67387                      Acidithiobacillus_ferrivorans_SS3_uid67387                      http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidithiobacillus_ferrivorans_SS3_uid67387.zip
Acidithiobacillus_ferrooxidans_ATCC_23270_uid57649              Acidithiobacillus_ferrooxidans_ATCC_23270_uid57649              http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidithiobacillus_ferrooxidans_ATCC_23270_uid57649.zip
Acidithiobacillus_ferrooxidans_ATCC_53993_uid58613              Acidithiobacillus_ferrooxidans_ATCC_53993_uid58613              http://sourceforge.net/projects/snpeff/files/databases/v3.2/snpEff_v3.2_Acidithiobacillus_ferrooxidans_ATCC_53993_uid58613.zip
...
...
</pre>

				<h3> SnpEff download </h3>
				<p>
				This command downloads and installs a database. 
				<br>
				<i class="icon-warning-sign"></i>Note that the database must be configured in <code>snpEff.config</code> and available at the download site.
				<br>
				Example: Download and install C.Elegans genome

<pre>
$ java -jar snpEff.jar download -v WBcel215.69
00:00:00.000	Downloading database for 'WBcel215.69'
00:00:00.002	Connecting to http://downloads.sourceforge.net/project/snpeff/databases/v3_1/snpEff_v3_1_WBcel215.69.zip
00:00:00.547	Copying file (type: application/zip, modified on: Sat Dec 01 20:59:55 EST 2012)
00:00:00.547	Local file name: 'snpEff_v3_1_WBcel215.69.zip'
00:00:01.949	Downloaded 1049506 bytes
00:00:03.624	Downloaded 2135266 bytes
00:00:05.067	Downloaded 3185026 bytes
00:00:06.472	Downloaded 4234786 bytes
00:00:07.877	Downloaded 5284546 bytes
00:00:09.580	Downloaded 6374626 bytes
00:00:11.005	Downloaded 7424386 bytes
00:00:12.410	Downloaded 8474146 bytes
00:00:13.815	Downloaded 9523906 bytes
00:00:15.358	Downloaded 10604226 bytes
00:00:16.761	Downloaded 11653666 bytes
00:00:18.168	Downloaded 12703426 bytes
00:00:19.573	Downloaded 13753186 bytes
00:00:21.198	Downloaded 14837506 bytes
00:00:22.624	Downloaded 15887266 bytes
00:00:24.029	Downloaded 16937026 bytes
00:00:25.434	Downloaded 17986786 bytes
00:00:26.864	Downloaded 19036546 bytes
00:00:28.269	Downloaded 20086306 bytes
00:00:29.155	Donwload finished. Total 20748168 bytes.
00:00:29.156	Local file name: '/home/pcingola//snpEff/data/WBcel215.69/snpEffectPredictor.bin'
00:00:29.156	Extracting file 'data/WBcel215.69/snpEffectPredictor.bin' to '/home/pcingola//snpEff/data/WBcel215.69/snpEffectPredictor.bin'
00:00:29.157	Creating local directory: '/home/pcingola/snpEff/data/WBcel215.69'
00:00:29.424	Unzip: OK
00:00:29.424	Done
</pre>

				</p>

				<h3> SnpEff dump </h3>
				<p>
				Dump the contents of a database to a text file, a BED file or a tab separated TXT file (that can be loaded into R).
				<br>
				<br>
				<b>BED file example</b>:
<pre>
$ java -jar snpEff.jar download -v GRCh37.70
$ java -Xmx4g -jar snpEff.jar dump -v -bed GRCh37.70 > GRCh37.70.bed
00:00:00.000	Reading database for genome 'GRCh37.70' (this might take a while)
00:00:32.476	done
00:00:32.477	Building interval forest
00:00:45.928	Done.
</pre>

				The output file looks like a typical BED file (chr \t start \t end \t name).
				<br>
				<i class="icon-warning-sign"></i>Keep in mind that BED file coordinates are zero based, semi-open intervals. 
				So a 2 base interval at (one-based) positions 100 and 101 is expressed as a BED interval [99 - 101].
				<br>
<pre>
$ head GRCh37.70.bed
1	0	249250621	Chromosome_1
1	111833483	111863188	Gene_ENSG00000134216
1	111853089	111863002	Transcript_ENST00000489524
1	111861741	111861861	Cds_CDS_1_111861742_111861861
1	111861948	111862090	Cds_CDS_1_111861949_111862090
1	111860607	111860731	Cds_CDS_1_111860608_111860731
1	111861114	111861300	Cds_CDS_1_111861115_111861300
1	111860305	111860427	Cds_CDS_1_111860306_111860427
1	111862834	111863002	Cds_CDS_1_111862835_111863002
1	111853089	111853114	Utr5prime_exon_1_111853090_111853114
</pre>

				<b>TXT file example</b>:
<pre>
$ java -Xmx4g -jar snpEff.jar dump -v -txt GRCh37.70 > GRCh37.70.txt
00:00:00.000	Reading database for genome 'GRCh37.70' (this might take a while)
00:00:31.961	done
00:00:31.962	Building interval forest
00:00:45.467	Done.
</pre>
				In this case, the ouput file looks like a typical BED file (chr \t start \t end \t name):
<pre>
$ head GRCh37.70.txt
chr start       end        strand  type         id                          geneName  geneId            numberOfTranscripts  canonicalTranscriptLength  transcriptId     cdsLength  numerOfExons  exonRank  exonSpliceType
1   1           249250622  +1      Chromosome   1                                                                                                       
1   111833484   111863189  +1      Gene         ENSG00000134216             CHIA      ENSG00000134216   10                   1431                                                                  
1   111853090   111863003  +1      Transcript   ENST00000489524             CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9                     
1   111861742   111861862  +1      Cds          CDS_1_111861742_111861861   CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9                     
1   111861949   111862091  +1      Cds          CDS_1_111861949_111862090   CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9                     
1   111853090   111853115  +1      Utr5prime    exon_1_111853090_111853114  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             1         ALTTENATIVE_3SS
1   111854311   111854341  +1      Utr5prime    exon_1_111854311_111854340  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             2         SKIPPED
1   111860608   111860732  +1      Exon         exon_1_111860608_111860731  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             5         RETAINED
1   111853090   111853115  +1      Exon         exon_1_111853090_111853114  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             1         ALTTENATIVE_3SS
1   111861742   111861862  +1      Exon         exon_1_111861742_111861861  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             7         RETAINED
</pre>

				The format is
				<table class="table table-striped">
					<tr> <th> Column </th><th> Meaning </th></tr>
					<tr> <td> chr </td> <td> Chromosome name</td> </tr>
					<tr> <td> start </td> <td> Marker start (one-based coordinate)</td> </tr>
					<tr> <td> end </td> <td> Marker end (one-based coordinate)</td> </tr>
					<tr> <td> strand </td> <td> Strand (positive or negative)</td> </tr>
					<tr> <td> type </td> <td> Type of marker (e.g. exon, transcript, etc.)</td> </tr>
					<tr> <td> id </td> <td> ID. E.g. if it's a Gene, then it may be ENSEBML's gene ID</td> </tr>
					<tr> <td> geneName </td> <td> Gene name, if marker is within a gene (exon, transcript, UTR, etc.), empty otherwise (e.g. intergenic) </td> </tr>
					<tr> <td> geneId </td> <td> Gene is, if marker is within a gene</td> </tr>
					<tr> <td> numberOfTranscripts </td> <td>Number of transcripts in the gene</td> </tr>
					<tr> <td> canonicalTranscriptLength </td> <td>CDS length of canonical transcript.</td> </tr>
					<tr> <td> transcriptId </td> <td> Transcript ID, if marker is within a transcript</td> </tr>
					<tr> <td> cdsLength </td> <td> CDS length of the transcript</td> </tr>
					<tr> <td> numerOfExons </td> <td> Number of exons in this transcript</td> </tr>
					<tr> <td> exonRank </td> <td> Exon rank, if marker is an exon</td> </tr>
					<tr> <td> exonSpliceType </td> <td> Exon splice type, if marker is an exon </td> </tr>
				</table>
				</p>

				<h3> SnpEff genes2bed </h3>
				<p>
				Dumps a selected set of genes as BED intervals.
				<br>
				<br>
				<i class="icon-warning-sign"></i>The functionality of this command is a subset of <code>SnpEff dump</code>, so it is likely to be deprecated in the future.
				<br>
				<br>
				Example:
<pre>
$ java -Xmx4g -jar snpEff.jar genes2bed GRCh37.66 DDX11L1 WASH7P
#chr	start	end	geneName;geneId
1	11868	14411	DDX11L1;ENSG00000223972
1	14362	29805	WASH7P;ENSG00000227232
</pre>
				</p>

				<h3> SnpEff cds &amp; SnpEff protein </h3>
				<p>
				These commands perform SnpEff database sanity checks.
				They calculate CDS and protein sequences from a SnpEff database and then compare the results to a FASTA file (having the "correct" sequences).
				<br>
				The commands are invoked automatically when building databases, so there is no need for the user to invoke them manually.
				</p>

				<h3> SnpEff len </h3>
				<p>
				Calculates the genomic length of every type of marker (Gene, Exon, Utr, etc.).
				Length is calculated by overlapping all markers and counting the number of bases (e.g. a base is counted as 'Exon' if any exon falls onto that base).
				This command also reports the probability of a Binomial model.
				<br>
				<br>
				<i class="icon-hand-right"></i> Parameter <code>-r num</code> adjusts the model for a read length of 'num' bases. That is, if two markers of the same type are closer than 'num' bases, it joins them by incliding the bases separating them.
				<br>
				<br>
				E.g.:
<pre>
$ java -Xmx1g -jar snpEff.jar len -r 100 BDGP5.69
marker                   size    count     raw_size raw_count    binomial_p
Cds                  22767006    56955     45406378    122117    0.13492635563570918
Chromosome          168736537       15    168736537        15    1.0
Downstream           49570138     5373    254095562     50830    0.29377240330587084
Exon                 31275946    61419     63230008    138474    0.18535372691689175
Gene                 82599213    11659     87017182     15222    0.4895158717166277
Intergenic           56792611    11637     56792611     11650    0.3365756581812509
Intron               55813748    42701    168836797    113059    0.33077452573297744
SpliceSiteAcceptor      97977    48983       226118    113059    5.806507691929223E-4
SpliceSiteDonor        101996    50981       226118    113059    6.044689657225808E-4
Transcript           82599213    11659    232066805     25415    0.4895158717166277
Upstream             52874082     5658    254044876     50830    0.3133528928592389
Utr3prime             5264120    13087     10828991     24324    0.031197274126824114
Utr5prime             3729197    19324      6368070     33755    0.02210070839607192
</pre>
				</p>
				Column meaning:
				<ul>
					<li> marker : Type of marker interval
					<li> size : Size of all intervals in the genome, after overlap and join.
					<li> count : Number of intervals in the genome, after overlap and join.
					<li> raw_size : Size of all intervals in the genome. Note that this could be larger than the genome.
					<li> raw_count : Number of intervals in the genome.
				</ul>

				<h3> Scripts </h3>
				<p>
				Several scripts are provided in the <code>scripts</code> directory.
				Here we briefly describe their functionality:
				<table class="table table-striped">
					<tr> <th> Script </th><th> Functionality </th></tr>
					<tr> <td> sam2fastq.pl </td> 
											<td> Convert a SAM input to a FASTQ output. Example:
<pre>
samtools view test.bam | ./scripts/bam2fastq.pl | head -n 12
@HWI-ST1220:76:D12CHACXX:7:2207:18986:95756
CGACAATGCACGACAGAGGAAGCAGAACAGATATTTAGATTGCCTCTCATT
+
CCCFFFFFGHHHHIIJIJJIJIJJIJJIIIJIIHIJJJIJJIJJJJJJIJJ
@HWI-ST1220:76:D12CHACXX:7:2206:4721:162268
ATATTATAGGGAGAAATATGATCGCGTATGCGAGAGTAGTGCCAACATATT
+
@@@DDD>DAB;?DGGEGGBCD>BFGI?FCFFBFGG@<?B*?BFB9FGII@E
@HWI-ST1220:76:D12CHACXX:7:1304:13069:17740
ATAGGGAGAAATATGATCGCGTATGCGAGAGTAGTGCCAACATATTGTGCT
+
CCCFFFFFHHHHHJJJJJJJJGIJJJJJJAGBGGIIIJJJJJJJJJJJIJG
</pre> </td> </tr>

					<tr> <td> fasta2tab.pl </td> <td> 
													Convert a fasta file to a two column tab-separated TXT file (name \t sequence)
Example (output truncated for brevity)
<pre>
$ zcat ce6.fa.gz | ./scripts/fasta2tab.pl 
chrI    gcctaagcctaagcctaagcctaagcctaagcctaagcctaagcct...
chrV    GAATTcctaagcctaagcctaagcctaagcctaagcctaagcctaa...
chrII   cctaagcctaagcctaagcctaagcctaagcctaagcctaagccta...
chrM    CAGTAAATAGTTTAATAAAAATATAGCATTTGGGTTGCTAAGATAT...
chrX    ctaagcctaagcctaagcctaagcctaagcctaagcctaagcctaa...
chrIV   cctaagcctaagcctaagcctaagcctaagcctaagcctaagccta...
chrIII  cctaagcctaagcctaagcctaagcctaagcctaagcctaagccta...
</pre> </td> </tr>

					<tr> <td> fastaSplit.pl </td> <td>	Split a multiple sequence FASTA file to individial files.
														Example: Creates one file per chromosome
<pre>
$ zcat ce6.fa.gz | ./scripts/fastaSplit.pl
Writing to chrI.fa
Writing to chrV.fa
Writing to chrII.fa
Writing to chrM.fa
Writing to chrX.fa
Writing to chrIV.fa
Writing to chrIII.fa
</pre> 
														</td> </tr>

					<tr> <td> hist.pl </td> <td>	Given a list of numbers (one per line), shows a histogram. Note: It requires R.
													<br>
													Example: Extract the file sizes in a directory and show a histogram
<pre>
$ ls -al scripts/ | tr -s " " | cut -f 5 -d " " | ./scripts/hist.pl 
</pre> 
													Creates the following plot
													<br>
													<img src=hist_example.png>
													</td> </tr>


					<tr> <td> maPlot.pl <br> 
								plot.pl <br> 
								qqplot.pl <br> 
								smoothScatter.pl
								</td> 
											<td>	Similar to 'hist.pl', these perform plots based on input from STDIN. 
													Note that in some cases, inputs are expected to be probabilities (qqplot.pl) or pairs of numbers (maPlot.pl).
<pre>
$ ls -al scripts/ | tr -s " " | cut -f 5 -d " " | ./scripts/plot.pl
</pre> 
													Creates the following plot
													<br>
													<img src=plot_example.png>
													</td> </tr>

					<tr> <td> queue.pl </td> <td>	Process a list of statements in parallel according to the number of CPUs available in the local machine </td> </tr>

					<tr> <td> splitChr.pl </td> <td> Splits a file by chromosome. Works on any tab separated file that the first column is CHR field (e.g. BED, VCF, etc.)
													<br>
													Example:
<pre>
$ cat large_test.vcf | ./scripts/splitChr.pl 
Input line 28. Creating file 'chr1.txt'
Input line 13332. Creating file 'chr2.txt'
Input line 22097. Creating file 'chr3.txt'
Input line 29289. Creating file 'chr4.txt'
Input line 34236. Creating file 'chr5.txt'
Input line 39899. Creating file 'chr6.txt'
Input line 47120. Creating file 'chr7.txt'
Input line 53371. Creating file 'chr8.txt'
Input line 57810. Creating file 'chr9.txt'
Input line 63005. Creating file 'chr10.txt'
Input line 68080. Creating file 'chr11.txt'
Input line 76629. Creating file 'chr12.txt'
Input line 83071. Creating file 'chr13.txt'
Input line 85124. Creating file 'chr14.txt'
Input line 89281. Creating file 'chr15.txt'
Input line 93215. Creating file 'chr16.txt'
Input line 99081. Creating file 'chr17.txt'
Input line 106405. Creating file 'chr18.txt'
Input line 108330. Creating file 'chr19.txt'
Input line 118568. Creating file 'chr20.txt'
Input line 121795. Creating file 'chr21.txt'
Input line 123428. Creating file 'chr22.txt'
Input line 126520. Creating file 'chrX.txt'
Input line 129094. Creating file 'chrY.txt'
Input line 129113. Creating file 'chrMT.txt'
</pre> </td> </tr>

					<tr> <td> uniqCount.pl </td> <td> Count number of unique lines. It's the same as doing <code>cat lines.tst | sort | uniq -c </code>, but much faster. Particularly useful for very large inputs. </td> </tr>

					<tr> <td> vcfEffOnePerLine.pl </td> <td> Splits EFF fields in a VCF file, creating mutiple lines, each one with only one effect. 
															<br>
															Very useful for filtering with SnpSift.
															<br>
															Example: </td> </tr>
				</table>
<pre>
$ cat test.stop.vcf 
1	897062	.	C	T	100.0	PASS	AC=1;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*|642|KLHL17||CODING|NM_198317|),UPSTREAM(MODIFIER||||576|PLEKHN1||CODING|NM_001160184|),UPSTREAM(MODIFIER||||611|PLEKHN1||CODING|NM_032129|),UPSTREAM(MODIFIER||||749|NOC2L||CODING|NM_015658|)

$ cat test.stop.vcf | ./scripts/vcfEffOnePerLine.pl 
1	897062	.	C	T	100.0	PASS	AC=1;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*|642|KLHL17||CODING|NM_198317|)
1	897062	.	C	T	100.0	PASS	AC=1;EFF=UPSTREAM(MODIFIER||||576|PLEKHN1||CODING|NM_001160184|)
1	897062	.	C	T	100.0	PASS	AC=1;EFF=UPSTREAM(MODIFIER||||611|PLEKHN1||CODING|NM_032129|)
1	897062	.	C	T	100.0	PASS	AC=1;EFF=UPSTREAM(MODIFIER||||749|NOC2L||CODING|NM_015658|)
</pre> 


				</p>
        </section>

      </div>
    </div>
  </div>



    <!-- Footer
    ================================================== -->
      <footer>
        <p class="text-center">&copy; <a class="body" href="http://www.linkedin.com/in/pablocingolani">Pablo Cingolani 2013</a></p>
      </footer>


    <!-- Placed at the end of the document so the pages load faster -->
    <script type="text/javascript" src="http://platform.twitter.com/widgets.js"></script>
    <script src="js/jquery.js"></script>
    <script src="js/bootstrap-transition.js"></script>
    <script src="js/bootstrap-alert.js"></script>
    <script src="js/bootstrap-modal.js"></script>
    <script src="js/bootstrap-dropdown.js"></script>
    <script src="js/bootstrap-scrollspy.js"></script>
    <script src="js/bootstrap-tab.js"></script>
    <script src="js/bootstrap-tooltip.js"></script>
    <script src="js/bootstrap-popover.js"></script>
    <script src="js/bootstrap-button.js"></script>
    <script src="js/bootstrap-collapse.js"></script>
    <script src="js/bootstrap-carousel.js"></script>
    <script src="js/bootstrap-typeahead.js"></script>
    <script src="js/bootstrap-affix.js"></script>

    <script src="js/holder/holder.js"></script>
    <script src="js/google-code-prettify/prettify.js"></script>

    <script src="js/application.js"></script>



  </body>
</html>
