
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="SnpEff and SnpSift">
      
      
        <meta name="author" content="Pablo Cingolani">
      
      
      
        <link rel="prev" href="../integration/">
      
      
        <link rel="next" href="../human_genomes/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.19">
    
    
      
        <title>Commands and utilities - SnpEff & SnpSift</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.eebd395e.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="default" data-md-color-accent="indigo">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#commands-and-utilities" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="SnpEff &amp; SnpSift" class="md-header__button md-logo" aria-label="SnpEff & SnpSift" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SnpEff & SnpSift
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Commands and utilities
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/pcingola/SnpEff" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    SnpEff
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../introduction/" class="md-tabs__link md-tabs__link--active">
        SnpEff
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../snpsift/introduction/" class="md-tabs__link">
        SnpSift
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../download/" class="md-tabs__link">
        Download & Install
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../examples/" class="md-tabs__link">
        Examples
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../about/" class="md-tabs__link">
        About
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="SnpEff &amp; SnpSift" class="md-nav__button md-logo" aria-label="SnpEff & SnpSift" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    SnpEff & SnpSift
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/pcingola/SnpEff" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    SnpEff
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
      
      
      
        <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../examples/" class="md-nav__link">
        Usage examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../license/" class="md-nav__link">
        License
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          SnpEff
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          SnpEff
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../running/" class="md-nav__link">
        Running SnpEff
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../commandline/" class="md-nav__link">
        Commands &amp; command line options
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../inputoutput/" class="md-nav__link">
        Input &amp; output files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../cansersamples/" class="md-nav__link">
        Cancer samples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../additionalann/" class="md-nav__link">
        Additional annotations
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../outputsummary/" class="md-nav__link">
        Output summary files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshooting/" class="md-nav__link">
        Troubleshooting
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_db/" class="md-nav__link">
        Building databases
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_db_gff_gtf/" class="md-nav__link">
        Build databses. GTF / GFF details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_reg/" class="md-nav__link">
        Building databases. Regulatory and Non-coding
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_pdb/" class="md-nav__link">
        Building databases. PDB and AlphaFold
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../integration/" class="md-nav__link">
        Integration. GATK and Galaxy
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Commands and utilities
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Commands and utilities
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#snpeff-closest" class="md-nav__link">
    SnpEff closest
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-count" class="md-nav__link">
    SnpEff count
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-databases" class="md-nav__link">
    SnpEff databases
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-download" class="md-nav__link">
    SnpEff download
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-dump" class="md-nav__link">
    SnpEff dump
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-genes2bed" class="md-nav__link">
    SnpEff genes2bed
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-cds-snpeff-protein" class="md-nav__link">
    SnpEff cds &amp; SnpEff protein
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#snpeff-len" class="md-nav__link">
    SnpEff len
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts" class="md-nav__link">
    Scripts
  </a>
  
    <nav class="md-nav" aria-label="Scripts">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sam2fastqpl" class="md-nav__link">
    sam2fastq.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fasta2tabpl" class="md-nav__link">
    fasta2tab.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fastasplitpl" class="md-nav__link">
    fastaSplit.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plothistogrampl" class="md-nav__link">
    plotHistogram.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotmapl-plotpl-plotqqpl-plotsmoothscatterpl" class="md-nav__link">
    plotMA.pl, plot.pl, plotQQ.pl, plotSmoothScatter.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#queuepl" class="md-nav__link">
    queue.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#splitchrpl" class="md-nav__link">
    splitChr.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#uniqcountpl" class="md-nav__link">
    uniqCount.pl
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcfeffoneperlinepl" class="md-nav__link">
    vcfEffOnePerLine.pl
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../human_genomes/" class="md-nav__link">
        Human Genomes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../faq/" class="md-nav__link">
        Frequently Asked Questions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          SnpSift
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          SnpSift
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/filter/" class="md-nav__link">
        SnpSift Filter
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/annotate/" class="md-nav__link">
        SnpSift Annotate
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/casecontrol/" class="md-nav__link">
        SnpSift CaseControl
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intervals/" class="md-nav__link">
        SnpSift Intervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intervalsindex/" class="md-nav__link">
        SnpSift Intervals Index
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/join/" class="md-nav__link">
        SnpSift Join
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/rmrefgen/" class="md-nav__link">
        SnpSift RmRefGen
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/tstv/" class="md-nav__link">
        SnpSift TsTv
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/extractfields/" class="md-nav__link">
        SnpSift Extract Fields
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/varianttype/" class="md-nav__link">
        SnpSift Variant Type
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/gwascatalog/" class="md-nav__link">
        SnpSift GWAS Catalog
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/dbnsfp/" class="md-nav__link">
        SnpSift dbNSFP
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/split/" class="md-nav__link">
        SnpSift Split
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/phastcons/" class="md-nav__link">
        SnpSift PhastCons
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/concordance/" class="md-nav__link">
        SnpSift Concordance
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/private/" class="md-nav__link">
        SnpSift Private
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/vcf2ped/" class="md-nav__link">
        SnpSift Vcf2Tped
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intersect/" class="md-nav__link">
        SnpSift Intersect
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/rminfo/" class="md-nav__link">
        SnpSift RmInfo
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/genesets/" class="md-nav__link">
        SnpSift GeneSets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/gt/" class="md-nav__link">
        SnpSift GT
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/vcfcheck/" class="md-nav__link">
        SnpSift VcfCheck
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/faq/" class="md-nav__link">
        Frequently asked questions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          Download & Install
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Download & Install
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../download/" class="md-nav__link">
        Download and install
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          Examples
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Examples
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../examples/" class="md-nav__link">
        Examples
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          About
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          About
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../about/" class="md-nav__link">
        About
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../help/" class="md-nav__link">
        Help and Bugs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../xiangyi_lu_donate/" class="md-nav__link">
        In Memory
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="commands-and-utilities">Commands and utilities</h1>
<p>SnpEff provides several other commands and utilities that can be useful for genomic data analysis.</p>
<p>Most of this manual was dedicated the <code>SnpEff eff</code> and <code>SnpEff build</code> commands, which annotate effects and build databases respectively.
Here we describe all the other commands and some scripts provided, that are useful for genomic data analysis.</p>
<h3 id="snpeff-closest">SnpEff closest</h3>
<p>Annotates using the closest genomic region (e.g. exon, transcript ID, gene name) and distance in bases.</p>
<p>Example:
<div class="highlight"><pre><span></span><code>$ java -Xmx8g -jar snpEff.jar closest GRCh37.66 test.vcf
##INFO=&lt;ID=CLOSEST,Number=4,Type=String,Description=&quot;Closest exon: Distance (bases), exons Id, transcript Id, gene name&quot;&gt;
1       12078   .       G       A       25.69   PASS    AC=2;AF=0.048;CLOSEST=0,exon_1_11869_12227,ENST00000456328,DDX11L1
1       16097   .       T       G       42.42   PASS    AC=9;AF=0.0113;CLOSEST=150,exon_1_15796_15947,ENST00000423562,WASH7P
1       40261   .       C       A       366.26  PASS    AC=30;AF=0.484;CLOSEST=4180,exon_1_35721_36081,ENST00000417324,FAM138A
1       63880   .       C       T       82.13   PASS    AC=10;AF=0.0400;CLOSEST=0,exon_1_62948_63887,ENST00000492842,OR4G11P
</code></pre></div></p>
<p>For instance, in the third line (1:16097 T G), it added the tag <code>CLOSEST=150,exon_1_15796_15947,ENST00000423562,WASH7P</code>
, which means that the variant is 150 bases away from exon "exon_1_15796_15947".
The exon belongs to transcript "ENST00000423562" of gene "WASH7P".</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>If multiple markers are available (at the same distance) the one belonging to the longest mRna transcript is shown.</p>
</div>
<p>The input can also be a BED file, the output file has the same information as CLOSEST info field, added to the fourth column of the output BED file:
<div class="highlight"><pre><span></span><code>$ snpeff closest -bed GRCh37.66 test.bed
1   12077   12078   line_1;0,exon_1_11869_12227,ENST00000456328,DDX11L1
1   16096   16097   line_2;150,exon_1_15796_15947,ENST00000423562,WASH7P
1   40260   40261   line_3;4180,exon_1_35721_36081,ENST00000417324,FAM138A
1   63879   63880   line_4;0,exon_1_62948_63887,ENST00000492842,OR4G11P
</code></pre></div></p>
<h3 id="snpeff-count">SnpEff count</h3>
<p>As the name suggests, <code>snpEff count</code> command counts how many reads and bases from a BAM file hit a gene, transcript, exon, intron, etc.
Input files can be in BAM, SAM, VCF, BED or BigBed formats.</p>
<p>A summary HTML file with charts is generated. Here are some examples:</p>
<p><img alt="snpeff_count_01" class="smallerimg center" src="../../images/snpeff_count_01.png" /></p>
<p><img alt="snpeff_count_02" class="smallerimg center" src="../../images/snpeff_count_02.png" /></p>
<p>If you need to count how many reads (and bases) from a BAM file hit each genomic region, you can use 'count' utility.</p>
<p>The command line is quite simple. E.g. in order to count how many reads (from N BAM files) hit regions of the human genome, you simply run:</p>
<div class="codehilite"><pre><span></span><code>java -Xmx8g -jar snpEff.jar count GRCh37.68 readsFile_1.bam readsFile_2.bam ...  readsFile_N.bam &gt; countReads.txt
</code></pre></div>

<p>The output is a TXT (tab-separated) file, that looks like this:
<div class="highlight"><pre><span></span><code>chr  start  end       type                IDs                         Reads:readsFile_1.bam  Bases:readsFile_1.bam  Reads:readsFile_2.bam  Bases:readsFile_2.bam ...
1    1      11873     Intergenic          DDX11L1                     130                    6631                   50                     2544
1    1      249250621 Chromosome          1                           2527754                251120400              2969569                328173439
1    6874   11873     Upstream            NR_046018;DDX11L1           130                    6631                   50                     2544
1    9362   14361     Downstream          NR_024540;WASH7P            243                    13702                  182                    9279
1    11874  12227     Exon                exon_1;NR_046018;DDX11L1    4                      116                    2                      102
1    11874  14408     Gene                DDX11L1                     114                    7121                   135                    6792
1    11874  14408     Transcript          NR_046018;DDX11L1           114                    7121                   135                    6792
1    12228  12229     SpliceSiteDonor     exon_1;NR_046018;DDX11L1    3                      6                      0                      0
1    12228  12612     Intron              intron_1;NR_046018;DDX11L1  13                     649                    0                      0
1    12611  12612     SpliceSiteAcceptor  exon_2;NR_046018;DDX11L1    0                      0                      0                      0
1    12613  12721     Exon                exon_2;NR_046018;DDX11L1    3                      24                     1                      51
1    12722  12723     SpliceSiteDonor     exon_2;NR_046018;DDX11L1    3                      6                      0                      0
1    12722  13220     Intron              intron_2;NR_046018;DDX11L1  22                     2110                   20                     987
1    13219  13220     SpliceSiteAcceptor  exon_3;NR_046018;DDX11L1    5                      10                     1                      2
1    13221  14408     Exon                exon_3;NR_046018;DDX11L1    82                     4222                   113                    5652
1    14362  14829     Exon                exon_11;NR_024540;WASH7P    37                     1830                   7                      357
1    14362  29370     Transcript          NR_024540;WASH7P            704                    37262                  524                    34377
1    14362  29370     Gene                WASH7P                      704                    37262                  524                    34377
1    14409  19408     Downstream          NR_046018;DDX11L1           122                    7633                   39                     4254
</code></pre></div>
The columns are:</p>
<ul>
<li>Column 1: Chromosome name</li>
<li>Column 2: Genomic region start</li>
<li>Column 3: Genomic region end</li>
<li>Column 4: Genomic region type (e.g. Exon, Gene, SpliceSiteDonor, etc.)</li>
<li>Column 5: ID (e.g. exon ID ; transcript ID; gene ID)</li>
<li>Column 6: Count of reads (in file readsFile_1.bam) intersecting genomic region.</li>
<li>Column 7: Count of bases (in file readsFile_1.bam) intersecting genomic region, i.e. each read is intersected and the resulting number of bases added.</li>
<li>Column ...:  (repeat count reads and bases for each BAM file provided)</li>
</ul>
<p><strong>Totals and Binomial model</strong></p>
<p>Using command line option <code>-p</code>, you can calculate p-values based on a Binomial model.
For example (output edited for the sake of brevity):
<div class="highlight"><pre><span></span><code>$ java -Xmx8g -jar snpEff.jar count -v BDGP5.69 fly.bam &gt; countReads.txt
00:00:00.000    Reading configuration file &#39;snpEff.config&#39;
...
00:00:12.148    Calculating probability model for read length 50
...
type               p.binomial             reads.fly  expected.fly  pvalue.fly
Chromosome         1.0                    205215     205215        1.0
Downstream         0.29531659795589793    59082      60603         1.0
Exon               0.2030262729897713     53461      41664         0.0
Gene               0.49282883664487515    110475     101136        0.0
Intergenic         0.33995644860241336    54081      69764         0.9999999963234701
Intron             0.3431415343615103     72308      70418         9.06236369003514E-19
RareAminoAcid      9.245222303207472E-7   3          0             9.879186871519377E-4
SpliceSiteAcceptor 0.014623209601955131   3142       3001          0.005099810118785825
SpliceSiteDonor    0.015279075154423956   2998       3135          0.9937690786738507
Transcript         0.49282883664487515    110475     101136        0.0
Upstream           0.31499087549896493    64181      64641         0.9856950416729887
Utr3prime          0.03495370828296416    8850       7173          1.1734134297889064E-84
Utr5prime          0.02765432673262785    8146       5675          7.908406840800345E-215
</code></pre></div></p>
<p>The columns in for this table are (in the previous example the input file was 'fly.bam' so fileName is 'fly'):</p>
<ul>
<li>type : Type of interval</li>
<li>p.binomial : Probability that a random read hits this 'type' of interval (in binomial model)</li>
<li>reads.fileName : Total number of reads in 'fileName' (user provided BAM/SAM file)</li>
<li>expected.fileName : Expected number of reads hitting this 'type' of interval (for user provided BAM/SAM file)</li>
<li>pvalue.fileName : p-value that 'reads.fileName' reads or more hit this 'type' of interval (for user provided BAM/SAM file)</li>
<li>Column ...:  (repeat last three column for each BAM/SAM file provided by the user)</li>
</ul>
<p><strong>User defined intervals</strong></p>
<p>You can add user defined intervals using <code>-i file.bed</code> command line option.
The option can be used multiple times, thus allowing multiple BED files to be added.</p>
<p>Example : You want to know how many reads intersect each peak from a peak detection algorithm:</p>
<div class="codehilite"><pre><span></span><code>java -Xmx8g -jar snpEff.jar count -i peaks.bed GRCh37.68 reads.bam
</code></pre></div>

<h3 id="snpeff-databases">SnpEff databases</h3>
<p>This command provides a list of configured databases, i.e. available in <code>snpEff.config</code> file.</p>
<p>Example:
<div class="highlight"><pre><span></span><code>$ java -jar snpEff.jar databases
Genome                                                      Organism                                                    Status    Bundle                        Database download link
------                                                      --------                                                    ------    ------                        ----------------------
129S1_SvImJ_v1.99                                           Mus_musculus_129s1svimj                                                                             https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_129S1_SvImJ_v1.99.zip
AIIM_Pcri_1.0.99                                            Pavo_cristatus                                                                                      https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AIIM_Pcri_1.0.99.zip
AKR_J_v1.99                                                 Mus_musculus_akrj                                                                                   https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AKR_J_v1.99.zip
AP006557.1                                                  SARS coronavirus TWH genomic RNA, complete genome.                                                                  https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AP006557.1.zip
AP006558.1                                                  SARS coronavirus TWJ genomic RNA, complete genome.                                                                  https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AP006558.1.zip
AP006559.1                                                  SARS coronavirus TWK genomic RNA, complete genome.                                                                  https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AP006559.1.zip
AP006560.1                                                  SARS coronavirus TWS genomic RNA, complete genome.                                                                  https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AP006560.1.zip
AP006561.1                                                  SARS coronavirus TWY genomic RNA, complete genome.                                                                  https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_AP006561.1.zip
...
</code></pre></div></p>
<h3 id="snpeff-download">SnpEff download</h3>
<p>This command downloads and installs a database.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that the database must be configured in <code>snpEff.config</code> and available at the download site.</p>
</div>
<p>Example: Download and install C.Elegans genome:
<div class="highlight"><pre><span></span><code>$ java -jar snpEff.jar download -v WBcel215.69
00:00:00.000    Downloading database for &#39;WBcel215.69&#39;
00:00:00.002    Connecting to http://downloads.sourceforge.net/project/snpeff/databases/v3_1/snpEff_v3_1_WBcel215.69.zip
00:00:00.547    Copying file (type: application/zip, modified on: Sat Dec 01 20:59:55 EST 2012)
00:00:00.547    Local file name: &#39;snpEff_v3_1_WBcel215.69.zip&#39;
00:00:01.949    Downloaded 1049506 bytes
00:00:03.624    Downloaded 2135266 bytes
00:00:05.067    Downloaded 3185026 bytes
00:00:06.472    Downloaded 4234786 bytes
00:00:07.877    Downloaded 5284546 bytes
00:00:09.580    Downloaded 6374626 bytes
00:00:11.005    Downloaded 7424386 bytes
00:00:12.410    Downloaded 8474146 bytes
00:00:13.815    Downloaded 9523906 bytes
00:00:15.358    Downloaded 10604226 bytes
00:00:16.761    Downloaded 11653666 bytes
00:00:18.168    Downloaded 12703426 bytes
00:00:19.573    Downloaded 13753186 bytes
00:00:21.198    Downloaded 14837506 bytes
00:00:22.624    Downloaded 15887266 bytes
00:00:24.029    Downloaded 16937026 bytes
00:00:25.434    Downloaded 17986786 bytes
00:00:26.864    Downloaded 19036546 bytes
00:00:28.269    Downloaded 20086306 bytes
00:00:29.155    Donwload finished. Total 20748168 bytes.
00:00:29.156    Local file name: &#39;/home/pcingola//snpEff/data/WBcel215.69/snpEffectPredictor.bin&#39;
00:00:29.156    Extracting file &#39;data/WBcel215.69/snpEffectPredictor.bin&#39; to &#39;/home/pcingola//snpEff/data/WBcel215.69/snpEffectPredictor.bin&#39;
00:00:29.157    Creating local directory: &#39;/home/pcingola/snpEff/data/WBcel215.69&#39;
00:00:29.424    Unzip: OK
00:00:29.424    Done
</code></pre></div></p>
<h3 id="snpeff-dump">SnpEff dump</h3>
<p>Dump the contents of a database to a text file, a BED file or a tab separated TXT file (that can be loaded into R).</p>
<p><strong>BED file example</strong>:
<div class="highlight"><pre><span></span><code>$ java -jar snpEff.jar download -v GRCh37.70
$ java -Xmx8g -jar snpEff.jar dump -v -bed GRCh37.70 &gt; GRCh37.70.bed
00:00:00.000    Reading database for genome &#39;GRCh37.70&#39; (this might take a while)
00:00:32.476    done
00:00:32.477    Building interval forest
00:00:45.928    Done.
</code></pre></div></p>
<p>The output file looks like a typical BED file (chr \t start \t end \t name).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Keep in mind that BED file coordinates are zero based, semi-open intervals.
So a 2 base interval at (one-based) positions 100 and 101 is expressed as a BED interval <code>[99 - 101]</code>.</p>
</div>
<div class="highlight"><pre><span></span><code>$ head GRCh37.70.bed
1   0   249250621   Chromosome_1
1   111833483   111863188   Gene_ENSG00000134216
1   111853089   111863002   Transcript_ENST00000489524
1   111861741   111861861   Cds_CDS_1_111861742_111861861
1   111861948   111862090   Cds_CDS_1_111861949_111862090
1   111860607   111860731   Cds_CDS_1_111860608_111860731
1   111861114   111861300   Cds_CDS_1_111861115_111861300
1   111860305   111860427   Cds_CDS_1_111860306_111860427
1   111862834   111863002   Cds_CDS_1_111862835_111863002
1   111853089   111853114   Utr5prime_exon_1_111853090_111853114
</code></pre></div>
<p><strong>TXT file example</strong>:
<div class="highlight"><pre><span></span><code>$ java -Xmx8g -jar snpEff.jar dump -v -txt GRCh37.70 &gt; GRCh37.70.txt
00:00:00.000    Reading database for genome &#39;GRCh37.70&#39; (this might take a while)
00:00:31.961    done
00:00:31.962    Building interval forest
00:00:45.467    Done.
</code></pre></div>
In this case, the output file looks like a typical BED file (chr \t start \t end \t name):
<div class="highlight"><pre><span></span><code>$ head GRCh37.70.txt
chr start       end        strand  type         id                          geneName  geneId            numberOfTranscripts  canonicalTranscriptLength  transcriptId     cdsLength  numerOfExons  exonRank  exonSpliceType
1   1           249250622  +1      Chromosome   1                                                                                                       
1   111833484   111863189  +1      Gene         ENSG00000134216             CHIA      ENSG00000134216   10                   1431                                                                  
1   111853090   111863003  +1      Transcript   ENST00000489524             CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9                     
1   111861742   111861862  +1      Cds          CDS_1_111861742_111861861   CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9                     
1   111861949   111862091  +1      Cds          CDS_1_111861949_111862090   CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9                     
1   111853090   111853115  +1      Utr5prime    exon_1_111853090_111853114  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             1         ALTTENATIVE_3SS
1   111854311   111854341  +1      Utr5prime    exon_1_111854311_111854340  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             2         SKIPPED
1   111860608   111860732  +1      Exon         exon_1_111860608_111860731  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             5         RETAINED
1   111853090   111853115  +1      Exon         exon_1_111853090_111853114  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             1         ALTTENATIVE_3SS
1   111861742   111861862  +1      Exon         exon_1_111861742_111861861  CHIA      ENSG00000134216   10                   1431                       ENST00000489524  862        9             7         RETAINED
</code></pre></div></p>
<p>The format is:</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td>chr</td>
<td>Chromosome name</td>
</tr>
<tr>
<td>start</td>
<td>Marker start (one-based coordinate)</td>
</tr>
<tr>
<td>end</td>
<td>Marker end (one-based coordinate)</td>
</tr>
<tr>
<td>strand</td>
<td>Strand (positive or negative)</td>
</tr>
<tr>
<td>type</td>
<td>Type of marker (e.g. exon, transcript, etc.)</td>
</tr>
<tr>
<td>id</td>
<td>ID. E.g. if it's a Gene, then it may be ENSEBML's gene ID</td>
</tr>
<tr>
<td>geneName</td>
<td>Gene name, if marker is within a gene (exon, transcript, UTR, etc.), empty otherwise (e.g. intergenic)</td>
</tr>
<tr>
<td>geneId</td>
<td>Gene is, if marker is within a gene</td>
</tr>
<tr>
<td>numberOfTranscripts</td>
<td>Number of transcripts in the gene</td>
</tr>
<tr>
<td>canonicalTranscriptLength</td>
<td>CDS length of canonical transcript</td>
</tr>
<tr>
<td>transcriptId</td>
<td>Transcript ID, if marker is within a transcript</td>
</tr>
<tr>
<td>cdsLength</td>
<td>CDS length of the transcript</td>
</tr>
<tr>
<td>numerOfExons</td>
<td>Number of exons in this transcript</td>
</tr>
<tr>
<td>exonRank</td>
<td>Exon rank, if marker is an exon</td>
</tr>
<tr>
<td>exonSpliceType</td>
<td>Exon splice type, if marker is an exon</td>
</tr>
</tbody>
</table>
<h3 id="snpeff-genes2bed">SnpEff genes2bed</h3>
<p>Dumps a selected set of genes as BED intervals.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The functionality of this command is a subset of <code>SnpEff dump</code>, so it is likely to be deprecated in the future.</p>
</div>
<p>Example:
<div class="highlight"><pre><span></span><code>$ java -Xmx8g -jar snpEff.jar genes2bed GRCh37.66 DDX11L1 WASH7P
#chr    start   end geneName;geneId
1   11868   14411   DDX11L1;ENSG00000223972
1   14362   29805   WASH7P;ENSG00000227232
</code></pre></div></p>
<h3 id="snpeff-cds-snpeff-protein">SnpEff cds &amp; SnpEff protein</h3>
<p>These commands perform SnpEff database sanity checks.
They calculate CDS and protein sequences from a SnpEff database and then compare the results to a FASTA file (having the "correct" sequences).</p>
<p>The commands are invoked automatically when building databases, so there is no need for the user to invoke them manually.</p>
<h3 id="snpeff-len">SnpEff len</h3>
<p>Calculates the genomic length of every type of marker (Gene, Exon, Utr, etc.).
Length is calculated by overlapping all markers and counting the number of bases (e.g. a base is counted as 'Exon' if any exon falls onto that base).
This command also reports the probability of a Binomial model.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>Parameter <code>-r num</code> adjusts the model for a read length of 'num' bases. That is, if two markers of the same type are closer than 'num' bases, it joins them by inclining the bases separating them.</p>
</div>
<p>E.g.:
<div class="highlight"><pre><span></span><code>$ java -Xmx1g -jar snpEff.jar len -r 100 BDGP5.69
marker                   size    count     raw_size raw_count    binomial_p
Cds                  22767006    56955     45406378    122117    0.13492635563570918
Chromosome          168736537       15    168736537        15    1.0
Downstream           49570138     5373    254095562     50830    0.29377240330587084
Exon                 31275946    61419     63230008    138474    0.18535372691689175
Gene                 82599213    11659     87017182     15222    0.4895158717166277
Intergenic           56792611    11637     56792611     11650    0.3365756581812509
Intron               55813748    42701    168836797    113059    0.33077452573297744
SpliceSiteAcceptor      97977    48983       226118    113059    5.806507691929223E-4
SpliceSiteDonor        101996    50981       226118    113059    6.044689657225808E-4
Transcript           82599213    11659    232066805     25415    0.4895158717166277
Upstream             52874082     5658    254044876     50830    0.3133528928592389
Utr3prime             5264120    13087     10828991     24324    0.031197274126824114
Utr5prime             3729197    19324      6368070     33755    0.02210070839607192
</code></pre></div>
Column meaning:</p>
<ul>
<li>marker : Type of marker interval</li>
<li>size : Size of all intervals in the genome, after overlap and join.</li>
<li>count : Number of intervals in the genome, after overlap and join.</li>
<li>raw_size : Size of all intervals in the genome. Note that this could be larger than the genome.</li>
<li>raw_count : Number of intervals in the genome.</li>
</ul>
<h3 id="scripts">Scripts</h3>
<p>Several scripts are provided in the <code>scripts</code> directory.
Here we briefly describe their functionality:</p>
<h4 id="sam2fastqpl">sam2fastq.pl</h4>
<p>Convert a SAM input to a FASTQ output. Example:
<div class="highlight"><pre><span></span><code>samtools view test.bam | ./scripts/bam2fastq.pl | head -n 12
@HWI-ST1220:76:D12CHACXX:7:2207:18986:95756
CGACAATGCACGACAGAGGAAGCAGAACAGATATTTAGATTGCCTCTCATT
+
CCCFFFFFGHHHHIIJIJJIJIJJIJJIIIJIIHIJJJIJJIJJJJJJIJJ
@HWI-ST1220:76:D12CHACXX:7:2206:4721:162268
ATATTATAGGGAGAAATATGATCGCGTATGCGAGAGTAGTGCCAACATATT
+
@@@DDD&gt;DAB;?DGGEGGBCD&gt;BFGI?FCFFBFGG@&lt;?B*?BFB9FGII@E
@HWI-ST1220:76:D12CHACXX:7:1304:13069:17740
ATAGGGAGAAATATGATCGCGTATGCGAGAGTAGTGCCAACATATTGTGCT
+
CCCFFFFFHHHHHJJJJJJJJGIJJJJJJAGBGGIIIJJJJJJJJJJJIJG
</code></pre></div></p>
<h4 id="fasta2tabpl">fasta2tab.pl</h4>
<p>Convert a fasta file to a two column tab-separated TXT file (name \t sequence)</p>
<p>Example (output truncated for brevity):
<div class="highlight"><pre><span></span><code>$ zcat ce6.fa.gz | ./scripts/fasta2tab.pl
chrI    gcctaagcctaagcctaagcctaagcctaagcctaagcctaagcct...
chrV    GAATTcctaagcctaagcctaagcctaagcctaagcctaagcctaa...
chrII   cctaagcctaagcctaagcctaagcctaagcctaagcctaagccta...
chrM    CAGTAAATAGTTTAATAAAAATATAGCATTTGGGTTGCTAAGATAT...
chrX    ctaagcctaagcctaagcctaagcctaagcctaagcctaagcctaa...
chrIV   cctaagcctaagcctaagcctaagcctaagcctaagcctaagccta...
chrIII  cctaagcctaagcctaagcctaagcctaagcctaagcctaagccta...
</code></pre></div></p>
<h4 id="fastasplitpl">fastaSplit.pl</h4>
<p>Split a multiple sequence FASTA file to individual files.</p>
<p>Example: Creates one file per chromosome:
<div class="highlight"><pre><span></span><code>$ zcat ce6.fa.gz | ./scripts/fastaSplit.pl
Writing to chrI.fa
Writing to chrV.fa
Writing to chrII.fa
Writing to chrM.fa
Writing to chrX.fa
Writing to chrIV.fa
Writing to chrIII.fa
</code></pre></div></p>
<h4 id="plothistogrampl">plotHistogram.pl</h4>
<p>Given a list of numbers (one per line), shows a histogram. Note: It requires R.</p>
<p>Example: Extract the file sizes in a directory and show a histogram</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>ls<span class="w"> </span>-al<span class="w"> </span>scripts/<span class="w"> </span><span class="p">|</span><span class="w"> </span>tr<span class="w"> </span>-s<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">5</span><span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>./scripts/hist.pl
</code></pre></div>

<p>Creates the following plot:
<img alt="hist_example" class="smallerimg center" src="../../images/hist_example.png" /></p>
<h4 id="plotmapl-plotpl-plotqqpl-plotsmoothscatterpl">plotMA.pl, plot.pl, plotQQ.pl, plotSmoothScatter.pl</h4>
<p>Similar to 'hist.pl', these perform plots based on input from STDIN.
Note that in some cases, inputs are expected to be probabilities (qqplot.pl) or pairs of numbers (maPlot.pl).</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>ls<span class="w"> </span>-al<span class="w"> </span>scripts/<span class="w"> </span><span class="p">|</span><span class="w"> </span>tr<span class="w"> </span>-s<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">5</span><span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot; &quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>./scripts/plot.pl
</code></pre></div>

<p>Creates the following plot:
<img alt="plot_example" class="smallerimg center" src="../../images/plot_example.png" /></p>
<h4 id="queuepl">queue.pl</h4>
<p>Process a list of statements in parallel according to the number of CPUs available in the local machine.</p>
<h4 id="splitchrpl">splitChr.pl</h4>
<p>Splits a file by chromosome. Works on any tab separated file that the first column is CHR field (e.g. BED, VCF, etc.)</p>
<p>Example:
<div class="highlight"><pre><span></span><code>$ cat large_test.vcf | ./scripts/splitChr.pl
Input line 28. Creating file &#39;chr1.txt&#39;
Input line 13332. Creating file &#39;chr2.txt&#39;
Input line 22097. Creating file &#39;chr3.txt&#39;
Input line 29289. Creating file &#39;chr4.txt&#39;
Input line 34236. Creating file &#39;chr5.txt&#39;
Input line 39899. Creating file &#39;chr6.txt&#39;
Input line 47120. Creating file &#39;chr7.txt&#39;
Input line 53371. Creating file &#39;chr8.txt&#39;
Input line 57810. Creating file &#39;chr9.txt&#39;
Input line 63005. Creating file &#39;chr10.txt&#39;
Input line 68080. Creating file &#39;chr11.txt&#39;
Input line 76629. Creating file &#39;chr12.txt&#39;
Input line 83071. Creating file &#39;chr13.txt&#39;
Input line 85124. Creating file &#39;chr14.txt&#39;
Input line 89281. Creating file &#39;chr15.txt&#39;
Input line 93215. Creating file &#39;chr16.txt&#39;
Input line 99081. Creating file &#39;chr17.txt&#39;
Input line 106405. Creating file &#39;chr18.txt&#39;
Input line 108330. Creating file &#39;chr19.txt&#39;
Input line 118568. Creating file &#39;chr20.txt&#39;
Input line 121795. Creating file &#39;chr21.txt&#39;
Input line 123428. Creating file &#39;chr22.txt&#39;
Input line 126520. Creating file &#39;chrX.txt&#39;
Input line 129094. Creating file &#39;chrY.txt&#39;
Input line 129113. Creating file &#39;chrMT.txt&#39;
</code></pre></div></p>
<h4 id="uniqcountpl">uniqCount.pl</h4>
<p>Count number of unique lines. It's the same as doing <code>cat lines.tst | sort | uniq -c</code>, but much faster. Particularly useful for very large inputs.</p>
<h4 id="vcfeffoneperlinepl">vcfEffOnePerLine.pl</h4>
<p>Splits EFF fields in a VCF file, creating multiple lines, each one with only one effect.</p>
<p>Very useful for filtering with SnpSift.</p>
<p>Example:</p>
<div class="highlight"><pre><span></span><code>$ cat test.stop.vcf
1   897062  .   C   T   100.0   PASS    AC=1;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*|642|KLHL17||CODING|NM_198317|),UPSTREAM(MODIFIER||||576|PLEKHN1||CODING|NM_001160184|),UPSTREAM(MODIFIER||||611|PLEKHN1||CODING|NM_032129|),UPSTREAM(MODIFIER||||749|NOC2L||CODING|NM_015658|)

$ cat test.stop.vcf | ./scripts/vcfEffOnePerLine.pl
1   897062  .   C   T   100.0   PASS    AC=1;EFF=STOP_GAINED(HIGH|NONSENSE|Cag/Tag|Q141*|642|KLHL17||CODING|NM_198317|)
1   897062  .   C   T   100.0   PASS    AC=1;EFF=UPSTREAM(MODIFIER||||576|PLEKHN1||CODING|NM_001160184|)
1   897062  .   C   T   100.0   PASS    AC=1;EFF=UPSTREAM(MODIFIER||||611|PLEKHN1||CODING|NM_032129|)
1   897062  .   C   T   100.0   PASS    AC=1;EFF=UPSTREAM(MODIFIER||||749|NOC2L||CODING|NM_015658|)
</code></pre></div>





                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Pablo Cingolani
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="http://www.linkedin.com/in/pablocingolani" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/pcingola/SnpEff" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "search.highlight", "search.suggest", "toc.integrate", "content.tabs.link", "content.code.annotate", "content.code.copy"], "search": "../../assets/javascripts/workers/search.74e28a9f.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.220ee61c.min.js"></script>
      
    
  </body>
</html>